---
title: "Divergence Stratigraphy using orthologr"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Divergence Stratigraphy using orthologr}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r, echo = FALSE, message = FALSE}
options(width = 750)
knitr::opts_chunk$set(
  comment = "#>",
  error = FALSE,
  tidy = FALSE)
```

The `orthologr` package allows you to perform __Divergence Stratigraphy__ for query and subject organisms
of interest. __Divergence Stratigraphy__ is the process of quantifying the selection pressure (in terms of amino acid sequence divergence) acting on orthologous genes between closely related species. The resulting sequence divergence map (short divergence map), stores the divergence stratum in the first column and the query_id of inferred orthologous genes in the second column ([Quint et al., 2012](http://www.nature.com/nature/journal/v490/n7418/full/nature11394.html); Drost et al., 2014).

Following steps are performed to obtain a standard divergence map based on the `divergence_stratigraphy()` function:

1) Orthology Inference using BLAST reciprocal best hit ("RBH") based on blastp

2) Pairwise global amino acid alignments of orthologous genes using the [Needleman-Wunsch algorithm](http://www.sciencedirect.com/science/article/pii/0022283670900574)

3) Codon alignments of orthologous genes using [PAL2NAL](http://www.bork.embl.de/pal2nal/)

4) dNdS estimation using [Comeron's method (1995)](http://link.springer.com/article/10.1007/BF00173196)

5) Assigning estimated dNdS values to divergence strata (deciles of all dNdS values)


In `orthologr` the [Needleman-Wunsch algorithm](http://www.sciencedirect.com/science/article/pii/0022283670900574), [PAL2NAL](http://www.bork.embl.de/pal2nal/) and [Comeron's method (1995)](http://link.springer.com/article/10.1007/BF00173196) 
are already included in the `orthologr` package and don't need to be installed. 

## Fast Installation guide for orthologr

```r
# install orthologr from CRAN
install.packages("orthologr", dependencies = TRUE)

# or the developer version from github
library(devtools)
install_github("HajkD/orthologr", build_vignettes = TRUE, dependencies = TRUE)

# On Windows, this won't work - see ?build_github_devtools
install_github("HajkD/orthologr", build_vignettes = TRUE, dependencies = TRUE)

# When working with Windows, first you need to install the
# R package: rtools -> install.packages("rtools")

# Afterwards you can install devtools -> install.packages("devtools")
# and then you can run:

devtools::install_github("HajkD/orthologr", build_vignettes = TRUE, dependencies = TRUE)

# and then call it from the library
library("orthologr", lib.loc = "C:/Program Files/R/R-3.1.1/library")



```

## Retrieving a divergence map

A divergence map quantifies for each protein coding gene of a given organism the
degree of selection pressure. The selection pressure is quantified by dNdS estimation.

To perform __Divergence Stratigraphy__ using `orthologr` you need the following prerequisites

* a CDS file covering all protein coding genes of the query organism of interest
* a CDS file covering all protein coding genes of the subject organism of interest

In the following example, we will use _Arabidopsis thaliana_ as query organism
and _Arabidopsis lyrata_ as subject organism.

First, we need to download the CDS sequences for all protein coding genes of _A. thaliana_
and _A. lyrata_.

The CDS retrieval can be done using a `Terminal` or by manual downloading the files

* `Arabidopsis_thaliana.TAIR10.23.cds.all.fa.gz`
* `Arabidopsis_lyrata.v.1.0.23.cds.all.fa.gz`

```shell

# download CDS file of A. thaliana
curl ftp://ftp.ensemblgenomes.org/pub/
plants/release-23/fasta/arabidopsis_thaliana/
cds/Arabidopsis_thaliana.TAIR10.23.cds.all.fa.gz 
-o Arabidopsis_thaliana.TAIR10.23.cds.all.fa.gz

# unzip the fasta file
gunzip -d Arabidopsis_thaliana.TAIR10.23.cds.all.fa.gz

# download CDS file of A. lyrata

curl ftp://ftp.ensemblgenomes.org/pub/plants/
release-23/fasta/arabidopsis_lyrata/cds/
Arabidopsis_lyrata.v.1.0.23.cds.all.fa.gz 
-o Arabidopsis_lyrata.v.1.0.23.cds.all.fa.gz

# unzip the fasta file
gunzip -d Arabidopsis_lyrata.v.1.0.23.cds.all.fa.gz

```
When the download is finished you need to unzip the files and
then start R to perform the following analyses:

```{r,eval=FALSE}
library(orthologr)

# compute the divergence map of A. thaliana
Athaliana_DM <- divergence_stratigraphy(
                         query_file = "path/to/Arabidopsis_thaliana.TAIR10.23.cds.all.fa",
                         subject_file = "path/to/Arabidopsis_lyrata.v.1.0.23.cds.all.fa",
                         eval = "1E-5", ortho_detection = "RBH",
                         comp_cores = 1, quiet = TRUE, clean_folders = TRUE )

```

In `divergence_stratigraphy()` the `query_file` and `subject_file` arguments take an
character string storing the path to the corresponding __fasta__ files containing the protein
sequences of these organisms. The `eval` is set to `1E-5` (default ; see [Quint et al., 2012](http://www.nature.com/nature/journal/v490/n7418/full/nature11394.html)) and BLAST reciprocal best hit is used for orthology inference. In case you work
with a multicore machine, you can set the `comp_cores` argument to any number of cores
supported by your machine. The `clean_folders` argument specifies whether the internal folder structure shall be deleted (cleaned) after processing is finished. In this case all output files returned by BLAST and the alignment tools are deleted.

You can run a test __Divergence Stratigraphy__ with 20 example genes stored in the `orthologr` package:

```{r,eval=FALSE}
library(orthologr)

# performing standard divergence stratigraphy
 divergence_stratigraphy(
      query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
      subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
      eval = "1E-5", ortho_detection = "RBH", dnds.threshold = 1,
      comp_cores = 1, quiet = TRUE, clean_folders = TRUE)

```


```

   divergence_strata    query_id
1                 10 AT1G01010.1
2                  9 AT1G01020.1
3                  5 AT1G01030.1
4                  4 AT1G01040.1
5                  1 AT1G01050.1
6                  9 AT1G01060.3
7                  6 AT1G01070.1
8                  8 AT1G01080.1
9                  2 AT1G01090.1
10                 7 AT1G01110.2
11                 2 AT1G01120.1
12                 3 AT1G01140.3
13                10 AT1G01150.1
14                 8 AT1G01160.1
15                 1 AT1G01170.2
16                 6 AT1G01180.1
17                 7 AT1G01190.1
18                 4 AT1G01200.1
19                 5 AT1G01210.1
20                 3 AT1G01220.1

```


Please note that performing __Divergence Stratigraphy__ using two large genomes might take
(even on a multicore machine) up to several minutes or hours (depending on the number of available computing cores).
On a 8 core machine with 3.4 GHz i7 processors the computation of an divergence map between _A. thaliana_ and _A. lyrata_
took 65 minutes.

When your diverge map computations are done and you have a divergence map of your query organism,
subsequent _phylotranscriptomics_ analyses can be done using the [myTAI](http://cran.r-project.org/web/packages/myTAI/index.html) package.

Please see [Introduction to the myTAI package](http://cran.r-project.org/web/packages/myTAI/vignettes/Introduction.html)
for more details.


In case you need to specify the PATH to your BLAST folder, you can use the `blast_path` argument:

```{r,eval=FALSE}
library(orthologr)

# performing standard divergence stratigraphy
 divergence_stratigraphy(
      query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
      subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
      blast_path = "here/the/path/to/blastp", eval = "1E-5", ortho_detection = "RBH", 
      dnds.threshold = 1, comp_cores = 1, quiet = TRUE, clean_folders = TRUE)

```


```

   divergence_strata    query_id
1                 10 AT1G01010.1
2                  9 AT1G01020.1
3                  5 AT1G01030.1
4                  4 AT1G01040.1
5                  1 AT1G01050.1
6                  9 AT1G01060.3
7                  6 AT1G01070.1
8                  8 AT1G01080.1
9                  2 AT1G01090.1
10                 7 AT1G01110.2
11                 2 AT1G01120.1
12                 3 AT1G01140.3
13                10 AT1G01150.1
14                 8 AT1G01160.1
15                 1 AT1G01170.2
16                 6 AT1G01180.1
17                 7 AT1G01190.1
18                 4 AT1G01200.1
19                 5 AT1G01210.1
20                 3 AT1G01220.1

```

### divergence_stratigraphy() and dNdS-maps

In case you are interested in the dNdS values of corresponding proteome comparisons returned by `divergence_stratigraphy`, you can either use the `dNdS()` function to obtain dNdS values for all protein coding genes of a given query and subject proteome or you can specify the `ds.values` argument: `ds.values = FALSE`.

```{r,eval=FALSE}
library(orthologr)

# performing standard divergence stratigraphy
 divergence_stratigraphy(
      query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
      subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
      eval = "1E-5", ortho_detection = "RBH",ds.values = FALSE, dnds.threshold = 1,
      comp_cores = 1, quiet = TRUE, clean_folders = TRUE)

```


```
      dNdS    query_id
1  0.41950 AT1G01010.1
2  0.38790 AT1G01020.1
3  0.11850 AT1G01030.1
4  0.11560 AT1G01040.1
5  0.00000 AT1G01050.1
6  0.39670 AT1G01060.3
7  0.17280 AT1G01070.1
8  0.32170 AT1G01080.1
9  0.04174 AT1G01090.1
10 0.26620 AT1G01110.2
11 0.02317 AT1G01120.1
12 0.04324 AT1G01140.3
13 0.64120 AT1G01150.1
14 0.37310 AT1G01160.1
15 0.00000 AT1G01170.2
16 0.16830 AT1G01180.1
17 0.17730 AT1G01190.1
18 0.11370 AT1G01200.1
19 0.13420 AT1G01210.1
20 0.10230 AT1G01220.1

```

The corresponding output is denoted as `dNdS-map`.