---
title: "dNdS estimation using orthologr"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{dNdS estimation using orthologr}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r, echo = FALSE, message = FALSE}
library(orthologr)
options(width = 750)
knitr::opts_chunk$set(
  comment = "#>",
  error = FALSE,
  tidy = FALSE)
```

The dN/dS ratio quantifies the mode and strength of selection acting on a pair of orthologous genes.
This selection pressure can be quantified by comparing synonymous substitution rates (dS) that are assumed to be neutral
with nonsynonymous substitution rates (dN), which are exposed to selection as they 
change the amino acid composition of a protein ([Mugal et al., 2013](http://mbe.oxfordjournals.org/content/31/1/212)).

The `orthologr` package provides a function named `dNdS()` to perform dNdS estimation on pairs of orthologous genes.
The `dNdS()` function takes the CDS files of two organisms of interest (`query_file` and `subject_file`) 
and computes the dNdS estimation values for orthologous gene pairs between these organisms.

The following pipieline resembles the dNdS estimation process:

 1) Orthology Inference: e.g. BLAST reciprocal best hit (RBH)
 2) Pairwise sequence alignment: e.g. clustalw for pairwise amino acid sequence alignments
 3) Codon Alignment: e.g. pal2nal program
 4) dNdS estimation: e.g. [Yang, Z. and Nielsen, R. (2000)](http://mbe.oxfordjournals.org/content/17/1/32.short) (YN)


__Note:__ it is assumed that when using `dNdS()` all corresponding multiple sequence alignment programs you
want to use are already installed on your machine and are executable via either
the default execution PATH or you specifically define the location of the executable file
via the `aa_aln_path` or `blast_path` argument that can be passed to `dNdS()`.

The following example shall illustrate a dNdS estimation process.


```{r,eval=FALSE}

library(orthologr)

# get a dNdS table using:
# 1) reciprocal best hit for orthology inference (RBH)
# 2) clustalw for pairwise amino acid alignments
# 3) pal2nal for codon alignments
# 4) Yang, Z. and Nielsen, R. (2000) (YN) for dNdS estimation
# 5) single core processing 'comp_cores = 1'
dNdS(query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
ortho_detection = "RBH", aa_aln_type = "multiple",
aa_aln_tool = "clustalw", codon_aln_tool = "pal2nal", 
dnds_est.method = "YN", comp_cores = 1, clean_folders = TRUE, quiet = TRUE)


```

The output includes `NA` values. To filter for `NA` values or a specific `dnds.threshold`,
you can use the `filter_dNdS()` function. The `filter_dNdS()` function takes the output data.table 
returned by `dNdS()` and filters the output by the following criteria:

1) all dN values having an NA value or omitted

2) all dS values having an NA value or omitted

3) all dNdS values >= the specified dnds.threshold are omitted


```{r,eval=FALSE}

library(orthologr)

# get dNdS estimated for orthologous genes between A. thaliana and A. lyrata

Ath_Aly_dnds <- dNdS(
                  query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
                  subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
                  dnds_est.method = "YN", comp_cores = 1, 
                  clean_folders = TRUE, quiet = TRUE
                  )

        
# filter for:
# 1) all dN values having an NA value or omitted
# 2) all dS values having an NA value or omitted
# 3) all dNdS values >= 2 are omitted

filter_dNdS(Ath_Aly_dnds, dnds.threshold = 2)


```


Instead if using a multiple alignment tool for pairwise alignments you can also
choose a global pairwise alignment of orthologous genes based on the Needleman-Wunsch algorithm.
For this purpose the argument `aa_aln_type` must be set to `aa_aln_type = "pairwise"` and `aa_aln_tool = "NW"` 
for Needleman-Wunsch.

```{r,eval=FALSE}

library(orthologr)

# get a dNdS table using:
# 1) reciprocal best hit for orthology inference (RBH)
# 2) pairwise amino acid alignments using Needleman-Wunsch
# 3) pal2nal for codon alignments
# 4) Comeron (1995) for dNdS estimation
# 5) single core processing 'comp_cores = 1'
dNdS(query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
ortho_detection = "RBH", aa_aln_type = "pairwise",
aa_aln_tool = "NW", codon_aln_tool = "pal2nal", 
dnds_est.method = "Comeron", comp_cores = 1, clean_folders = TRUE, quiet = TRUE)


```

The `dNdS()` function can be used choosing the folllowing options:

* `ortho_detection` : `RBH` (BLAST best reciprocal hit), `BH` (BLAST best reciprocal hit), `PO` (ProteinOrtho), `OrthoMCL` (OrthoMCL), `IP` (InParanoid)
* `aa_aln_type` : `multiple` or `pairwise`
* `aa_aln_tool` : `clustalw`, `t_coffee`, `muscle`, `clustalo`, `mafft`, and `NW` (in case `aa_aln_type = "pairwise"`)
* `codon_aln_tool` : `pal2nal`
* `dnds_est.method` : `Li`, `Comeron`, `NG`, `LWL`, `LPB`, `MLWL`, `YN`, and `MYN`

Please see `?dNdS` for details.

In case your BLAST program, or multiple alignment program can not be executed from the default execution PATH
you can specify the `aa_aln_path` or `blast_path` arguments.

```{r,eval=FALSE}

library(orthologr)

# using the `aa_aln_path` or `blast_path` arguments
dNdS(query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
ortho_detection = "RBH",blast_path = "here/path/to/blastp",
aa_aln_type = "multiple", aa_aln_tool = "clustalw", 
aa_aln_path = "here/path/to/clustalw",
codon_aln_tool = "pal2nal", dnds_est.method = "Comeron", 
comp_cores = 1, clean_folders = TRUE, quiet = TRUE)


```


## Advanced options

Additional arguments that can be passed to `dNdS()` allow to use more advanced 
options of several interface programs.

To pass additional parameters to the interface programs, you can use the
`blast_params` and `aa_aln_params`. The `aa_aln_params` argument assumes
that when you chose e.g. `aa_aln_tool = "mafft"` you will pass the corresponding
addtional paramers in MAFFT notation.

```{r,eval=FALSE}

library(orthologr)

# get dNdS estimated for orthologous genes between A. thaliana and A. lyrata
# using additional parameters:

# get a dNdS table using:
# 1) reciprocal best hit for orthology inference (RBH)
# 2) multiple amino acid alignments using MAFFT
# 3) pal2nal for codon alignments
# 4) Comeron (1995) for dNdS estimation
# 5) single core processing 'comp_cores = 1'
Ath_Aly_dnds <- dNdS(
                  query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
                  subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
                  ortho_detection = "RBH",blast_params = "-matrix BLOSUM80",
                  aa_aln_tool = "mafft",aa_aln_params = "--maxiterate 1 --clustalout",
                  dnds_est.method = "Comeron", comp_cores = 1, clean_folders = TRUE, quiet = TRUE
                  )

        
# filter for:
# 1) all dN values having an NA value or omitted
# 2) all dS values having an NA value or omitted
# 3) all dNdS values >= 0.1 are omitted

filter_dNdS(Ath_Aly_dnds, dnds.threshold = 0.1)


```

Here `blast_params` and `aa_aln_params` take an character string specifying the parameters
that shall passed to BLAST and MAFFT. The notation of these parameters must follow the 
command line call of the stand alone versions of BLAST and MAFFT: e.g. `blast_params = "blast_params = -matrix BLOSUM80"`
and `aa_aln_params = "--maxiterate 1 --clustalout"`.

