---
title: "dNdS estimation using orthologr"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Estimating dNdS}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r, echo = FALSE, message = FALSE}
options(width = 750)
knitr::opts_chunk$set(
  comment = "#>",
  error = FALSE,
  tidy = FALSE)
```

The dN/dS ratio quantifies the mode and strength of selection acting on a pair of orthologous genes.
This selection pressure can be quantified by comparing synonymous substitution rates (dS) that are assumed to be neutral
with nonsynonymous substitution rates (dN), which are exposed to selection as they 
change the amino acid composition of a protein ([Mugal et al., 2013](http://mbe.oxfordjournals.org/content/31/1/212)).

The `orthologr` package provides a function named `dNdS()` to perform dNdS estimation on pairs of orthologous genes.
The `dNdS()` function takes the CDS files of two organisms of interest (`query_file` and `subject_file`) 
and computes the dNdS estimation values for orthologous gene pairs between these organisms.

__Note:__ the following dNdS estimation methods are based on KaKs_Calculator:

* "NG": Nei, M. and Gojobori, T. (1986)

* "LWL": Li, W.H., et al. (1985)

* "LPB": Li, W.H. (1993) and Pamilo, P. and Bianchi, N.O. (1993)

* "MLWL" (Modified LWL), MLPB (Modified LPB): Tzeng, Y.H., et al. (2004)

* "YN": Yang, Z. and Nielsen, R. (2000)

* "MYN" (Modified YN): Zhang, Z., et al. (2006)

It is assumed that when you choose one of these dNdS estimation methods you have
KaKs_Calculator installed on your machine and it can be executed from the
default execution `PATH`.

The following pipeline resembles and example dNdS estimation procedure:

 1) Orthology Inference: e.g. BLAST reciprocal best hit (RBH)
 2) Pairwise sequence alignment: e.g. clustalw for pairwise amino acid sequence alignments
 3) Codon Alignment: e.g. pal2nal program
 4) dNdS estimation: e.g. [Yang, Z. and Nielsen, R. (2000)](http://mbe.oxfordjournals.org/content/17/1/32.short) (YN)


__Note:__ it is assumed that when using `dNdS()` all corresponding multiple sequence alignment programs you
want to use are already installed on your machine and are executable via either
the default execution PATH or you specifically define the location of the executable file
via the `aa_aln_path` or `blast_path` argument that can be passed to `dNdS()`.

The following example shall illustrate a dNdS estimation process.


```{r,eval=FALSE}

library(orthologr)

# get a dNdS table using:
# 1) reciprocal best hit for orthology inference (RBH)
# 2) clustalw for pairwise amino acid alignments
# 3) pal2nal for codon alignments
# 4) Yang, Z. and Nielsen, R. (2000) (YN) for dNdS estimation
# 5) single core processing 'comp_cores = 1'
dNdS(query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
     subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
     ortho_detection = "RBH", aa_aln_type = "multiple",
     aa_aln_tool = "clustalw", codon_aln_tool = "pal2nal", 
     dnds_est.method = "YN", comp_cores = 1, clean_folders = TRUE, quiet = TRUE)


```


```
       query_id            subject_id         dN        dS      dNdS method
 1: AT1G01010.1 333554|PACid_16033839 0.10581700 0.2844350 0.3720250     YN
 2: AT1G01020.1 470181|PACid_16064328 0.04164150 0.0951677 0.4375590     YN
 3: AT1G01030.1 470180|PACid_16054974 0.01664670 0.1163900 0.1430260     YN
 4: AT1G01040.1 333551|PACid_16057793 0.01421700 0.1314360 0.1081670     YN
 5: AT1G01050.1 909874|PACid_16064489         NA 0.2092450 0.0000000     YN
 6: AT1G01060.3 470177|PACid_16043374 0.04387800 0.1131710 0.3877130     YN
 7: AT1G01070.1 918864|PACid_16052578 0.02028020 0.0960773 0.2110820     YN
 8: AT1G01080.1 909871|PACid_16053217 0.03930610 0.0995795 0.3947210     YN
 9: AT1G01090.1 470171|PACid_16052860 0.00992436 0.2496940 0.0397461     YN
10: AT1G01110.2 333544|PACid_16034284 0.03292970 0.1293160 0.2546450     YN
11: AT1G01120.1 918858|PACid_16049140 0.00356132 0.1225410 0.0290623     YN
12: AT1G01140.3 470161|PACid_16036015 0.00582238 0.1354990 0.0429699     YN
13: AT1G01150.1 918855|PACid_16037307 0.13565500 0.1962460 0.6912480     YN
14: AT1G01160.1 918854|PACid_16044153 0.11558300 0.1929560 0.5990120     YN
15: AT1G01170.2 311317|PACid_16052302 0.00557175 0.2903370 0.0191906     YN
16: AT1G01180.1 909860|PACid_16056125 0.04065370 0.1557400 0.2610360     YN
17: AT1G01190.1 311315|PACid_16059488 0.02849220 0.1538610 0.1851810     YN
18: AT1G01200.1 470156|PACid_16041002 0.01983450 0.1512510 0.1311360     YN
19: AT1G01210.1 311313|PACid_16057125 0.02106910 0.1433630 0.1469630     YN
20: AT1G01220.1 470155|PACid_16047984 0.01530070 0.1446480 0.1057780     YN

```
The output includes `NA` values. To filter for `NA` values or a specific `dnds.threshold`,
you can use the `filter_dNdS()` function. The `filter_dNdS()` function takes the output data.table 
returned by `dNdS()` and filters the output by the following criteria:

1) all dN values having an NA value or omitted

2) all dS values having an NA value or omitted

3) all dNdS values >= the specified dnds.threshold are omitted


```{r,eval=FALSE}

library(orthologr)

# get dNdS estimated for orthologous genes between A. thaliana and A. lyrata

Ath_Aly_dnds <- dNdS(
                  query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
                  subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
                  dnds_est.method = "YN", comp_cores = 1, 
                  clean_folders = TRUE, quiet = TRUE
                  )

        
# filter for:
# 1) all dN values having an NA value or omitted
# 2) all dS values having an NA value or omitted
# 3) all dNdS values >= 2 are omitted

filter_dNdS(Ath_Aly_dnds, dnds.threshold = 2)


```

```

      query_id            subject_id         dN        dS      dNdS method
1  AT1G01010.1 333554|PACid_16033839 0.10581700 0.2844350 0.3720250     YN
2  AT1G01020.1 470181|PACid_16064328 0.04164150 0.0951677 0.4375590     YN
3  AT1G01030.1 470180|PACid_16054974 0.01664670 0.1163900 0.1430260     YN
4  AT1G01040.1 333551|PACid_16057793 0.01421700 0.1314360 0.1081670     YN
5  AT1G01060.3 470177|PACid_16043374 0.04387800 0.1131710 0.3877130     YN
6  AT1G01070.1 918864|PACid_16052578 0.02028020 0.0960773 0.2110820     YN
7  AT1G01080.1 909871|PACid_16053217 0.03930610 0.0995795 0.3947210     YN
8  AT1G01090.1 470171|PACid_16052860 0.00992436 0.2496940 0.0397461     YN
9  AT1G01110.2 333544|PACid_16034284 0.03292970 0.1293160 0.2546450     YN
10 AT1G01120.1 918858|PACid_16049140 0.00356132 0.1225410 0.0290623     YN
11 AT1G01140.3 470161|PACid_16036015 0.00582238 0.1354990 0.0429699     YN
12 AT1G01150.1 918855|PACid_16037307 0.13565500 0.1962460 0.6912480     YN
13 AT1G01160.1 918854|PACid_16044153 0.11558300 0.1929560 0.5990120     YN
14 AT1G01170.2 311317|PACid_16052302 0.00557175 0.2903370 0.0191906     YN
15 AT1G01180.1 909860|PACid_16056125 0.04065370 0.1557400 0.2610360     YN
16 AT1G01190.1 311315|PACid_16059488 0.02849220 0.1538610 0.1851810     YN
17 AT1G01200.1 470156|PACid_16041002 0.01983450 0.1512510 0.1311360     YN
18 AT1G01210.1 311313|PACid_16057125 0.02106910 0.1433630 0.1469630     YN
19 AT1G01220.1 470155|PACid_16047984 0.01530070 0.1446480 0.1057780     YN

```

Instead if using a multiple alignment tool for pairwise alignments you can also
choose a global pairwise alignment of orthologous genes based on the Needleman-Wunsch algorithm.
For this purpose the argument `aa_aln_type` must be set to `aa_aln_type = "pairwise"` and `aa_aln_tool = "NW"` 
for Needleman-Wunsch.

```{r,eval=FALSE}

library(orthologr)

# get a dNdS table using:
# 1) reciprocal best hit for orthology inference (RBH)
# 2) pairwise amino acid alignments using Needleman-Wunsch
# 3) pal2nal for codon alignments
# 4) Comeron (1995) for dNdS estimation
# 5) single core processing 'comp_cores = 1'
dNdS(query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
     subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
     ortho_detection = "RBH", aa_aln_type = "pairwise",
     aa_aln_tool = "NW", codon_aln_tool = "pal2nal", 
     dnds_est.method = "Comeron", comp_cores = 1, clean_folders = TRUE, quiet = TRUE)


```


```

       query_id            subject_id       dN     dS    dNdS
 1: AT1G01010.1 333554|PACid:16033839 0.106400 0.2537 0.41950
 2: AT1G01020.1 470181|PACid:16064328 0.040230 0.1037 0.38790
 3: AT1G01030.1 470180|PACid:16054974 0.014990 0.1265 0.11850
 4: AT1G01040.1 333551|PACid:16057793 0.013470 0.1165 0.11560
 5: AT1G01050.1 909874|PACid:16064489 0.000000 0.1750 0.00000
 6: AT1G01060.3 470177|PACid:16043374 0.044950 0.1133 0.39670
 7: AT1G01070.1 918864|PACid:16052578 0.018300 0.1059 0.17280
 8: AT1G01080.1 909871|PACid:16053217 0.033980 0.1056 0.32170
 9: AT1G01090.1 470171|PACid:16052860 0.009104 0.2181 0.04174
10: AT1G01110.2 333544|PACid:16034284 0.032480 0.1220 0.26620
11: AT1G01120.1 918858|PACid:16049140 0.003072 0.1326 0.02317
12: AT1G01140.3 470161|PACid:16036015 0.005672 0.1312 0.04324
13: AT1G01150.1 918855|PACid:16037307 0.130000 0.2028 0.64120
14: AT1G01160.1 918854|PACid:16044153 0.104600 0.2804 0.37310
15: AT1G01170.2 311317|PACid:16052302 0.000000 0.3064 0.00000
16: AT1G01180.1 909860|PACid:16056125 0.029680 0.1763 0.16830
17: AT1G01190.1 311315|PACid:16059488 0.028690 0.1618 0.17730
18: AT1G01200.1 470156|PACid:16041002 0.019050 0.1675 0.11370
19: AT1G01210.1 311313|PACid:16057125 0.020670 0.1540 0.13420
20: AT1G01220.1 470155|PACid:16047984 0.015690 0.1533 0.10230

```


The `dNdS()` function can be used choosing the following options:

* `ortho_detection` : `RBH` (BLAST best reciprocal hit), `BH` (BLAST best reciprocal hit), `PO` (ProteinOrtho), `OrthoMCL` (OrthoMCL), `IP` (InParanoid)
* `aa_aln_type` : `multiple` or `pairwise`
* `aa_aln_tool` : `clustalw`, `t_coffee`, `muscle`, `clustalo`, `mafft`, and `NW` (in case `aa_aln_type = "pairwise"`)
* `codon_aln_tool` : `pal2nal`
* `dnds_est.method` : `Li`, `Comeron`, `NG`, `LWL`, `LPB`, `MLWL`, `YN`, and `MYN`

Please see `?dNdS` for details.

In case your BLAST program, or multiple alignment program can not be executed from the default execution PATH
you can specify the `aa_aln_path` or `blast_path` arguments.

```{r,eval=FALSE}

library(orthologr)

# using the `aa_aln_path` or `blast_path` arguments
dNdS(query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
     subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
     ortho_detection = "RBH",blast_path = "here/path/to/blastp",
     aa_aln_type = "multiple", aa_aln_tool = "clustalw", 
     aa_aln_path = "here/path/to/clustalw",
     codon_aln_tool = "pal2nal", dnds_est.method = "Comeron", 
     comp_cores = 1, clean_folders = TRUE, quiet = TRUE)


```


## Advanced options

Additional arguments that can be passed to `dNdS()` allow to use more advanced 
options of several interface programs.

To pass additional parameters to the interface programs, you can use the
`blast_params` and `aa_aln_params`. The `aa_aln_params` argument assumes
that when you chose e.g. `aa_aln_tool = "mafft"` you will pass the corresponding
additional parameters in MAFFT notation.

```{r,eval=FALSE}

library(orthologr)

# get dNdS estimated for orthologous genes between A. thaliana and A. lyrata
# using additional parameters:

# get a dNdS table using:
# 1) reciprocal best hit for orthology inference (RBH)
# 2) multiple amino acid alignments using MAFFT
# 3) pal2nal for codon alignments
# 4) Comeron (1995) for dNdS estimation
# 5) single core processing 'comp_cores = 1'
Ath_Aly_dnds <- dNdS(
                  query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
                  subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
                  ortho_detection = "RBH",blast_params = "-matrix BLOSUM80",
                  aa_aln_tool = "mafft",aa_aln_params = "--maxiterate 1 --clustalout",
                  dnds_est.method = "Comeron", comp_cores = 1, clean_folders = TRUE, quiet = TRUE
                  )

        
# filter for:
# 1) all dN values having an NA value or omitted
# 2) all dS values having an NA value or omitted
# 3) all dNdS values >= 0.1 are omitted

filter_dNdS(Ath_Aly_dnds, dnds.threshold = 0.1)


```


```

     query_id      subject_id       dN     dS    dNdS
1 AT1G01050.1 909874|PACid:16 0.000000 0.1750 0.00000
2 AT1G01090.1 470171|PACid:16 0.009843 0.2150 0.04579
3 AT1G01120.1 918858|PACid:16 0.003072 0.1326 0.02317
4 AT1G01140.3 470161|PACid:16 0.005672 0.1312 0.04324
5 AT1G01170.2 311317|PACid:16 0.008750 0.2827 0.03095
6 AT1G01220.1 470155|PACid:16 0.015210 0.1533 0.09919

```
Here `blast_params` and `aa_aln_params` take an character string specifying the parameters
that shall passed to BLAST and MAFFT. The notation of these parameters must follow the 
command line call of the stand alone versions of BLAST and MAFFT: e.g. `blast_params = "blast_params = -matrix BLOSUM80"`
and `aa_aln_params = "--maxiterate 1 --clustalout"`.

