## ---- echo = FALSE, message = FALSE-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
options(width = 750)
knitr::opts_chunk$set(
  comment = "#>",
  error = FALSE,
  tidy = FALSE)

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  system("blastp -version")
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  library(dplyr)
#  
#  # performing a BLAST search using blastp (default)
#  hit_tbl <- blast(query_file   = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#                   subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'))
#  
#  
#  glimpse(hit_tbl)
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # performing a BLAST search using blastp (default)
#  hit_tbl <- blast(query_file   = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#                   subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
#                   path         = "/path/to/blastp")
#  
#  
#  hit_tbl
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # access columns: query_id, subject_id, evalue, and bit_score
#  hit_tbl[ , list(query_id, subject_id, evalue, bit_score)]
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # BLAST computations using the comp_cores parameter: here with 2 cores
#  blast(query_file   = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#        subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
#        comp_cores   = 2)
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # install.packages(c("plyr","dplyr"), dependencies = TRUE)
#  library(plyr)
#  library(dplyr)
#  
#  # perform a blastp search of 20 A. thaliana genes against 1000 A. lyrata genes
#  hit_tbl <- blast(query_file   = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#                   subject_file = system.file('seqs/ortho_lyra_cds_1000.fasta', package = 'orthologr'))
#  
#  # determine 'one-to-many' and 'one-to-one' gene relationships
#  rel_hit_tbl <- ddply(.data = hit_tbl, .variables = "query_id", .fun = nrow)
#  colnames(rel_hit_tbl)[2] <- "n_genes"
#  
#  rel_hit_tbl
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # classify into 'one-to-one' relationships
#  one_to_one <- filter(rel_hit_tbl,n_genes == 1)
#  
#  # classify into 'one-to-many' relationships
#  one_to_many <- filter(rel_hit_tbl,n_genes > 1)
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # look at one_to_one
#  one_to_one
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # look at one_to_many
#  one_to_many
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # look at the evalue, perc_identity, and alig_length of one_to_one genes
#  oo_genes <- dplyr::filter(hit_tbl,query_id %in% one_to_one[ , "query_id"])
#  
#  oo_genes[ , list(query_id,subject_id,evalue,perc_identity,alig_length)]
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # look at the evalue, perc_identity, and alig_length of one_to_one genes
#  oo_genes <- dplyr::filter(hit_tbl,query_id %in% one_to_one[ , "query_id"])
#  
#  true_orthologs <- dplyr::filter(oo_genes,evalue < 1e-5, perc_identity > 80, alig_length > 300)
#  
#  true_orthologs[ , list(query_id,subject_id,evalue,perc_identity,alig_length)]
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  one_to_many
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  hit_tbl["AT1G01200.1", list(query_id,subject_id,evalue,perc_identity,alig_length)]
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  rbh_hit_tbl <- blast_rec(query_file   = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#                           subject_file = system.file('seqs/ortho_lyra_cds_1000.fasta', package = 'orthologr'))
#  
#  
#  rbh_hit_tbl
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  rbh_hit_tbl["AT1G01200.1" , ]
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  hit_tbl["AT1G01070.1", list(query_id,subject_id,evalue,perc_identity,alig_length)]
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  rbh_hit_tbl["AT1G01070.1" , ]
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # read CDS sequences of the 20 example query genes of A. thaliana
#  Ath.cds <- read.cds(file   = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#                      format = "fasta")
#  
#  # read CDS sequences of the 1000 example subject genes of A. lyrata
#  Aly.cds <- read.cds(file   = system.file('seqs/ortho_lyra_cds_1000.fasta', package = 'orthologr'),
#                      format = "fasta")
#  
#  
#  # show the sequence of gene AT1G01070.1
#  Ath.cds["AT1G01070.1" , seqs]
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  library(Biostrings)
#  
#  # perform 3 global alignments between:  AT1G01070.1 and 918864|PACid:16052578,
#  # 919693|PACid:16048878, 919961|PACid:16062329
#  sapply(Aly.cds[ hit_tbl[ "AT1G01070.1", subject_id], seqs ], pairwiseAlignment,
#         pattern = Ath.cds["AT1G01070.1" , seqs],
#         type    = "global" )
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # show the length distribution of all genes
#  # stored in "Ath.cds"
#  Ath.cds[ , nchar(seqs)]
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  Ath.cds["AT1G01070.1" , nchar(seqs)]
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  Ath.cds <- read.cds(system.file('seqs/ortho_thal_cds_1000.fasta', package = 'orthologr'),
#                      format = "fasta")
#  
#  
#  hist(Ath.cds[ , nchar(seqs)], breaks = 100)
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # performing gene orthology inference using the best hit (BH) method
#  blast_best(query_file    = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#             subject_file  = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
#             clean_folders = TRUE)
#  
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # BLAST+ best hit search
#  best_hit_tbl <- blast_best(query_file      = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#                             subject_file    = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
#                             detailed_output = TRUE)
#  
#  dplyr::glimpse(best_hit_tbl)
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # performing gene orthology inference using the reciprocal best hit (RBH) method
#  blast_rec(query_file   = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#            subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'))
#  
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # running blast_rec() using detailed_output = TRUE
#  rbh <- blast_rec(query_file      = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#                   subject_file    = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
#                   detailed_output = TRUE)
#  
#  
#  dplyr::glimpse(rbh)

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  ab <- advanced_blast(query_file  = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#                       subject_file    = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
#                       blast_algorithm = "blastp",
#                       blast_params    = "-evalue 1E-5 -num_threads 1")
#  
#  
#  dplyr::glimpse(ab)
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # when performing an advanced BLAST search, you can easily select the best hit
#  library(dplyr)
#  
#  advB <- advanced_blast(
#             query_file      = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#             subject_file    = system.file('seqs/ortho_lyra_cds_1000.fasta', package = 'orthologr'),
#             seq_type        = "cds",
#             blast_algorithm = "blastp",
#             blast_params    = "-evalue 1E-5 -num_threads 1")
#  
#  best_hit <- advB %>% group_by(query_id) %>% do(min(evalue))
#  
#  
#  best_hit
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # DELTA-BLAST
#  #
#  # you can also use deltablast to perform BLAST searches
#  # make sure you have specified the the db_path argument to the cdd_deltablast.* files
#  advanced_blast(query_file      = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#                 subject_file    = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
#                 blast_algorithm = "deltablast",
#                 db_path         = "path/to/cdd_files",
#                 blast_params    = "-evalue 1E-5 -num_threads 2")
#  
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # perform a delta-blast serach between A. thaliana and A. lyrata genes
#  delta.blast(
#              query_file   = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#              subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
#              cdd.path     = "path/to/cdd/database/folder",
#              comp_cores   = 1)
#  
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # perform a delta-blast serach between A. thaliana and the cdd database
#  delta.blast(
#              query_file   = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#              subject_file = "cdd_delta",
#              cdd.path     = "path/to/cdd/database/folder",
#              comp_cores   = 1)
#  

## ---- eval = FALSE----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # install the 'biomartr' package from CRAN
#  install.packages("biomartr")
#  
#  # load biomartr
#  library("biomartr")
#  
#  # download the NCBI nr database to your local machine to a folder
#  # names 'DB'
#  sapply(listDatabases("nr"),download_database)

## ---- eval = FALSE----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  nr.res <- blast.nr( query_file      = system.file('seqs/aa_seqs.fasta', package = 'orthologr'),
#                      nr.path         = "DB",
#                      seq_type        = "protein",
#                      max.target.seqs = 20,
#                      comp_cores      = 2 )
#  
#  nr.res

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  nr.res[ , "subject_id"] <- unlist(lapply(stringr::str_split(nr.res[ , subject_id],"[|]"), function(x) x[2]))
#  
#  nr.res

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # using a SQLite database to store the BLAST output and
#  # select some example data
#  
#  library(dplyr)
#  
#  sqlE <- advanced_blast(
#               query_file      = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
#               subject_file    = system.file('seqs/ortho_lyra_cds_1000.fasta', package = 'orthologr'),
#               seq_type        = "cds",
#               blast_algorithm = "blastp",
#               blast_params    = "-evalue 1E-5 -num_threads 1",
#               sql_database    = TRUE)
#  
#  head(sqlE)
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # look at the structure of sqlE
#  glimpse(sqlE)
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  # select all rows that have an evalue of zero
#  filter(sqlE, V12 > 1e-15)
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # select the best hit using the evalue criterion
#  sqlE %>% group_by(V1) %>% do(best_hit_eval = min(V12))
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  head(set_blast(file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'))[[1]] , 2)
#  

## ----eval=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#  
#  # make the A. thaliana genome to a blast-able database
#  advanced_makedb( database_file = system.file('seqs/ortho_thal_aa.fasta', package = 'orthologr'),
#                   params        = "-input_type fasta -dbtype prot -hash_index" )
#  
#  

