<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="date" content="2016-03-04" />

<title>BLAST searches using orthologr</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>


<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div class="fluid-row" id="header">


<h1 class="title">BLAST searches using orthologr</h1>
<h4 class="date"><em>2016-03-04</em></h4>

</div>


<div id="getting-started" class="section level2">
<h2>Getting Started</h2>
<p>The <code>orthologr</code> package provides several interface functions to perform BLAST searches.</p>
<p>First, users need to make sure that they have <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">BLAST</a> installed on their machine. Please follow <a href="https://github.com/HajkD/orthologr/blob/master/vignettes/Install.Rmd#install-blast">these</a> instructions to install BLAST on your manine.</p>
</div>
<div id="performing-blast-searches" class="section level2">
<h2>Performing BLAST Searches</h2>
<p>The <code>orthologr</code> package stores 20 example genes (orthologs) between <em>Arabidopsis thaliana</em> and <em>Arabidopsis lyrata</em>. The following example BLAST search shall illustrate a simple search with standard parameters provided by the <code>blast()</code> function.</p>
<p>When running the subsequent functions please make sure you can call BLAST+ from your console either in the standard <code>PATH</code> or in case you have BLAST+ installed in a separate folder, please specify the <code>path</code> argument that can be passed to <code>blast()</code>.</p>
<p>To check whether BLAST+ can be executed from the default <code>PATH</code> (<code>usr/bin/local</code> on UNIX systems), you can run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">system</span>(<span class="st">&quot;blastp -version&quot;</span>)</code></pre></div>
<p>This should return something like this:</p>
<pre><code>blastp: 2.2.29+
Package: blast 2.2.29, build Dec 10 2013 15:51:59
</code></pre>
<p>If everything works properly, you can get started with you first BLAST+ search.</p>
</div>
<div id="the-blast-function" class="section level2">
<h2>The blast() function</h2>
<p>The <code>blast()</code> function provides the easiest way to perform a BLAST search.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(dplyr)

<span class="co"># performing a BLAST search using blastp (default)</span>
hit_tbl &lt;-<span class="st"> </span><span class="kw">blast</span>(<span class="dt">query_file   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))


<span class="kw">glimpse</span>(hit_tbl)</code></pre></div>
<pre><code>Variables:
$ query_id      (chr) &quot;AT1G01010.1&quot;, &quot;AT1G01020.1&quot;, &quot;AT1G01030.1&quot;,...
$ subject_id    (chr) &quot;333554|PACid:16033839&quot;, &quot;470181|PACid:16064...
$ perc_identity (dbl) 73.99, 91.06, 95.54, 91.98, 100.00, 89.51, 9...
$ alig_length   (dbl) 469, 246, 359, 1970, 213, 648, 366, 300, 434...
$ mismatches    (dbl) 80, 22, 12, 85, 0, 58, 14, 22, 8, 34, 4, 6, ...
$ gap_openings  (dbl) 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, 1, 3, 2, 1...
$ q_start       (dbl) 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2,...
$ q_end         (dbl) 430, 246, 359, 1910, 213, 646, 366, 294, 429...
$ s_start       (dbl) 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4...
$ s_end         (dbl) 466, 246, 355, 1963, 213, 640, 362, 299, 433...
$ evalue        (dbl) 0e+00, 7e-166, 0e+00, 0e+00, 2e-160, 0e+00, ...
$ bit_score     (dbl) 627, 454, 698, 3704, 437, 1037, 696, 491, 85...
</code></pre>
<p>As you can see, the hit table shows the output of the BLAST+ search. The <code>blast()</code> function runs <code>blastp</code> as default BLAST+ algorithm. Different BLAST+ algorithms can be selected by specifying the <code>blast_algorithm</code> argument, e.g. <code>blast_algorithm = &quot;tblastn&quot;</code>. See <code>?blast</code> for further details. The <code>blast()</code> function returns the BLAST arguments: <code>query_id</code>, <code>subject_id</code>, <code>perc_identity</code>, <code>alig_length</code>, <code>mismatches</code>, <code>gap_openings</code>, <code>q_start</code>, <code>q_end</code>, <code>s_start</code>, <code>s_end</code>, <code>evalue</code>, and <code>bit_score</code>.</p>
<p>Since <code>blast()</code> stores the hit table returned by BLAST in a data.table object, you can access each column, using the <a href="https://github.com/Rdatatable/data.table/wiki">data.table notation</a>.</p>
<p>In case you need to specify the <code>PATH</code> to BLAST+ please use the <code>path</code> argument:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># performing a BLAST search using blastp (default)</span>
hit_tbl &lt;-<span class="st"> </span><span class="kw">blast</span>(<span class="dt">query_file   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">path         =</span> <span class="st">&quot;/path/to/blastp&quot;</span>)


hit_tbl</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># access columns: query_id, subject_id, evalue, and bit_score</span>
hit_tbl[ , <span class="kw">list</span>(query_id, subject_id, evalue, bit_score)]</code></pre></div>
<pre><code>
       query_id            subject_id evalue bit_score
 1: AT1G01010.1 333554|PACid:16033839  0e+00       627
 2: AT1G01020.1 470181|PACid:16064328 7e-166       454
 3: AT1G01030.1 470180|PACid:16054974  0e+00       698
 4: AT1G01040.1 333551|PACid:16057793  0e+00      3704
 5: AT1G01050.1 909874|PACid:16064489 2e-160       437
 6: AT1G01060.3 470177|PACid:16043374  0e+00      1037
 7: AT1G01070.1 918864|PACid:16052578  0e+00       696
 8: AT1G01080.1 909871|PACid:16053217 1e-178       491
 9: AT1G01090.1 470171|PACid:16052860  0e+00       859
10: AT1G01110.2 333544|PACid:16034284  0e+00       972
11: AT1G01120.1 918858|PACid:16049140  0e+00      1092
12: AT1G01140.3 470161|PACid:16036015  0e+00       918
13: AT1G01150.1 918855|PACid:16037307 3e-150       421
14: AT1G01160.1 918854|PACid:16044153  1e-93       268
15: AT1G01170.2 311317|PACid:16052302  3e-54       158
16: AT1G01180.1 909860|PACid:16056125  0e+00       576
17: AT1G01190.1 311315|PACid:16059488  0e+00      1036
18: AT1G01200.1 470156|PACid:16041002 3e-172       470
19: AT1G01210.1 311313|PACid:16057125  7e-76       215
20: AT1G01220.1 470155|PACid:16047984  0e+00      2106
</code></pre>
<p>The <code>blast()</code> function also allows you to pass additional parameters to the BLAST+ search using the <code>blast_params</code> argument. In the following example, a remote BLAST+ search is performed.</p>
<pre><code>
blast(query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
                 subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
                 blast_params = &quot;-max_target_seqs 1&quot;)</code></pre>
<p>In all cases the default <code>e-value</code> BLAST+ searches is <code>1E-5</code> and the default <code>blast_algorithm</code> is <code>blastp</code>.</p>
<p>Since BLAST+ searches can be computationally expensive, it is possible to specify the <code>comp_cores</code> argument when working with an multicore machine.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># BLAST computations using the comp_cores parameter: here with 2 cores</span>
<span class="kw">blast</span>(<span class="dt">query_file   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
      <span class="dt">subject_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
      <span class="dt">comp_cores   =</span> <span class="dv">2</span>)</code></pre></div>
<p>The <code>query_file</code> and <code>subject_file</code> arguments specify the path to the corresponding fasta files storing the <code>CDS</code> files, <code>amino acid</code> files, or <code>genome</code> files of the query organism and subject organism of interest. Make sure that when using <code>CDS</code>files, <code>amino acid</code> files, or <code>genome</code> files the corresponding argument <code>seq_type</code> must be adapted according to the input data format.</p>
<p>Use :</p>
<ul>
<li><code>CDS</code> files -&gt; <code>seq_type = &quot;cds&quot;</code></li>
<li><code>amino acid</code> files -&gt; <code>seq_type = &quot;protein&quot;</code></li>
<li><code>genome</code> files -&gt; <code>seq_type = &quot;dna&quot;</code></li>
</ul>
<p>The <code>format</code> argument specifies the input file format, e.g. “fasta” or “gbk”. The <code>blast_algorithm</code> argument specifies the BLAST program (algorithm) that shall be used to perform BLAST searches, e.g. “blastp”,“blastn”,“tblastn”,etc. Again, the <code>eval</code> argument defines the default e-value that shall be chosen as best hit threshold.</p>
<div id="using-the-split-apply-combine-strategy-for-a-blast-hit-table" class="section level3">
<h3>Using the split-apply-combine strategy for a BLAST hit table</h3>
<p>All <code>blast</code> functions implemented in <code>orthologr</code> can easily be processed using the <a href="http://www.jstatsoft.org/v40/i01/paper">split-apply-combine strategy</a> to detect for example <code>one-to-one</code>, <code>one-to-many</code>, and <code>many-to-many</code> gene homology relationships.</p>
<p>Here a simple example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install.packages(c(&quot;plyr&quot;,&quot;dplyr&quot;), dependencies = TRUE)</span>
<span class="kw">library</span>(plyr)
<span class="kw">library</span>(dplyr)

<span class="co"># perform a blastp search of 20 A. thaliana genes against 1000 A. lyrata genes</span>
hit_tbl &lt;-<span class="st"> </span><span class="kw">blast</span>(<span class="dt">query_file   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))

<span class="co"># determine 'one-to-many' and 'one-to-one' gene relationships</span>
rel_hit_tbl &lt;-<span class="st"> </span><span class="kw">ddply</span>(<span class="dt">.data =</span> hit_tbl, <span class="dt">.variables =</span> <span class="st">&quot;query_id&quot;</span>, <span class="dt">.fun =</span> nrow)
<span class="kw">colnames</span>(rel_hit_tbl)[<span class="dv">2</span>] &lt;-<span class="st"> &quot;n_genes&quot;</span>

rel_hit_tbl</code></pre></div>
<pre><code>      query_id n_genes
1  AT1G01010.1       4
2  AT1G01020.1       1
3  AT1G01030.1       1
4  AT1G01040.1       1
5  AT1G01050.1       1
6  AT1G01060.3       2
7  AT1G01070.1       3
8  AT1G01080.1       4
9  AT1G01090.1       1
10 AT1G01110.2       1
11 AT1G01120.1       3
12 AT1G01140.3      36
13 AT1G01150.1       1
14 AT1G01160.1       1
15 AT1G01170.2       1
16 AT1G01180.1       1
17 AT1G01190.1       6
18 AT1G01200.1       8
19 AT1G01210.1       1
20 AT1G01220.1       1
</code></pre>
<p>Now you can sort genes into classes: <code>one-to-one</code> and <code>one-to-many</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># classify into 'one-to-one' relationships</span>
one_to_one &lt;-<span class="st"> </span><span class="kw">filter</span>(rel_hit_tbl,n_genes ==<span class="st"> </span><span class="dv">1</span>)

<span class="co"># classify into 'one-to-many' relationships</span>
one_to_many &lt;-<span class="st"> </span><span class="kw">filter</span>(rel_hit_tbl,n_genes &gt;<span class="st"> </span><span class="dv">1</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at one_to_one</span>
one_to_one </code></pre></div>
<pre><code>      query_id n_genes
1  AT1G01020.1       1
2  AT1G01030.1       1
3  AT1G01040.1       1
4  AT1G01050.1       1
5  AT1G01090.1       1
6  AT1G01110.2       1
7  AT1G01150.1       1
8  AT1G01160.1       1
9  AT1G01170.2       1
10 AT1G01180.1       1
11 AT1G01210.1       1
12 AT1G01220.1       1
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at one_to_many</span>
one_to_many</code></pre></div>
<pre><code>     query_id n_genes
1 AT1G01010.1       4
2 AT1G01060.3       2
3 AT1G01070.1       3
4 AT1G01080.1       4
5 AT1G01120.1       3
6 AT1G01140.3      36
7 AT1G01190.1       6
8 AT1G01200.1       8
</code></pre>
<p>Now we can treat classes: <code>one_to_one</code> and <code>one_to_many</code> differently:</p>
</div>
<div id="one-to-one-genes" class="section level3">
<h3>one-to-one genes</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at the evalue, perc_identity, and alig_length of one_to_one genes</span>
oo_genes &lt;-<span class="st"> </span>dplyr::<span class="kw">filter</span>(hit_tbl,query_id %in%<span class="st"> </span>one_to_one[ , <span class="st">&quot;query_id&quot;</span>])

oo_genes[ , <span class="kw">list</span>(query_id,subject_id,evalue,perc_identity,alig_length)]</code></pre></div>
<pre><code>      query_id            subject_id evalue perc_identity alig_length
1  AT1G01020.1 470181|PACid:16064328 3e-164         91.06         246
2  AT1G01030.1 470180|PACid:16054974  0e+00         95.54         359
3  AT1G01040.1 333551|PACid:16057793  0e+00         91.98        1970
4  AT1G01050.1 909874|PACid:16064489 1e-158        100.00         213
5  AT1G01090.1 470171|PACid:16052860  0e+00         96.77         434
6  AT1G01110.2 333544|PACid:16034284  0e+00         93.56         528
7  AT1G01150.1 918855|PACid:16037307 1e-148         72.63         285
8  AT1G01160.1 918854|PACid:16044153  5e-92         84.92         179
9  AT1G01170.2 311317|PACid:16052302  1e-52         85.57          97
10 AT1G01180.1 909860|PACid:16056125  0e+00         92.58         310
11 AT1G01210.1 311313|PACid:16057125  3e-74         95.33         107
12 AT1G01220.1 470155|PACid:16047984  0e+00         96.69        1056
</code></pre>
<p>Now you could filter for additional criteria to define a first set of true orthologs. In this example we define true orthologs as <code>one_to_one</code> genes having a minimum alignment length of 300, a perc_identity of &gt; 80 percent and an e-value &lt; 1E-5.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at the evalue, perc_identity, and alig_length of one_to_one genes</span>
oo_genes &lt;-<span class="st"> </span>dplyr::<span class="kw">filter</span>(hit_tbl,query_id %in%<span class="st"> </span>one_to_one[ , <span class="st">&quot;query_id&quot;</span>])

true_orthologs &lt;-<span class="st"> </span>dplyr::<span class="kw">filter</span>(oo_genes,evalue &lt;<span class="st"> </span><span class="fl">1e-5</span>, perc_identity &gt;<span class="st"> </span><span class="dv">80</span>, alig_length &gt;<span class="st"> </span><span class="dv">300</span>)

true_orthologs[ , <span class="kw">list</span>(query_id,subject_id,evalue,perc_identity,alig_length)]</code></pre></div>
<pre><code>      query_id            subject_id evalue perc_identity alig_length
1: AT1G01030.1 470180|PACid:16054974      0         95.54         359
2: AT1G01040.1 333551|PACid:16057793      0         91.98        1970
3: AT1G01090.1 470171|PACid:16052860      0         96.77         434
4: AT1G01110.2 333544|PACid:16034284      0         93.56         528
5: AT1G01180.1 909860|PACid:16056125      0         92.58         310
6: AT1G01220.1 470155|PACid:16047984      0         96.69        1056
</code></pre>
<p>This way we could filter out a high confidence set of orthologous genes from the <code>one_to_one</code> class of genes.</p>
<p>In reality most orthology inference programs and methods perform way more complicated and sophisticated analyses to distinguish true orthologs from true paralogs (in-paralogs, out-paralogs, etc.). These subsequent analyses can also be performed using the above introduced split-apply-combine strategy.</p>
<p>Note, that you can perform self-BLAST searches <code>blast(query,query)</code> and <code>blast(subject,subject)</code> to distinguish between orthologous and paralogous genes.</p>
<p>Now we continue with the <code>one_to_many</code> class of genes.</p>
</div>
<div id="one-to-many-genes" class="section level3">
<h3>one-to-many genes</h3>
<p>Here we want to address the question how to deal with multiple hits returned by BLAST+ .</p>
<p>Again we investigate all <code>one_to_many</code> genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">one_to_many</code></pre></div>
<pre><code>     query_id n_genes
1 AT1G01010.1       4
2 AT1G01060.3       2
3 AT1G01070.1       3
4 AT1G01080.1       4
5 AT1G01120.1       3
6 AT1G01140.3      36
7 AT1G01190.1       6
8 AT1G01200.1       8
</code></pre>
<p>When looking at gene_id <code>AT1G01200.1</code> we see that it was found 8 times in the corresponding subject set of <em>A. lyrata</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hit_tbl[<span class="st">&quot;AT1G01200.1&quot;</span>, <span class="kw">list</span>(query_id,subject_id,evalue,perc_identity,alig_length)]</code></pre></div>
<pre><code>      query_id            subject_id evalue perc_identity alig_length
1: AT1G01200.1 470156|PACid:16041002 2e-170         95.80         238
2: AT1G01200.1 909905|PACid:16035105  7e-06         21.64         171
3: AT1G01200.1 910431|PACid:16035207  8e-74         52.97         219
4: AT1G01200.1 918732|PACid:16054958  2e-50         44.56         193
5: AT1G01200.1 919287|PACid:16060536  1e-68         58.10         179
6: AT1G01200.1 919355|PACid:16050170  7e-72         53.30         212
7: AT1G01200.1 919721|PACid:16036935  9e-80         59.31         204
8: AT1G01200.1 919852|PACid:16055066  4e-07         24.03         154
</code></pre>
<p>Now we have to decide which hit shall be considered as potential <em>ortholog</em>. In this example <code>subject_id</code> <code>470156|PACid:16041002</code> has the highest <code>perc_identity</code> as well as the lowest e-value <code>2e-170</code>. So a straightforward approach would be to choose subject gene <code>470156|PACid:16041002</code> as potential ortholog of query gene <code>AT1G01200.1</code>.</p>
<p>We can validate this approach by running a reciprocal best hit search with <code>blast_rec()</code>and compare the output of gene <code>AT1G01200.1</code> with our choice <code>470156|PACid:16041002</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbh_hit_tbl &lt;-<span class="st"> </span><span class="kw">blast_rec</span>(<span class="dt">query_file   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                         <span class="dt">subject_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))


rbh_hit_tbl</code></pre></div>
<pre><code>      query_id            subject_id evalue
1  AT1G01010.1 333554|PACid:16033839  0e+00
2  AT1G01020.1 470181|PACid:16064328 3e-164
3  AT1G01030.1 470180|PACid:16054974  0e+00
4  AT1G01040.1 333551|PACid:16057793  0e+00
5  AT1G01050.1 909874|PACid:16064489 1e-158
6  AT1G01060.3 470177|PACid:16043374  0e+00
7  AT1G01070.1 918864|PACid:16052578  0e+00
8  AT1G01080.1 909871|PACid:16053217 5e-177
9  AT1G01090.1 470171|PACid:16052860  0e+00
10 AT1G01110.2 333544|PACid:16034284  0e+00
11 AT1G01120.1 918858|PACid:16049140  0e+00
12 AT1G01140.3 470161|PACid:16036015  0e+00
13 AT1G01150.1 918855|PACid:16037307 1e-148
14 AT1G01160.1 918854|PACid:16044153  5e-92
15 AT1G01170.2 311317|PACid:16052302  1e-52
16 AT1G01180.1 909860|PACid:16056125  0e+00
17 AT1G01190.1 311315|PACid:16059488  0e+00
18 AT1G01200.1 470156|PACid:16041002 2e-170
19 AT1G01210.1 311313|PACid:16057125  3e-74
20 AT1G01220.1 470155|PACid:16047984  0e+00
</code></pre>
<p>When we now look at gene <code>AT1G01200.1</code> we find that indeed subject gene <code>470156|PACid:16041002</code> has been detected as potential ortholog having the same evalue <code>2e-170</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbh_hit_tbl[<span class="st">&quot;AT1G01200.1&quot;</span> , ]</code></pre></div>
<pre><code>     query_id            subject_id evalue
1 AT1G01200.1 470156|PACid:16041002 2e-170
</code></pre>
<p>Nevertheless there might be cases in which it is hard to decide for or against the best hit compared with all other hits.</p>
<p>For example we can investigate gene <code>AT1G01070.1</code> :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hit_tbl[<span class="st">&quot;AT1G01070.1&quot;</span>, <span class="kw">list</span>(query_id,subject_id,evalue,perc_identity,alig_length)]</code></pre></div>
<pre><code>      query_id            subject_id evalue perc_identity alig_length
1: AT1G01070.1 918864|PACid:16052578  0e+00         95.08         366
2: AT1G01070.1 919693|PACid:16048878  2e-67         32.87         356
3: AT1G01070.1 919961|PACid:16062329  0e+00         79.29         338
</code></pre>
<p>Here both e-values for subject genes <code>918864|PACid:16052578</code> and <code>919961|PACid:16062329</code> are the same and only <code>perc_identity</code> and <code>alig_length</code> differ. The <em>reciprocal best hit</em> approach chose gene <code>918864|PACid:16052578</code> which also had the highest <code>perc_identity</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbh_hit_tbl[<span class="st">&quot;AT1G01070.1&quot;</span> , ]</code></pre></div>
<pre><code>     query_id            subject_id evalue
1 AT1G01070.1 918864|PACid:16052578      0
</code></pre>
<p>But since the <code>blast_rec()</code> function was implemented to choose the bidirectional best hit based on the e-value, in border line cases a different gene as expected could be chosen.</p>
<p>An alternative analysis that can be performed with these three candidate subject genes is the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read CDS sequences of the 20 example query genes of A. thaliana</span>
Ath.cds &lt;-<span class="st"> </span><span class="kw">read.cds</span>(<span class="dt">file   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                    <span class="dt">format =</span> <span class="st">&quot;fasta&quot;</span>)

<span class="co"># read CDS sequences of the 1000 example subject genes of A. lyrata</span>
Aly.cds &lt;-<span class="st"> </span><span class="kw">read.cds</span>(<span class="dt">file   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                    <span class="dt">format =</span> <span class="st">&quot;fasta&quot;</span>)


<span class="co"># show the sequence of gene AT1G01070.1</span>
Ath.cds[<span class="st">&quot;AT1G01070.1&quot;</span> , seqs]</code></pre></div>
<pre><code>[1] &quot;atggctggagatatgcaaggagtgagagtagtagaaaaatattcaccggtcatagtgatggtgatgtcaaatgta
gcgatgggttcggtgaatgcacttgtgaagaaagctcttgatgttggtgtgaaccatatggtcattggtgcttatcgaat
ggctatttccgctttaattttggttccctttgcctatgtcttggaaaggaaaacaagaccacaaataacgtttaggctaa
tggtcgatcatttcgtcagtggccttctcggggcgagtttgatgcagtttttctttttgcttggtctgtcgtacacgtca
gcaactgtttcgtgtgctttggtaagcatgttgcctgcaatcaccttcgctttggcccttattttcaggactgaaaatgt
gaagattctaaagaccaaagcaggaatgttgaaggtgattggaactttgatctgtataagtggagctttgttcttaacat
tttacaaaggcccacaaatatcaaactctcactctcactctcacggtggggcttcccacaacaacaacgatcaagacaag
gccaataattggcttcttggatgtctttatttaaccataggaacagtgttgctatctctatggatgttgtttcaagggac
tttaagtattaagtacccttgcaaatactcgagcacttgtcttatgtcaattttcgcggcatttcaatgtgctctcttga
gcctttacaagagcagagacgttaatgattggatcatagatgatagattcgttatcaccgtcatcatatacgctggagtg
gtaggacaagcaatgacgacggttgcaacaacatgggggattaaaaaattaggagctgtgttcgcatcggcgtttttccc
acttactctcatttcggctactctatttgatttcctcattttacacactcctttataccttggaagtgtgattggatcac
tagtgaccataacgggtctctacatgttcttgtgggggaagaacaaagaaacggaatcatcaactgcattgtcttcagga
atggataacgaagctcaatatactactcctaataaggataacgactctaagtcgcccgtttaa&quot;
</code></pre>
<p>Now you can perform a global alignment between the CDS sequences of <code>AT1G01070.1</code> and the three subject genes as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Biostrings)

<span class="co"># perform 3 global alignments between:  AT1G01070.1 and 918864|PACid:16052578, </span>
<span class="co"># 919693|PACid:16048878, 919961|PACid:16062329</span>
<span class="kw">sapply</span>(Aly.cds[ hit_tbl[ <span class="st">&quot;AT1G01070.1&quot;</span>, subject_id], seqs ], pairwiseAlignment, 
       <span class="dt">pattern =</span> Ath.cds[<span class="st">&quot;AT1G01070.1&quot;</span> , seqs],
       <span class="dt">type    =</span> <span class="st">&quot;global&quot;</span> )</code></pre></div>
<pre><code>$...
Global PairwiseAlignmentsSingleSubject (1 of 1)
pattern: [1] atggctggagatatgcaaggagtgagagta...aaggataacgactctaagtcgcccgtttaa 
subject: [1] atgggtgaaggtatgattggagtgagagta...aaggataacgactctaagtcgcccgtttaa 
score: 1768.965 

$...
Global PairwiseAlignmentsSingleSubject (1 of 1)
pattern: [1] atggctgga---gatatgcaaggagtgaga...----cgac----tctaagtcgcccgtttaa 
subject: [1] atggctaaatcagatatgc------tg---...ggttccacaaggtctatatcgcc---ttaa 
score: -2318.726 

$...
Global PairwiseAlignmentsSingleSubject (1 of 1)
pattern: [1] atggctggagatatgcaaggagtgagagta...aaggataacgactctaagtcgcccgtttaa 
subject: [1] atgagtgaggatatgggaggagtgaaagta...----------------------------aa 
score: 486.462 
</code></pre>
<p><strong>Note</strong>: To obtain the score value, you need to specify the <code>scoreOnly = TRUE</code> in the <code>pairwiseAlignment</code> function.</p>
<p>As you can see, subject gene <code>918864|PACid:16052578</code> also has the highest global alignment score <code>1768.965</code> based on the Needleman-Wunsch algorithm. This strategy might help you to differentiate between border line cases.</p>
<p>The examples shown above shall demonstrate the use cases that can be performed using the <code>blast</code> functions implemented in <code>orthologr</code>.</p>
<p>Another useful analysis can be to take the length of the initial query genes into account using the <code>nchar()</code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show the length distribution of all genes</span>
<span class="co"># stored in &quot;Ath.cds&quot;</span>
Ath.cds[ , <span class="kw">nchar</span>(seqs)]</code></pre></div>
<pre><code>
[1] 1290  738 1077 5730  639 1938 1098  882 1287 1584 1587 1356 1038  588  252 1437 1608  714  321 3168
</code></pre>
<p>Or the length of a specific gene:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Ath.cds[<span class="st">&quot;AT1G01070.1&quot;</span> , <span class="kw">nchar</span>(seqs)]</code></pre></div>
<pre><code>[1] 1098</code></pre>
<p>This way you can easily visualize the length distribution of genes stored in your query organism file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Ath.cds &lt;-<span class="st"> </span><span class="kw">read.cds</span>(<span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                    <span class="dt">format =</span> <span class="st">&quot;fasta&quot;</span>)
                    
                    
<span class="kw">hist</span>(Ath.cds[ , <span class="kw">nchar</span>(seqs)], <span class="dt">breaks =</span> <span class="dv">100</span>)</code></pre></div>
</div>
</div>
<div id="the-blast_best-function" class="section level2">
<h2>The blast_best() function</h2>
<p>For some analyses it is sufficient to perform BLAST+ best hit searches. The <code>blast_best()</code> function is optimized to perform BLAST+ best hit searches (only based on the minimum e-value) and returns the best hit when performing a BLAST+ search of a query organisms (or set of query genes) against a subject organism (or set of subject genes).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># performing gene orthology inference using the best hit (BH) method</span>
<span class="kw">blast_best</span>(<span class="dt">query_file    =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
           <span class="dt">subject_file  =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
           <span class="dt">clean_folders =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>       query_id            subject_id evalue
 1: AT1G01010.1 333554|PACid:16033839  0e+00
 2: AT1G01020.1 470181|PACid:16064328 7e-166
 3: AT1G01030.1 470180|PACid:16054974  0e+00
 4: AT1G01040.1 333551|PACid:16057793  0e+00
 5: AT1G01050.1 909874|PACid:16064489 2e-160
 6: AT1G01060.3 470177|PACid:16043374  0e+00
 7: AT1G01070.1 918864|PACid:16052578  0e+00
 8: AT1G01080.1 909871|PACid:16053217 1e-178
 9: AT1G01090.1 470171|PACid:16052860  0e+00
10: AT1G01110.2 333544|PACid:16034284  0e+00
11: AT1G01120.1 918858|PACid:16049140  0e+00
12: AT1G01140.3 470161|PACid:16036015  0e+00
13: AT1G01150.1 918855|PACid:16037307 3e-150
14: AT1G01160.1 918854|PACid:16044153  1e-93
15: AT1G01170.2 311317|PACid:16052302  3e-54
16: AT1G01180.1 909860|PACid:16056125  0e+00
17: AT1G01190.1 311315|PACid:16059488  0e+00
18: AT1G01200.1 470156|PACid:16041002 3e-172
19: AT1G01210.1 311313|PACid:16057125  7e-76
20: AT1G01220.1 470155|PACid:16047984  0e+00
</code></pre>
<p>The <code>blast_best()</code> function returns: <code>query_id</code>, <code>subject_id</code>, and <code>eval</code>.</p>
<p>In case you need more parameters returned by a BLAST+ best hit search, you can specify the <code>detailed_output</code> argument (<code>detailed_output = TRUE</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># BLAST+ best hit search</span>
best_hit_tbl &lt;-<span class="st"> </span><span class="kw">blast_best</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                           <span class="dt">subject_file    =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                           <span class="dt">detailed_output =</span> <span class="ot">TRUE</span>)

dplyr::<span class="kw">glimpse</span>(best_hit_tbl)</code></pre></div>
<pre><code>Variables:
$ query_id      (chr) &quot;AT1G01010.1&quot;, &quot;AT1G01020.1&quot;, &quot;AT1G01030.1&quot;,...
$ subject_id    (chr) &quot;333554|PACid:16033839&quot;, &quot;470181|PACid:16064...
$ perc_identity (dbl) 73.99, 91.06, 95.54, 91.98, 100.00, 89.51, 9...
$ alig_length   (dbl) 469, 246, 359, 1970, 213, 648, 366, 300, 434...
$ mismatches    (dbl) 80, 22, 12, 85, 0, 58, 14, 22, 8, 34, 4, 6, ...
$ gap_openings  (dbl) 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, 1, 3, 2, 1...
$ q_start       (dbl) 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2,...
$ q_end         (dbl) 430, 246, 359, 1910, 213, 646, 366, 294, 429...
$ s_start       (dbl) 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4...
$ s_end         (dbl) 466, 246, 355, 1963, 213, 640, 362, 299, 433...
$ evalue        (dbl) 0e+00, 7e-166, 0e+00, 0e+00, 2e-160, 0e+00, ...
$ bit_score     (dbl) 627, 454, 698, 3704, 437, 1037, 696, 491, 85...
</code></pre>
</div>
<div id="the-blast_rec-function" class="section level2">
<h2>The blast_rec() function</h2>
<p>The <code>blast_rec()</code> function was implemented to optimize BLAST+ reciprocal best hit searches (only based on the minimum e-value). BLAST+ reciprocal best hit searches are used to perform orthology inference.</p>
<p>Running <code>blast_rec()</code> using default parameter settings:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># performing gene orthology inference using the reciprocal best hit (RBH) method</span>
<span class="kw">blast_rec</span>(<span class="dt">query_file   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))</code></pre></div>
<pre><code>      query_id            subject_id evalue
1  AT1G01010.1 333554|PACid:16033839  0e+00
2  AT1G01020.1 470181|PACid:16064328 7e-166
3  AT1G01030.1 470180|PACid:16054974  0e+00
4  AT1G01040.1 333551|PACid:16057793  0e+00
5  AT1G01050.1 909874|PACid:16064489 2e-160
6  AT1G01060.3 470177|PACid:16043374  0e+00
7  AT1G01070.1 918864|PACid:16052578  0e+00
8  AT1G01080.1 909871|PACid:16053217 1e-178
9  AT1G01090.1 470171|PACid:16052860  0e+00
10 AT1G01110.2 333544|PACid:16034284  0e+00
11 AT1G01120.1 918858|PACid:16049140  0e+00
12 AT1G01140.3 470161|PACid:16036015  0e+00
13 AT1G01150.1 918855|PACid:16037307 3e-150
14 AT1G01160.1 918854|PACid:16044153  1e-93
15 AT1G01170.2 311317|PACid:16052302  3e-54
16 AT1G01180.1 909860|PACid:16056125  0e+00
17 AT1G01190.1 311315|PACid:16059488  0e+00
18 AT1G01200.1 470156|PACid:16041002 3e-172
19 AT1G01210.1 311313|PACid:16057125  7e-76
20 AT1G01220.1 470155|PACid:16047984  0e+00
</code></pre>
<p>Again you can specify the <code>detailed_output</code> argument to get more parameters returned by <code>blast_rec()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># running blast_rec() using detailed_output = TRUE</span>
rbh &lt;-<span class="st"> </span><span class="kw">blast_rec</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file    =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">detailed_output =</span> <span class="ot">TRUE</span>)


dplyr::<span class="kw">glimpse</span>(rbh)</code></pre></div>
<pre><code>Variables:
$ query_id      (chr) &quot;AT1G01010.1&quot;, &quot;AT1G01020.1&quot;, &quot;AT1G01030.1&quot;,...
$ subject_id    (chr) &quot;333554|PACid:16033839&quot;, &quot;470181|PACid:16064...
$ perc_identity (dbl) 73.99, 91.06, 95.54, 91.98, 100.00, 89.51, 9...
$ alig_length   (dbl) 469, 246, 359, 1970, 213, 648, 366, 300, 434...
$ mismatches    (dbl) 80, 22, 12, 85, 0, 58, 14, 22, 8, 34, 4, 6, ...
$ gap_openings  (dbl) 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, 1, 3, 2, 1...
$ q_start       (dbl) 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2,...
$ q_end         (dbl) 430, 246, 359, 1910, 213, 646, 366, 294, 429...
$ s_start       (dbl) 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4...
$ s_end         (dbl) 466, 246, 355, 1963, 213, 640, 362, 299, 433...
$ evalue        (dbl) 0e+00, 7e-166, 0e+00, 0e+00, 2e-160, 0e+00, ...
$ bit_score     (dbl) 627, 454, 698, 3704, 437, 1037, 696, 491, 85...
</code></pre>
</div>
<div id="the-advanced_blast-function" class="section level2">
<h2>The advanced_blast() function</h2>
<p>The <code>advanced_blast()</code> function was implemented to allow you to perform BLAST+ searches in a flexible environment. All parameters that shall be passed to the corresponding BLAST+ search need to be specified using the <code>blast_params</code> argument. In case you work with very large hit tables, that do not fit into memory, you can specify the <code>sql_database</code> argument to store the corresponding hit table in a SQLite database and access it via the <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/databases.html">dplyr database notation</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ab &lt;-<span class="st"> </span><span class="kw">advanced_blast</span>(<span class="dt">query_file  =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                     <span class="dt">subject_file    =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                     <span class="dt">blast_algorithm =</span> <span class="st">&quot;blastp&quot;</span>, 
                     <span class="dt">blast_params    =</span> <span class="st">&quot;-evalue 1E-5 -num_threads 1&quot;</span>)


dplyr::<span class="kw">glimpse</span>(ab)</code></pre></div>
<pre><code>
Variables:
$ query_id                (chr) &quot;AT1G01010.1&quot;, &quot;AT1G01020.1&quot;, &quot;AT1...
$ subject_id              (chr) &quot;333554|PACid:16033839&quot;, &quot;470181|P...
$ perc_identity           (dbl) 73.99, 91.06, 95.54, 91.98, 100.00...
$ num_ident_matches       (dbl) 347, 224, 343, 1812, 213, 580, 348...
$ alig_length             (dbl) 469, 246, 359, 1970, 213, 648, 366...
$ mismatches              (dbl) 80, 22, 12, 85, 0, 58, 14, 22, 8, ...
$ gap_openings            (dbl) 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, ...
$ q_start                 (dbl) 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1...
$ q_end                   (dbl) 430, 246, 359, 1910, 213, 646, 366...
$ s_start                 (dbl) 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1...
$ s_end                   (dbl) 466, 246, 355, 1963, 213, 640, 362...
$ evalue                  (dbl) 0e+00, 7e-166, 0e+00, 0e+00, 2e-16...
$ bit_score               (dbl) 627, 454, 698, 3704, 437, 1037, 69...
$ score_raw               (dbl) 1617, 1169, 1801, 9604, 1125, 2681...
$ query_coverage_per_subj (dbl) 100, 100, 100, 99, 100, 100, 100, ...
</code></pre>
<p>As you can see, the <code>advanced_blast()</code> function returns more parameters than all other <code>blast()</code> functions.</p>
<div id="selecting-the-best-hit-using-advanced_blast" class="section level3">
<h3>Selecting the best hit using advanced_blast()</h3>
<p>The flexibility of <code>advanced_blast()</code> enables you to perform all previously introduced analyses (<code>blast()</code>, <code>blast_best()</code>, <code>blast_rec()</code>, etc.) as well as a variety of addtional analyses. The following example shall illustrate how to perform BLAST best hit searches using <code>advanced_blast()</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># when performing an advanced BLAST search, you can easily select the best hit</span>
<span class="kw">library</span>(dplyr)

advB &lt;-<span class="st"> </span><span class="kw">advanced_blast</span>(
           <span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
           <span class="dt">subject_file    =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
           <span class="dt">seq_type        =</span> <span class="st">&quot;cds&quot;</span>,
           <span class="dt">blast_algorithm =</span> <span class="st">&quot;blastp&quot;</span>,
           <span class="dt">blast_params    =</span> <span class="st">&quot;-evalue 1E-5 -num_threads 1&quot;</span>)

best_hit &lt;-<span class="st"> </span>advB %&gt;%<span class="st"> </span><span class="kw">group_by</span>(query_id) %&gt;%<span class="st"> </span><span class="kw">do</span>(<span class="kw">min</span>(evalue))


best_hit</code></pre></div>
<pre><code>
      query_id min(evalue)
1  AT1G01010.1       0e+00
2  AT1G01020.1      3e-164
3  AT1G01030.1       0e+00
4  AT1G01040.1       0e+00
5  AT1G01050.1      1e-158
6  AT1G01060.3       0e+00
7  AT1G01070.1       0e+00
8  AT1G01080.1      5e-177
9  AT1G01090.1       0e+00
10 AT1G01110.2       0e+00
11 AT1G01120.1       0e+00
12 AT1G01140.3       0e+00
13 AT1G01150.1      1e-148
14 AT1G01160.1       5e-92
15 AT1G01170.2       1e-52
16 AT1G01180.1       0e+00
17 AT1G01190.1       0e+00
18 AT1G01200.1      2e-170
19 AT1G01210.1       3e-74
20 AT1G01220.1       0e+00
</code></pre>
<p>The <code>blast_params</code> argument allows you to specify all parameters that shall be passed to the corresponding <code>blast_algorithm</code> based on the <a href="http://www.ncbi.nlm.nih.gov/books/NBK1763/">NCBI blast stand-alone notation</a>.</p>
</div>
<div id="using-delta-blast-in-advanced_blast" class="section level3">
<h3>Using DELTA-BLAST in advanced_blast()</h3>
<p><a href="http://www.biomedcentral.com/content/pdf/1745-6150-7-12.pdf">Domain enhanced lookup time accelerated BLAST (DELTA-BLAST)</a> is a new BLAST algorithm provided by NCBI. It was introduced as a useful program for the detection of remote protein homologs.</p>
<p>You can download the Conserved Domain Database (CDD) file <code>cdd.tar.gz</code> from <a href="ftp://ftp.ncbi.nlm.nih.gov/pub/mmdb/cdd">NCBI</a> and store all <code>cdd_deltablast.* files</code> in a folder: <code>path/to/cdd_database/folder</code>.</p>
<p>More information on how to install and use DELTA-BLAST can be found <a href="http://www.ncbi.nlm.nih.gov/books/NBK1763/#CmdLineAppsManual.Performing_a_DELTABLAS">here</a>. Make sure that when using DELTA-BLAST in <code>advanced_blast()</code> the <code>db_path</code> argument denoting the path to the folder storing the corresponding <code>cdd_deltablast.* files</code> need to be specified.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># DELTA-BLAST</span>
<span class="co">#</span>
<span class="co"># you can also use deltablast to perform BLAST searches</span>
<span class="co"># make sure you have specified the the db_path argument to the cdd_deltablast.* files</span>
<span class="kw">advanced_blast</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
               <span class="dt">subject_file    =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
               <span class="dt">blast_algorithm =</span> <span class="st">&quot;deltablast&quot;</span>,
               <span class="dt">db_path         =</span> <span class="st">&quot;path/to/cdd_files&quot;</span>,
               <span class="dt">blast_params    =</span> <span class="st">&quot;-evalue 1E-5 -num_threads 2&quot;</span>)</code></pre></div>
<pre><code>Adding sequences from FASTA; added 20 sequences in 0.010026 seconds.
       query_id            subject_id perc_identity num_ident_matches alig_length mismatches
 1: AT1G01010.1 333554|PACid:16033839         73.77               346         469         81
 2: AT1G01020.1 470181|PACid:16064328         91.06               224         246         22
 3: AT1G01030.1 470180|PACid:16054974         95.26               342         359         13
 4: AT1G01040.1 333551|PACid:16057793         91.57              1803        1969         95
 5: AT1G01050.1 909874|PACid:16064489        100.00               213         213          0
 6: AT1G01060.3 470177|PACid:16043374         89.51               580         648         58
 7: AT1G01070.1 918864|PACid:16052578         93.44               342         366         20
 8: AT1G01080.1 909871|PACid:16053217         78.60               235         299         59
 9: AT1G01090.1 470171|PACid:16052860         96.77               420         434          8
10: AT1G01110.2 333544|PACid:16034284         93.56               494         528         34
11: AT1G01120.1 918858|PACid:16049140         99.24               525         529          4
12: AT1G01140.3 470161|PACid:16036015         98.45               446         453          6
13: AT1G01150.1 470177|PACid:16043374         13.33                10          75         51
14: AT1G01150.1 918855|PACid:16037307         71.88               207         288         71
15: AT1G01160.1 918854|PACid:16044153         78.65               140         178         32
16: AT1G01170.2 311317|PACid:16052302         85.57                83          97          0
17: AT1G01180.1 909860|PACid:16056125         92.58               287         310         20
18: AT1G01190.1 311315|PACid:16059488         94.00               501         533         31
19: AT1G01200.1 470156|PACid:16041002         95.80               228         238         10
20: AT1G01210.1 311313|PACid:16057125         95.33               102         107          5
21: AT1G01220.1 470155|PACid:16047984         96.69              1021        1056         35
       query_id            subject_id perc_identity num_ident_matches alig_length mismatches
    gap_openings q_start q_end s_start s_end evalue bit_score score_raw query_coverage_per_subj
 1:            9       1   430       1   466  0e+00       637      1643                     100
 2:            0       1   246       1   246  3e-88       257       657                     100
 3:            2       1   359       1   355  0e+00       502      1294                     100
 4:            8       6  1910       2  1963  0e+00      2362      6123                      99
 5:            0       1   213       1   213 2e-128       357       917                     100
 6:            5       1   646       1   640  0e+00      1003      2592                     100
 7:            2       1   366       1   362  2e-68       214       544                     100
 8:            1       1   294       1   299  4e-77       232       592                     100
 9:            3       1   429       1   433  0e+00       658      1698                     100
10:            0       1   528       1   528  0e+00       763      1971                     100
11:            0       1   529       1   529  0e+00       747      1928                     100
12:            1       1   452       1   453 9e-179       504      1298                     100
13:            3     215   289       8    68  1e-07        43       100                      22
14:            3       5   292      16   293 3e-120       344       883                      83
15:            1       4   175       2   179  4e-45       143       361                      88
16:            1       2    84       4   100  5e-54       158       399                      99
17:            1       1   307       1   310 8e-156       441      1135                      64
18:            1       1   533       6   537  0e+00       522      1345                      99
19:            0       1   238       1   238 2e-114       323       828                     100
20:            0       1   107       1   107  2e-40       124       313                     100
21:            0       1  1056       1  1056  0e+00      1765      4571                     100
    gap_openings q_start q_end s_start s_end evalue bit_score score_raw query_coverage_per_subj
</code></pre>
<p>A more intuitive way (but less flexible) to perform delta-blast searches is to use the <code>delta.blast()</code> function. The <code>delta.blast()</code> function works in two ways.</p>
<ol style="list-style-type: decimal">
<li><p>Use the <code>cdd_delta</code> database as additional information by performing <code>deltablast</code> searches between <code>query organism A</code> and <code>subject organism B</code>.</p></li>
<li><p>Perform a <code>deltablast</code> search between <code>query organism A</code> and the <code>cdd_delta</code> database</p></li>
</ol>
<p>The following example illustrates the first option (A vs B via cdd_delta).</p>
<p>Here the <code>cdd.path</code> argument specifies the path to the <code>cdd_deltablast.* files</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># perform a delta-blast serach between A. thaliana and A. lyrata genes</span>
<span class="kw">delta.blast</span>(
            <span class="dt">query_file   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
            <span class="dt">subject_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
            <span class="dt">cdd.path     =</span> <span class="st">&quot;path/to/cdd/database/folder&quot;</span>,
            <span class="dt">comp_cores   =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>
       query_id            subject_id perc_identity num_ident_matches alig_length mismatches gap_openings q_start q_end s_start s_end evalue
 1: AT1G01010.1 333554|PACid:16033839         73.77               346         469         81            9       1   430       1   466  0e+00
 2: AT1G01020.1 470181|PACid:16064328         91.06               224         246         22            0       1   246       1   246  3e-88
 3: AT1G01030.1 470180|PACid:16054974         95.26               342         359         13            2       1   359       1   355  0e+00
 4: AT1G01040.1 333551|PACid:16057793         91.57              1803        1969         95            8       6  1910       2  1963  0e+00
 5: AT1G01050.1 909874|PACid:16064489        100.00               213         213          0            0       1   213       1   213 2e-128
 6: AT1G01060.3 470177|PACid:16043374         89.51               580         648         58            5       1   646       1   640  0e+00
 7: AT1G01070.1 918864|PACid:16052578         93.44               342         366         20            2       1   366       1   362  2e-68
 8: AT1G01080.1 909871|PACid:16053217         78.60               235         299         59            1       1   294       1   299  4e-77
 9: AT1G01090.1 470171|PACid:16052860         96.77               420         434          8            3       1   429       1   433  0e+00
10: AT1G01110.2 333544|PACid:16034284         93.56               494         528         34            0       1   528       1   528  0e+00
11: AT1G01120.1 918858|PACid:16049140         99.24               525         529          4            0       1   529       1   529  0e+00
12: AT1G01140.3 470161|PACid:16036015         98.45               446         453          6            1       1   452       1   453 9e-179
13: AT1G01150.1 470177|PACid:16043374         13.33                10          75         51            3     215   289       8    68  1e-07
14: AT1G01150.1 918855|PACid:16037307         71.88               207         288         71            3       5   292      16   293 3e-120
15: AT1G01160.1 918854|PACid:16044153         78.65               140         178         32            1       4   175       2   179  4e-45
16: AT1G01170.2 311317|PACid:16052302         85.57                83          97          0            1       2    84       4   100  5e-54
17: AT1G01180.1 909860|PACid:16056125         92.58               287         310         20            1       1   307       1   310 8e-156
18: AT1G01190.1 311315|PACid:16059488         94.00               501         533         31            1       1   533       6   537  0e+00
19: AT1G01200.1 470156|PACid:16041002         95.80               228         238         10            0       1   238       1   238 2e-114
20: AT1G01210.1 311313|PACid:16057125         95.33               102         107          5            0       1   107       1   107  2e-40
21: AT1G01220.1 470155|PACid:16047984         96.69              1021        1056         35            0       1  1056       1  1056  0e+00
       query_id            subject_id perc_identity num_ident_matches alig_length mismatches gap_openings q_start q_end s_start s_end evalue
    bit_score score_raw query_coverage_per_subj
 1:       637      1643                     100
 2:       257       657                     100
 3:       502      1294                     100
 4:      2362      6123                      99
 5:       357       917                     100
 6:      1003      2592                     100
 7:       214       544                     100
 8:       232       592                     100
 9:       658      1698                     100
10:       763      1971                     100
11:       747      1928                     100
12:       504      1298                     100
13:        43       100                      22
14:       344       883                      83
15:       143       361                      88
16:       158       399                      99
17:       441      1135                      64
18:       522      1345                      99
19:       323       828                     100
20:       124       313                     100
21:      1765      4571                     100
    bit_score score_raw query_coverage_per_subj</code></pre>
<p>The following example illustrates the second option (A vs cdd_delta). The difference is the specification of argument <code>subject_file = &quot;cdd_delta&quot;</code> instead of a second subject organism:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># perform a delta-blast serach between A. thaliana and the cdd database</span>
<span class="kw">delta.blast</span>(
            <span class="dt">query_file   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
            <span class="dt">subject_file =</span> <span class="st">&quot;cdd_delta&quot;</span>,
            <span class="dt">cdd.path     =</span> <span class="st">&quot;path/to/cdd/database/folder&quot;</span>,
            <span class="dt">comp_cores   =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>        query_id     subject_id perc_identity num_ident_matches alig_length mismatches gap_openings q_start q_end s_start s_end evalue
  1: AT1G01010.1 gnl|CDD|251251         40.15                55         137         71            4       5   139       3   130  5e-79
  2: AT1G01020.1 gnl|CDD|252420         35.82                72         201        109            5       6   193       1   194  4e-55
  3: AT1G01030.1 gnl|CDD|197383         29.70                30         101         66            1      54   154       1    96  1e-42
  4: AT1G01030.1 gnl|CDD|214977         35.64                36         101         57            2      56   156       1    93  3e-39
  5: AT1G01030.1 gnl|CDD|251248         37.62                38         101         59            2      56   156       1    97  4e-44
 ---                                                                                                                                  
828: AT1G01220.1 gnl|CDD|235243         15.28                22         144        105            3     742   879       9   141  1e-09
829: AT1G01220.1 gnl|CDD|235407         17.91                65         363        248           14     717  1046      24   369  4e-70
830: AT1G01220.1 gnl|CDD|237766         13.83                39         282        187           10     757  1035      32   260  2e-13
831: AT1G01220.1 gnl|CDD|249743         35.29                24          68         43            1     818   885       1    67  1e-17
832: AT1G01220.1 gnl|CDD|256035         33.33                11          33         16            1     716   748      17    43  1e-10
     bit_score score_raw query_coverage_per_subj
  1:     243.0       620                      31
  2:     177.0       450                      76
  3:     145.0       366                      28
  4:     136.0       342                      28
  5:     149.0       376                      28
 ---                                            
828:      59.2       142                      13
829:     239.0       610                      31
830:      70.8       172                      26
831:      79.6       195                       6
832:      59.2       142                       3
</code></pre>
</div>
<div id="performing-blastp-searches-against-the-ncbi-nr-database-with-blast.nr" class="section level3">
<h3>Performing BLASTP searches against the NCBI nr database with <code>blast.nr()</code></h3>
<p>The <code>blast.nr()</code> function implemented in <code>orthologr</code> allows users to perform local <code>blastp</code> searches against the NCBI nr database.</p>
<p>For this purpose, the first step is to download the NCBI nr database to a local machine using the <a href="http://cran.r-project.org/package=biomartr">biomartr</a> package.</p>
<p>Please notice that the NCBI nr database contains about 30 files storing approx. 500 MB data per file. So when downloading this large database via <code>biomartr</code> users need to make sure that they are constantly connected to the internet.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install the 'biomartr' package from CRAN</span>
<span class="kw">install.packages</span>(<span class="st">&quot;biomartr&quot;</span>)

<span class="co"># load biomartr</span>
<span class="kw">library</span>(<span class="st">&quot;biomartr&quot;</span>)

<span class="co"># download the NCBI nr database to your local machine to a folder</span>
<span class="co"># names 'DB'</span>
<span class="kw">sapply</span>(<span class="kw">listDatabases</span>(<span class="st">&quot;nr&quot;</span>),download_database)</code></pre></div>
<p>After downloading all NCBI nr files to the <code>DB</code> folder users need to go to the <code>DB</code> folder and unpack all <code>env_nr*</code> files and store them in the same <code>DB</code> folder. In case users whish to specify a different folder for downloading NCBI nr they can specify the <code>path</code> argument of the <code>?download_database</code> function implemented in <code>biomartr</code>.</p>
<p>Now users can use the <code>blast.nr()</code> function to perform BLASTP searches against NCBI nr.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nr.res &lt;-<span class="st"> </span><span class="kw">blast.nr</span>( <span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/aa_seqs.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                    <span class="dt">nr.path         =</span> <span class="st">&quot;DB&quot;</span>,
                    <span class="dt">seq_type        =</span> <span class="st">&quot;protein&quot;</span>,
                    <span class="dt">max.target.seqs =</span> <span class="dv">20</span>,
                    <span class="dt">comp_cores      =</span> <span class="dv">2</span> )

nr.res</code></pre></div>
<pre><code>                query_id                  subject_id perc_identity num_ident_matches alig_length mismatches gap_openings
1: 333554|PACid:16033839 gi|139291046|gb|ECE46929.1|         40.00                24          60         35            1
2: 333554|PACid:16033839 gi|140103111|gb|ECJ55606.1|         46.03                29          63         34            0
   n_gaps pos_match ppos q_start q_end q_len qcov qcovhsp                                                       query_seq
1:      1        39 65.0       8    66   246   24      24    CVGC-GFRVKSLFIQYSPGNIRLMKCGNCKEVADEYIECERMIIFIDLILHRPKVYRHVL
2:      0        47 74.6       8    70   246   26      26 CVGCGFRVKSLFIQYSPGNIRLMKCGNCKEVADEYIECERMIIFIDLILHRPKVYRHVLYNAI
   s_start s_end s_len                                                     subject_seq evalue bit_score score_raw
1:       9    68    71    CVECMCDENESIYKKYSEGNLRLTRCIRCSDFVDRYVEYDNVLIVLDLVLHKNPAYRHLL  1e-07      54.3       129
2:      48   110   114 CVECGKSVQQVFKMFGKGNIRLSRCKSCHSISDKYVEYEFVLIFIDLLLHKTQVYRHLIFNRL  8e-14      72.0       175</code></pre>
<p>Here, the <code>nr.path</code> argument specifies the folder path to the NCBI nr database. In case users unpacked all NCBI nr files to the <code>DB</code> folder created by <code>download_database()</code> the <code>nr.path</code> argument can be specified as <code>nr.path = &quot;DB&quot;</code>.</p>
<p>The corresponding <code>gi id</code> can then be clipped by running:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">nr.res[ , <span class="st">&quot;subject_id&quot;</span>] &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">lapply</span>(stringr::<span class="kw">str_split</span>(nr.res[ , subject_id],<span class="st">&quot;[|]&quot;</span>), function(x) x[<span class="dv">2</span>]))

nr.res</code></pre></div>
<pre><code>                query_id subject_id perc_identity num_ident_matches alig_length mismatches
1: 333554|PACid:16033839  139291046         40.00                24          60         35
2: 333554|PACid:16033839  140103111         46.03                29          63         34
   gap_openings n_gaps pos_match ppos q_start q_end q_len qcov qcovhsp
1:            1      1        39 65.0       8    66   246   24      24
2:            0      0        47 74.6       8    70   246   26      26
                                                         query_seq s_start s_end s_len
1:    CVGC-GFRVKSLFIQYSPGNIRLMKCGNCKEVADEYIECERMIIFIDLILHRPKVYRHVL       9    68    71
2: CVGCGFRVKSLFIQYSPGNIRLMKCGNCKEVADEYIECERMIIFIDLILHRPKVYRHVLYNAI      48   110   114
                                                       subject_seq evalue bit_score score_raw
1:    CVECMCDENESIYKKYSEGNLRLTRCIRCSDFVDRYVEYDNVLIVLDLVLHKNPAYRHLL  1e-07      54.3       129
2: CVECGKSVQQVFKMFGKGNIRLSRCKSCHSISDKYVEYEFVLIFIDLLLHKTQVYRHLIFNRL  8e-14      72.0       175</code></pre>
<p>By performing this <code>gi</code> clipping the initial <code>gi id</code> returned by blasting against the <code>NCBI nr</code> database <code>gi|139291046|gb|ECE46929.1|</code> is clipped to the <code>gi id</code> <code>139291046</code> for which corresponding <code>taxonomy ids</code> can be extracted from <a href="ftp://ftp.ncbi.nlm.nih.gov/pub/taxonomy/">NCBI Taxonomy</a>.</p>
</div>
<div id="storing-very-large-hit-tables-in-a-sqlite-database-with-advanced_blast" class="section level3">
<h3>Storing very large hit tables in a SQLite database with advanced_blast()</h3>
<p>The <code>advanced_blast()</code> function can also store the BLAST output CSV file in a SQLite database. This works only with <code>dplyr</code> version &gt;= 0.3 . To store the BLAST output in an SQLite database and to receive an SQLite connection as specified by <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/databases.html">tbl()</a> in dplyr please use the <code>sql_database = TRUE</code> argument.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># using a SQLite database to store the BLAST output and</span>
<span class="co"># select some example data</span>

<span class="kw">library</span>(dplyr)

sqlE &lt;-<span class="st"> </span><span class="kw">advanced_blast</span>(
             <span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
             <span class="dt">subject_file    =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
             <span class="dt">seq_type        =</span> <span class="st">&quot;cds&quot;</span>,
             <span class="dt">blast_algorithm =</span> <span class="st">&quot;blastp&quot;</span>,
             <span class="dt">blast_params    =</span> <span class="st">&quot;-evalue 1E-5 -num_threads 1&quot;</span>, 
             <span class="dt">sql_database    =</span> <span class="ot">TRUE</span>)

<span class="kw">head</span>(sqlE)</code></pre></div>
<pre><code>           V1                    V2    V3  V4  V5  V6 V7 V8  V9 V10 V11    V12   V13  V14 V15
1 AT1G01010.1 333554|PACid:16033839 73.99 347 469  80  8  1 430   1 466  0e+00 627.0 1617 100
2 AT1G01010.1 909883|PACid:16051052 34.46 122 354 162 12  1 343   1 295  7e-43 152.0  384  80
3 AT1G01010.1 311334|PACid:16035114 43.64  72 165  81  4  1 162   1 156  1e-32 123.0  309  38
4 AT1G01010.1 918785|PACid:16057277 33.75  54 160  80  6  6 153  10 155  8e-21  87.8  216  34
5 AT1G01020.1 470181|PACid:16064328 91.06 224 246  22  0  1 246   1 246 3e-164 454.0 1169 100
6 AT1G01030.1 470180|PACid:16054974 95.54 343 359  12  2  1 359   1 355  0e+00 698.0 1801 100

</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at the structure of sqlE</span>
<span class="kw">glimpse</span>(sqlE)</code></pre></div>
<pre><code>Variables:
$ V1  (chr) &quot;AT1G01010.1&quot;, &quot;AT1G01010.1&quot;, &quot;AT1G01010.1&quot;, &quot;AT1G01010.1&quot;, &quot;AT1G01020.1...
$ V2  (chr) &quot;333554|PACid:16033839&quot;, &quot;909883|PACid:16051052&quot;, &quot;311334|PACid:16035114...
$ V3  (dbl) 73.99, 34.46, 43.64, 33.75, 91.06, 95.54, 91.98, 100.00, 89.51, 64.52, 9...
$ V4  (int) 347, 122, 72, 54, 224, 343, 1812, 213, 580, 40, 348, 268, 117, 271, 29, ...
$ V5  (int) 469, 354, 165, 160, 246, 359, 1970, 213, 648, 62, 366, 338, 356, 300, 95...
$ V6  (int) 80, 162, 81, 80, 22, 12, 85, 0, 58, 22, 14, 68, 232, 22, 66, 91, 47, 8, ...
$ V7  (int) 8, 12, 4, 6, 0, 2, 10, 0, 5, 0, 2, 1, 3, 2, 0, 3, 0, 3, 0, 0, 5, 5, 1, 6...
$ V8  (int) 1, 1, 1, 6, 1, 1, 6, 1, 1, 13, 1, 1, 16, 1, 105, 109, 110, 1, 1, 1, 32, ...
$ V9  (int) 430, 343, 162, 153, 246, 359, 1910, 213, 646, 74, 366, 338, 365, 294, 19...
$ V10 (int) 1, 1, 1, 10, 1, 1, 2, 1, 1, 1048, 1, 1, 8, 1, 336, 13, 76, 1, 1, 1, 20, ...
$ V11 (int) 466, 295, 156, 155, 246, 355, 1963, 213, 640, 1109, 362, 336, 362, 299, ...
$ V12 (dbl) 0e+00, 7e-43, 1e-32, 8e-21, 3e-164, 0e+00, 0e+00, 1e-158, 0e+00, 2e-22, ...
$ V13 (dbl) 627.0, 152.0, 123.0, 87.8, 454.0, 698.0, 3704.0, 437.0, 1037.0, 98.2, 69...
$ V14 (int) 1617, 384, 309, 216, 1169, 1801, 9604, 1125, 2681, 243, 1797, 1409, 551,...
$ V15 (int) 100, 80, 38, 34, 100, 100, 99, 100, 100, 10, 100, 92, 96, 100, 48, 48, 2...
</code></pre>
<p>This way you can perfom filtering steps on very large tables.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select all rows that have an evalue of zero</span>
<span class="kw">filter</span>(sqlE, V12 &gt;<span class="st"> </span><span class="fl">1e-15</span>)</code></pre></div>
<pre><code>Source: sqlite 3.8.6 [blast_sql_db.sqlite3]
From: hit_tbl [64 x 15]
Filter: V12 &gt; 1e-15 

            V1                    V2     V3  V4  V5  V6 V7  V8  V9  V10  V11    V12   V13  V14 V15
1  AT1G01010.1 909883|PACid:16051052  34.46 122 354 162 12   1 343    1  295  7e-43 152.0  384  80
2  AT1G01010.1 311334|PACid:16035114  43.64  72 165  81  4   1 162    1  156  1e-32 123.0  309  38
3  AT1G01010.1 918785|PACid:16057277  33.75  54 160  80  6   6 153   10  155  8e-21  87.8  216  34
4  AT1G01020.1 470181|PACid:16064328  91.06 224 246  22  0   1 246    1  246 3e-164 454.0 1169 100
5  AT1G01050.1 909874|PACid:16064489 100.00 213 213   0  0   1 213    1  213 1e-158 437.0 1125 100
6  AT1G01060.3 311286|PACid:16063880  64.52  40  62  22  0  13  74 1048 1109  2e-22  98.2  243  10
7  AT1G01070.1 919693|PACid:16048878  32.87 117 356 232  3  16 365    8  362  2e-67 216.0  551  96
8  AT1G01080.1 909871|PACid:16053217  90.33 271 300  22  2   1 294    1  299 5e-177 491.0 1264 100
9  AT1G01080.1 470321|PACid:16050357  30.53  29  95  66  0 105 199  336  430  6e-11  58.2  139  48
10 AT1G01080.1 470321|PACid:16050357  25.18  35 139  91  3 109 244   13  141  4e-06  43.1  100  48
..         ...                   ...    ... ... ... ... .. ... ...  ...  ...    ...   ...  ... ...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># select the best hit using the evalue criterion</span>
sqlE %&gt;%<span class="st"> </span><span class="kw">group_by</span>(V1) %&gt;%<span class="st"> </span><span class="kw">do</span>(<span class="dt">best_hit_eval =</span> <span class="kw">min</span>(V12))</code></pre></div>
<pre><code>Source: sqlite 3.8.6 [blast_sql_db.sqlite3]
From: &lt;derived table&gt; [?? x 2]

            V1 best_hit_eval
1  AT1G01010.1         0e+00
2  AT1G01020.1        3e-164
3  AT1G01030.1         0e+00
4  AT1G01040.1         0e+00
5  AT1G01050.1        1e-158
6  AT1G01060.3         0e+00
7  AT1G01070.1         0e+00
8  AT1G01080.1        5e-177
9  AT1G01090.1         0e+00
10 AT1G01110.2         0e+00
..         ...           ...</code></pre>
<p>Since the BLAST output does not return a CSV header, the corresponding SQLite table the BLAST output in columns named V1 - V15. In SQLite, renaming the columns can only be done by creating a new table which is very time consuming. The dplyr::rename() function does not support to rename multiple columns and therefore this “not so elegent way” is implemented in the current version of <code>orthologr</code>. Future versions will tackle this issue.</p>
</div>
<div id="the-set_blast-function" class="section level3">
<h3>The set_blast() function</h3>
<p>The <code>set_blast()</code>function reads a file storing a specific sequence type, such as “cds”, “protein”, or “dna” in a standard sequence file format such as “fasta”, etc. and depending of the makedb parameter either creates a blast-able database, or returns the corresponding protein sequences as data.table object for further BLAST searches.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">set_blast</span>(<span class="dt">file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))[[<span class="dv">1</span>]] , <span class="dv">2</span>)</code></pre></div>
<pre><code>
       geneids
1: AT1G01010.1
2: AT1G01020.1

seqs
1: atggaggatcaagttgggtttgggttccgtccgaacgacgaggagctcgttggtcactatctccgtaacaaaatcgaaggaaacact
agccgcgacgttgaagtagccatcagcgaggtcaacatctgtagctacgatccttggaacttgcgcttccagtcaaagtacaaatcgaga
gatgctatgtggtacttcttctctcgtagagaaaacaacaaagggaatcgacagagcaggacaacggtttctggtaaatggaagcttacc
ggagaatctgttgaggtcaaggaccagtggggattttgtagtgagggctttcgtggtaagattggtcataaaagggttttggtgttcctc
gatggaagataccctgacaaaaccaaatctgattgggttatccacgagttccactacgacctcttaccagaacatcagaggacatatgtc
atctgcagacttgagtacaagggtgatgatgcggacattctatctgcttatgcaatagatcccactcccgcttttgtccccaatatgact
agtagtgcaggttctgtggtcaaccaatcacgtcaacgaaattcaggatcttacaacacttactctgagtatgattcagcaaatcatggc
cagcagtttaatgaaaactctaacattatgcagcagcaaccacttcaaggatcattcaaccctctccttgagtatgattttgcaaatcac
ggcggtcagtggctgagtgactatatcgacctgcaacagcaagttccttacttggcaccttatgaaaatgagtcggagatgatttggaag
catgtgattgaagaaaattttgagtttttggtagatgaaaggacatctatgcaacagcattacagtgatcaccggcccaaaaaacctgtg
tctggggttttgcctgatgatagcagtgatactgaaactggatcaatgattttcgaagacacttcgagctccactgatagtgttggtagt
tcagatgaaccgggccatactcgtatagatgatattccatcattgaacattattgagcctttgcacaattataaggcacaagagcaacca
aagcagcagagcaaagaaaaggtgataagttcgcagaaaagcgaatgcgagtggaaaatggctgaagactcgatcaagatacctccatcc
accaacacggtgaagcagagctggattgttttggagaatgcacagtggaactatctcaagaacatgatcattggtgtcttgttgttcatc
tccgtcattagttggatcattcttgttggttaa
2:
atggcggcgagtgaacacagatgcgtgggatgtggttttagggtaaagtcattgttcattcaatactctccgggtaacattcgtctcatg
aaatgcggaaattgcaaggaagtagcagatgagtacatcgagtgtgaacgcatgattattttcatcgatttaatccttcacagaccaaag
gtatatagacacgtcctctacaatgcaattaatccagcaactgtcaatattcagcatctgttgtggaagttggtcttcgcctatcttctt
ctagactgttatagaagcttgctactgagaaaaagtgatgaagaatcgagcttttctgatagccctgttcttctatctataaaggttctg
attggtgtcttatctgcaaacgctgcatttatcatctcttttgccattgcgactaagggtttgctaaatgaagtttccagaagaagagag
attatgttggggatattcatctctagttacttcaagatatttctgcttgcgatgttggtatgggaattcccaatgtcagtgatttttttt
gtcgatatacttctcttaacatcaaactccatggctcttaaagtgatgactgaatcaacaatgaccagatgcatagccgtatgcttaatc
gcgcacttgattagattcttggtgggtcagatttttgagccgacaatatttttgatacaaattggatctctgttgcaatatatgtcttat
tttttcagaatcgtatga

aa
1: MEDQVGFGFRPNDEELVGHYLRNKIEGNTSRDVEVAISEVNICSYDPWNLRFQSKYKSRDAMWYFFSRRENNKGNRQSRTTVSGKWK
LTGESVEVKDQWGFCSEGFRGKIGHKRVLVFLDGRYPDKTKSDWVIHEFHYDLLPEHQRTYVICRLEYKGDDADILSAYAIDPTPAFVPN
MTSSAGSVVNQSRQRNSGSYNTYSEYDSANHGQQFNENSNIMQQQPLQGSFNPLLEYDFANHGGQWLSDYIDLQQQVPYLAPYENESEMI
WKHVIEENFEFLVDERTSMQQHYSDHRPKKPVSGVLPDDSSDTETGSMIFEDTSSSTDSVGSSDEPGHTRIDDIPSLNIIEPLHNYKAQE
QPKQQSKEKVISSQKSECEWKMAEDSIKIPPSTNTVKQSWIVLENAQWNYLKNMIIGVLLFISVISWIILVG*
2:
MAASEHRCVGCGFRVKSLFIQYSPGNIRLMKCGNCKEVADEYIECERMIIFIDLILHRPKVYRHVLYNAINPATVNIQHLLWKLVFAYLL
LDCYRSLLLRKSDEESSFSDSPVLLSIKVLIGVLSANAAFIISFAIATKGLLNEVSRRREIMLGIFISSYFKIFLLAMLVWEFPMSVIFF
VDILLLTSNSMALKVMTESTMTRCIAVCLIAHLIRFLVGQIFEPTIFLIQIGSLLQYMSYFFRIV*</code></pre>
</div>
<div id="the-advanced_makedb-function" class="section level3">
<h3>The advanced_makedb() function</h3>
<p>The <code>advanced_makedb</code> function provides a simple, but powerful interface between the R language and <code>makeblastdb</code>. You can specify the <code>params</code> argument to pass all parameters defined for <code>makeblastdb</code> to the corresponding <code>makeblastdb</code> call.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># make the A. thaliana genome to a blast-able database</span>
<span class="kw">advanced_makedb</span>( <span class="dt">database_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">params        =</span> <span class="st">&quot;-input_type fasta -dbtype prot -hash_index&quot;</span> )</code></pre></div>
<pre><code>Building a new DB, current time: 11/10/2014 16:56:58
New DB name:   _blast_db/ortho_thal_aa.fasta
New DB title:  _blast_db/ortho_thal_aa.fasta
Sequence type: Protein
Keep Linkouts: T
Keep MBits: T
Maximum file size: 1000000000B
Adding sequences from FASTA; added 20 sequences in 0.0583858 seconds.
</code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
