<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="date" content="2016-03-04" />

<title>Installing Prerequisite Tools</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>


<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div class="fluid-row" id="header">


<h1 class="title">Installing Prerequisite Tools</h1>
<h4 class="date"><em>2016-03-04</em></h4>

</div>


<div id="installing-prerequisite-tools" class="section level2">
<h2>Installing Prerequisite Tools</h2>
<p>Most <code>orthologr</code> functions are interface functions that pass data to common bioinformatics tools, internally call the corresponding tool, and read their output as R object. For this purpose, when using interface functions in <code>orthologr</code> users need to install the underlying bioinformatics tools to obain accurate results.</p>
<p>The following sections provide step by step instructions or guidance on installing all bioinformatics tools for which R interface functions are implemented in <code>orthologr</code>.</p>
<p><strong>Some tools are not trivial to install, so please read the corresponding sections carefully and execute test cases that are presented in each section.</strong></p>
</div>
<div id="programming-languages" class="section level2">
<h2>Programming Languages</h2>
<p>The following bioinformatics tools you are going to install are based on the these programming languages:</p>
<ul>
<li><p><a href="http://www.cran.r-project.org"><strong>R</strong></a> &gt;= 3.1.1</p></li>
<li><p><a href="http://isocpp.org/about"><strong>C++11</strong></a></p></li>
<li><p><a href="https://www.perl.org"><strong>Perl</strong></a> &gt;= 5.12</p></li>
</ul>
<p>Please make sure these programming languages are installed and executable on the machines you are going run <code>orthologr</code> on.</p>
</div>
<div id="installing-orthology-inference-tools" class="section level2">
<h2>Installing Orthology Inference Tools</h2>
<p>The <code>orthologr</code> package provides interfaces to the following bioinformatics tools to perform <a href="http://link.springer.com/protocol/10.1007%2F978-1-61779-582-4_9">orthology inference</a>:</p>
<div id="install-blast" class="section level3">
<h3>Install <code>BLAST</code></h3>
<p><a href="http://www.ncbi.nlm.nih.gov/guide/howto/run-blast-local/%5D"><strong>BLAST</strong></a> = Basic Local Alignment Search Tool finds regions of similarity between biological sequences and is also used as underlying paradigm of most fast orthology inference methods.</p>
<ol style="list-style-type: decimal">
<li><p>Go to <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/" class="uri">ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/</a> and download the system specific BLAST program.</p></li>
<li><p>Install BLAST :</p></li>
</ol>
<ul>
<li>On a Windows machine (see <a href="http://www.ncbi.nlm.nih.gov/books/NBK52637/">installation manual: Windows</a>) -&gt; Please carefully read the <code>Environment Variables</code> section of the <code>installation manual: Windows</code> and make sure the execution <code>PATH</code> variable is set correctly.</li>
<li>On a Unix machine (see <a href="http://www.ncbi.nlm.nih.gov/books/NBK52640/">installation manual: Unix</a>) -&gt; Please carefully read the <code>Configuration</code> section of the <code>installation manual: Unix</code> and make sure the execution <code>PATH</code> variable is set correctly to <code>usr/local/bin</code>.</li>
</ul>
<p>For example for Linux systems open the <code>Terminal</code> application and run (Thanks to Alexander Gabel):</p>
<pre><code># download BLAST+ version 2.2.31
wget ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.2.31+-x64-linux.tar.gz

# extract the compiled version of BLAST
tar zxvpf ncbi-blast+2.2.29-x64-linux.tar.gz

# copy BLAST files to `usr/local/bin`
cp ncbi-blast-2.2.31+/bin/* usr/local/bin</code></pre>
<p>Alternatively users can set the system call to the BLAST programs by specifying the <code>PATH</code> variable (this is useful, because it allows an easier update of BLAST versions instead of deleting all BLAST programs from <code>usr/local/bin</code>):</p>
<pre><code># open vim text editor
vi .bash_profile

# type 'Shift' then I to edit the file .bash_profile
# and specify the export PATH
export PATH=${PATH}:/path/to/downloaded/blast/folder/ncbi-blast-2.2.31+/bin

# type 'ESC' then ':' then 'w' then 'q' to save and quit the .bash_profile file 

# log out from your server with
exit

# log in again and type
blastp -version</code></pre>
<p>Now users should see the BLAST command line options.</p>
</div>
<div id="personal-experience" class="section level3">
<h3>Personal Experience</h3>
<p>Based on our personal experience the installation of BLAST works best when copy/pasting the BLAST executables to the path <code>usr/local/bin</code>. In detail you can run the following steps to copy/paste the BLAST executables to <code>usr/local/bin</code> (on Unix systems). However, updating BLAST will then need to manually delete all previous BLAST programs from <code>usr/local/bin</code> :</p>
<p>Open the Terminal application on your system and type:</p>
<pre><code>open /usr/local/bin 
</code></pre>
<p>Next, copy/paste the <code>blastp</code>, <code>makeblastdb</code>, etc files (BLAST executables) from your BLAST folder to <code>/usr/local/bin</code>. To do so you will need to enter the system password to allow the copy process.</p>
<p>After installing the BLAST program you can open an R session and type the following command to check whether or not BLAST can be executed from R.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># test whether blastp is correctly installed on your machine</span>
<span class="kw">system</span>(<span class="st">&quot;blastp -version&quot;</span>)</code></pre></div>
<pre><code>blastp: 2.2.30+
Package: blast 2.2.30, build Oct 27 2014 17:10:51</code></pre>
<p><strong>You should see this output if BLAST was installed correctly.</strong></p>
<p>In case you find the following output:</p>
<pre><code>sh: blastp: command not found</code></pre>
<p>You should return to <code>step 2)</code> and install BLAST so that it can be executed from the default execution <code>PATH</code>.</p>
<p>These interface functions to BLAST+ are implemented in <code>orthologr</code>:</p>
<ul>
<li><code>blast()</code> : Interface function to BLAST+</li>
<li><code>blast_best()</code> : Perform a BLAST+ best hit search</li>
<li><code>blast_rec()</code> : Perform a BLAST+ reciprocal best hit (RBH) search</li>
<li><code>set_blast()</code> : Preparing the parameters and databases for subsequent BLAST+ searches</li>
<li><code>blast.nr()</code> : Perform a BLASTp search against NCBI nr</li>
<li><code>delta.blast()</code> : Perform a DELTA-BLAST Search</li>
<li><code>advanced_blast()</code> : Advanced interface function to BLAST+</li>
<li><code>advanced_makedb()</code> : Advanced interface function to makeblastdb</li>
</ul>
</div>
<div id="install-proteinortho" class="section level3">
<h3>Install <code>ProteinOrtho</code></h3>
<p><a href="https://www.bioinf.uni-leipzig.de/Software/proteinortho/"><strong>ProteinOrtho</strong></a> is an orthology inference tool for large-scale analysis. Please follow <a href="http://www.bioinf.uni-leipzig.de/Software/proteinortho/manual.html">these</a> step by step instructions to install ProteinOrtho so that it is executable from your default execution <code>PATH</code>, e.g. <code>usr/local/bin</code>.</p>
<p>The following interface function to ProteinOrtho is implemented in <code>orthologr</code>:</p>
<ul>
<li><code>ProteinOrtho()</code>: Interface function to ProteinOrtho</li>
</ul>
</div>
<div id="install-orthomcl" class="section level3">
<h3>Install <code>OrthoMCL</code></h3>
<ul>
<li><a href="http://www.orthomcl.org/orthomcl/"><strong>OrthoMCL</strong></a> : An Orthology Inference tool based on the OrthoMCL Algorithm detecting ortholog groups using all-versus all BLAST of all compared protein sequences. Please follow <a href="http://www.orthomcl.org/common/downloads/software/v2.0/UserGuide.txt">these</a> step by step instructions to install OrthoMCL so that it is executable from your default execution <code>PATH</code>, e.g. <code>usr/local/bin</code>.</li>
</ul>
<p>The following interface function to OrthoMCL is implemented in <code>orthologr</code>:</p>
<ul>
<li><code>OrthoMCL()</code>: Interface function to OrthoMCL</li>
</ul>
</div>
<div id="install-ggsearch-and-ssearch" class="section level3">
<h3>Install <code>GGSearch</code> and <code>SSearch</code></h3>
<p>This section will be covered soon.</p>
</div>
<div id="install-hmmer" class="section level3">
<h3>Install <code>HMMER</code></h3>
<p>This section will be covered soon.</p>
</div>
<div id="multiple-alignment-tools" class="section level3">
<h3>Multiple Alignment Tools</h3>
<p>The <code>orthologr</code> package also provides interfaces to the following Multiple Alignment Tools. Nevertheless, non of them have to be installed if the corresponding interface functions are not used.</p>
</div>
<div id="install-clustalw2" class="section level3">
<h3>Install <code>ClustalW2</code></h3>
<p>To install <code>ClustalW2</code> please go to the <a href="http://www.clustal.org/clustal2/">ClustalW</a> homepage and download the corresponding <a href="http://www.clustal.org/download/current/">clustalw2 program</a> matching your operating system.</p>
<p>After downloading and unpacking the <code>clustalw2</code> program, please go to the clustalw-2.1 folder and open a <code>Terminal</code> application to type (in this example for Mac OS X):</p>
<pre><code># copy clustalw2 files to `usr/local/bin`
cp clustalw2 usr/local/bin</code></pre>
</div>
<div id="install-t-coffee" class="section level3">
<h3>Install <code>T-Coffee</code></h3>
<p>To install <code>T-Coffee</code> please go to the <a href="http://www.tcoffee.org/Projects/tcoffee/">T-Coffee</a> homepage and download the corresponding <a href="http://www.tcoffee.org/Packages/Stable/Latest/">T-Coffee program</a> matching your operating system.</p>
</div>
<div id="install-muscle" class="section level3">
<h3>Install <code>MUSCLE</code></h3>
<ul>
<li><a href="http://www.drive5.com/muscle/"><strong>MUSCLE</strong></a> : Fast and accurate multiple alignment tool of nucleic acid and protein sequences</li>
</ul>
</div>
<div id="install-clustalo" class="section level3">
<h3>Install <code>ClustalO</code></h3>
<ol style="list-style-type: decimal">
<li><p>Download the <a href="http://argtable.sourceforge.net/">argtable</a> program.</p></li>
<li><p>Unzip the file.</p></li>
<li><p>Run within the argtable folder:</p></li>
</ol>
<p>``` ./configure</p>
<p>make</p>
<p>make check</p>
<p>sudo make install ```</p>
<ol start="4" style="list-style-type: decimal">
<li><p>Download <a href="http://www.clustal.org/omega/">ClustalO</a>.</p></li>
<li><p>Unzip the folder and run within the folder:</p></li>
</ol>
<p>``` ./configure</p>
<p>make</p>
<p>sudo make install ```</p>
</div>
<div id="install-mafft" class="section level3">
<h3>Install <code>MAFFT</code></h3>
<ul>
<li><a href="http://mafft.cbrc.jp/alignment/software/"><strong>MAFFT</strong></a> : A tool for multiple sequence alignment and phylogeny</li>
</ul>
<p>In <code>orthologr</code> the function <code>multi_aln()</code> provides interfaces to all of these multiple alignment tools as well as an pairwise alignment interface to the <a href="http://www.bioconductor.org/packages/release/bioc/html/Biostrings.html">Biostrings</a> package performing a <a href="http://www.sciencedirect.com/science/article/pii/0022283670900574">Needleman-Wunsch algorithm</a>.</p>
</div>
<div id="codon-alignment-tools" class="section level3">
<h3>Codon Alignment Tools</h3>
<p>The codon alignment tool <a href="http://www.bork.embl.de/pal2nal/">Pal2Nal</a> is already integrated in the <code>orthologr</code> package and doesn’t need to be installed.</p>
<ul>
<li><a href="http://www.bork.embl.de/pal2nal/"><strong>Pal2Nal</strong></a></li>
</ul>
<p>You don’t need to worry about downloading and installing <strong>PAL2NAL</strong>, it is already included in the <code>orthologr</code> package. The corresponding function <code>codon_aln()</code> takes a protein alignment and the corresponding coding sequences and returns a codon alignment by calling <a href="http://www.bork.embl.de/pal2nal/"><strong>Pal2Nal</strong></a> from inside of the <code>orthologr</code> package.</p>
</div>
<div id="dnds-estimation-methods" class="section level3">
<h3>dNdS Estimation Methods</h3>
<p>dNdS estimation is a method to quantify the selection pressure acting on a specific protein sequence determined by pairwise comparisons of amino acid substitutions between two protein sequences and their corresponding codon alignments. Different models have been proposed to estimate this ratio quantifying selection pressure on proteins. The <code>orthologr</code> package includes the most common dNdS estimation methods.</p>
<p>Starting with an codon alignment returned by <code>codon_aln()</code> the function <code>dNdS()</code> computes the the dN, dS, and dNdS values of pairs of proteins.</p>
<p>Based on implementations provided by <a href="http://www.molpopgen.org/software/libsequence.html">gestimator</a>, <a href="https://www.cran.r-project.org/web/packages/ape/index.html">ape</a>, and <a href="https://code.google.com/p/kaks-calculator/">KaKs_Calculator</a>, the following dNdS Estimation Methods are available in <code>orthologr</code>:</p>
<ul>
<li><p><a href="http://link.springer.com/article/10.1007/BF02407308#page-1">Li</a> : Li’s method (1993) -&gt; provided by the <a href="https://cran.r-project.org/web/packages/ape/index.html">ape package</a></p></li>
<li><p><a href="http://link.springer.com/article/10.1007/BF00173196">Comeron</a> : Comeron’s method (1995)</p></li>
<li><p><a href="http://mbe.oxfordjournals.org/content/3/5/418.short">NG</a> : Nei, M. and Gojobori, T. (1986)</p></li>
<li><p><a href="http://mbe.oxfordjournals.org/content/2/2/150.short">LWL</a> : Li, W.H., et al. (1985)</p></li>
<li><p><a href="http://mbe.oxfordjournals.org/content/21/12/2290.short">MLWL</a> (Modified LWL), MLPB (Modified LPB): Tzeng, Y.H., et al. (2004)</p></li>
<li><p><a href="http://mbe.oxfordjournals.org/content/17/1/32.short">YN</a> : Yang, Z. and Nielsen, R. (2000)</p></li>
<li><p><a href="http://www.biomedcentral.com/1471-2148/6/44/">MYN</a> (Modified YN): Zhang, Z., et al. (2006)</p></li>
</ul>
<p>For this purpose you need to have <strong>KaKs_Calculator</strong> installed on your system and executable from your default <code>PATH</code>, e,g, <code>/usr/local/bin/</code>.</p>
</div>
<div id="cds-prediction-pipeline" class="section level3">
<h3>CDS prediction pipeline</h3>
<p><strong>The <code>predict_cds()</code> function is not finished yet.</strong></p>
<p>It is also possible to run <strong>Divergence Stratigraphy</strong> or any other orthology inference method starting with non-annotated genomes.</p>
<p>For this purpose <code>orthologr</code> provides a function <code>predict_cds()</code> that allows you to perform gene prediction (CDS prediction) based on the gene prediction programs <a href="http://bioinf.uni-greifswald.de/augustus/">Augustus</a> and <a href="http://exon.gatech.edu/GeneMark/">GeneMark</a> and also allows you to translate the predicted genes to amino acids: genes -&gt; cds -&gt; amino acids that can subsequently be used for <strong>Divergence Stratigraphy</strong> or <strong>Orthology Inference</strong>.</p>
<ul>
<li><p><a href="http://bioinf.uni-greifswald.de/augustus/">Augustus</a> : Prediction of regions within eukaryotic genomic sequences that can be denoted as “predicted genes”</p></li>
<li><p><code>augustus()</code>: Interface function for the gene prediction program augustus</p></li>
<li><p><a href="http://exon.gatech.edu/GeneMark/">GeneMark</a>: A family of gene prediction programs for all kingdoms of life</p></li>
<li><p><code>genemark()</code> : Interface function for the gene prediction program GeneMark</p></li>
</ul>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
