<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="date" content="2016-03-04" />

<title>Orthology Inference using orthologr</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>


<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>



<div class="fluid-row" id="header">


<h1 class="title">Orthology Inference using orthologr</h1>
<h4 class="date"><em>2016-03-04</em></h4>

</div>


<p>Orthology Inference is the process of detecting <a href="https://www.biostars.org/p/1595/">orthologous genes</a> between a query organism and a set of subject organisms. Motivated by the discussion on which orthology inference method, paradigm or program is the <a href="https://www.biostars.org/p/7568/">most accurate</a> one, the <code>orthologr</code> package provides several interface functions to common orthology inference tools such as <a href="http://www.orthomcl.org/orthomcl/">OrthoMCL</a>, <a href="https://www.bioinf.uni-leipzig.de/Software/proteinortho/">ProteinOrtho</a>, and <a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download">BLAST</a> <em>best reciprocal hit</em>, <a href="http://omabrowser.org/Algorithm.html">OMA</a>, <a href="http://fasta.bioch.virginia.edu/fasta_www2/fasta_list2.shtml">FASTA</a>, and <a href="http://hmmer.janelia.org/">HMM based methods</a>. Future versions of the <code>orthologr</code> package will also implement interfaces to non-BLAST based methods such as <a href="http://orthology.phylomedb.org/">MetaPhOrs</a>, <a href="http://www.ensembl.org/info/docs/api/compara/index.html">EnsemblCompara</a>, <a href="http://phylomedb.org/">PhylomeDB</a>, etc and will allow to benchmark the goodness of the corresponding orthology inference run via <a href="http://orthology.benchmarkservice.org/cgi-bin/gateway.pl">Orthology Benchmarking</a>.</p>
<p>To perform orthology inference you can start with the <code>orthologs()</code> function provided by <code>orthologr</code>. The <code>orthologs()</code> function takes nucleotide or protein sequences stored in fasta files for a set of organisms and performs orthology inference to detect orthologous genes within the given organisms based on selected orthology inference programs.</p>
<p>The following interfaces are (yet) implemented in the <code>orthologs()</code> function:</p>
<div id="blast-based-methods" class="section level3">
<h3>BLAST based methods:</h3>
<ul>
<li><p>BLASTp best hit (BH)</p></li>
<li><p>BLASTp reciprocal best hit (RBH)</p></li>
<li><p>DELTA-BLAST reciprocal best hit (DELTA)</p></li>
<li><p>ProteinOrtho</p></li>
</ul>
<p>Using a simple example stored in the package environment of <code>orthologr</code> you can get an impression on how to use the <code>orthologs()</code> function.</p>
<p><strong>Note:</strong> it is assumed that when using <code>orthologs()</code> all corresponding programs you want to use are already installed on your machine and are executable via either the default execution PATH or you specifically define the location of the executable file via the <code>path</code> argument that can be passed to <code>orthologs()</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># perform orthology inference using BLAST reciprocal best hit</span>
<span class="co"># and fasta sequence files storing protein sequences</span>
<span class="kw">orthologs</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">seq_type        =</span> <span class="st">&quot;protein&quot;</span>, 
          <span class="dt">ortho_detection =</span> <span class="st">&quot;RBH&quot;</span>, 
          <span class="dt">clean_folders   =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>      query_id            subject_id evalue
1  AT1G01010.1 333554|PACid:16033839  0e+00
2  AT1G01020.1 470181|PACid:16064328 7e-166
3  AT1G01030.1 470180|PACid:16054974  0e+00
4  AT1G01040.1 333551|PACid:16057793  0e+00
5  AT1G01050.1 909874|PACid:16064489 2e-160
6  AT1G01060.3 470177|PACid:16043374  0e+00
7  AT1G01070.1 918864|PACid:16052578  0e+00
8  AT1G01080.1 909871|PACid:16053217 1e-178
9  AT1G01090.1 470171|PACid:16052860  0e+00
10 AT1G01110.2 333544|PACid:16034284  0e+00
11 AT1G01120.1 918858|PACid:16049140  0e+00
12 AT1G01140.3 470161|PACid:16036015  0e+00
13 AT1G01150.1 918855|PACid:16037307 3e-150
14 AT1G01160.1 918854|PACid:16044153  1e-93
15 AT1G01170.2 311317|PACid:16052302  3e-54
16 AT1G01180.1 909860|PACid:16056125  0e+00
17 AT1G01190.1 311315|PACid:16059488  0e+00
18 AT1G01200.1 470156|PACid:16041002 3e-172
19 AT1G01210.1 311313|PACid:16057125  7e-76
20 AT1G01220.1 470155|PACid:16047984  0e+00
</code></pre>
<p>This small example returns 20 orthologous genes between <em>Arabidopsis thaliana</em> and <em>Arabidopsis lyrata</em>. As you can see, the <code>query_file</code> and <code>subject_files</code> arguments take the proteomes of <em>Arabidopsis thaliana</em> (<code>query_file</code>) and <em>Arabidopsis lyrata</em> (<code>subject_files</code>) stored in fasta files. The <code>seq_type</code> argument specifies that you will pass protein sequences (proteomes) to the <code>orthologs()</code> function. In case you only have either genomes (DNA sequences) or CDS files, you can modify the <code>seq_type</code> argument to <code>seq_type = &quot;dna&quot;</code> (when working with only genome data) or <code>seq_type = &quot;cds&quot;</code> (when working with CDS files). Internally the <code>orthologs()</code> function will perform a CDS prediction using <code>predict_cds()</code> and will furthermore translate the predicted CDS sequences into protein sequences. Analogously when <code>seq_type = &quot;cds&quot;</code> is specified, internally the <code>orthologs()</code> function will translate all CDS sequences into protein sequences to run orthology inference based on protein sequences.</p>
<p><strong>Note</strong>: future versions of <code>orthologr</code> will allow to perform orthology inference using DNA sequences. Nevertheless, since most orthology inference methods or paradigms rely on protein sequences, the first version of <code>orthologr</code> will follow this paradigm.</p>
<p>In case you have to specify the path to the corresponding orthology inference method you can use the <code>path</code> argument as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># using an external execution path</span>
<span class="kw">orthologs</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">seq_type        =</span> <span class="st">&quot;protein&quot;</span>, 
          <span class="dt">ortho_detection =</span> <span class="st">&quot;RBH&quot;</span>, 
          <span class="dt">path            =</span> <span class="st">&quot;here/path/to/blastp&quot;</span>, 
          <span class="dt">clean_folders   =</span> <span class="ot">FALSE</span>,
          <span class="dt">comp_cores      =</span> <span class="dv">1</span>)</code></pre></div>
<p>When you are working on a multicore machine, you can also specify the <code>comp_cores</code> argument that will allow you to run all analyses in parallel (to speed up computations).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># running orthology inference in parallel using 2 cores</span>
<span class="kw">orthologs</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">seq_type        =</span> <span class="st">&quot;protein&quot;</span>, 
          <span class="dt">ortho_detection =</span> <span class="st">&quot;RBH&quot;</span>, 
          <span class="dt">clean_folders   =</span> <span class="ot">FALSE</span>,
          <span class="dt">comp_cores      =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>
      query_id            subject_id evalue
1  AT1G01010.1 333554|PACid:16033839  0e+00
2  AT1G01020.1 470181|PACid:16064328 7e-166
3  AT1G01030.1 470180|PACid:16054974  0e+00
4  AT1G01040.1 333551|PACid:16057793  0e+00
5  AT1G01050.1 909874|PACid:16064489 2e-160
6  AT1G01060.3 470177|PACid:16043374  0e+00
7  AT1G01070.1 918864|PACid:16052578  0e+00
8  AT1G01080.1 909871|PACid:16053217 1e-178
9  AT1G01090.1 470171|PACid:16052860  0e+00
10 AT1G01110.2 333544|PACid:16034284  0e+00
11 AT1G01120.1 918858|PACid:16049140  0e+00
12 AT1G01140.3 470161|PACid:16036015  0e+00
13 AT1G01150.1 918855|PACid:16037307 3e-150
14 AT1G01160.1 918854|PACid:16044153  1e-93
15 AT1G01170.2 311317|PACid:16052302  3e-54
16 AT1G01180.1 909860|PACid:16056125  0e+00
17 AT1G01190.1 311315|PACid:16059488  0e+00
18 AT1G01200.1 470156|PACid:16041002 3e-172
19 AT1G01210.1 311313|PACid:16057125  7e-76
20 AT1G01220.1 470155|PACid:16047984  0e+00
</code></pre>
<p>In this case 2 cores are being used to perform parallel processing, the <code>clean_folders</code> argument specifies that all files returned by the corresponding orthology inference method are removed after analyses.</p>
</div>
<div id="program-specific-use-of-the-orthologs-function" class="section level2">
<h2>Program specific use of the orthologs() function</h2>
<p>In this section small examples will illustrate the use of the <code>orthologs()</code> function for each orthology inference program.</p>
<div id="blastp-best-hit" class="section level3">
<h3>BLASTp best hit</h3>
<p>The BLASTp best hit method is a uni-directional BLAST best hit search of a <code>query organism A</code> against a <code>subject organism B</code> based on the <code>e-value</code>.</p>
<p>Orthology Inference using BLASTp best hit can be performed by specifying the argument <code>ortho_detection = &quot;BH&quot;</code> and one computing core <code>comp_core = 1</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># perform orthology inference using BLAST best hit</span>
<span class="co"># and fasta sequence files storing protein sequences</span>
<span class="kw">orthologs</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">seq_type        =</span> <span class="st">&quot;protein&quot;</span>, 
          <span class="dt">ortho_detection =</span> <span class="st">&quot;BH&quot;</span>, 
          <span class="dt">clean_folders   =</span> <span class="ot">TRUE</span>,
          <span class="dt">detailed_output =</span> <span class="ot">FALSE</span>,
          <span class="dt">comp_cores      =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>       query_id            subject_id evalue
 1: AT1G01010.1 333554|PACid:16033839  0e+00
 2: AT1G01020.1 470181|PACid:16064328 7e-166
 3: AT1G01030.1 470180|PACid:16054974  0e+00
 4: AT1G01040.1 333551|PACid:16057793  0e+00
 5: AT1G01050.1 909874|PACid:16064489 2e-160
 6: AT1G01060.3 470177|PACid:16043374  0e+00
 7: AT1G01070.1 918864|PACid:16052578  0e+00
 8: AT1G01080.1 909871|PACid:16053217 1e-178
 9: AT1G01090.1 470171|PACid:16052860  0e+00
10: AT1G01110.2 333544|PACid:16034284  0e+00
11: AT1G01120.1 918858|PACid:16049140  0e+00
12: AT1G01140.3 470161|PACid:16036015  0e+00
13: AT1G01150.1 918855|PACid:16037307 3e-150
14: AT1G01160.1 918854|PACid:16044153  1e-93
15: AT1G01170.2 311317|PACid:16052302  3e-54
16: AT1G01180.1 909860|PACid:16056125  0e+00
17: AT1G01190.1 311315|PACid:16059488  0e+00
18: AT1G01200.1 470156|PACid:16041002 3e-172
19: AT1G01210.1 311313|PACid:16057125  7e-76
20: AT1G01220.1 470155|PACid:16047984  0e+00</code></pre>
<p>The resulting table stores the orthologous gene pairs and the corresponding e-value of the best hit. By specifying <code>detailed_output = TRUE</code> more information of the BLAST result can be obtained.</p>
</div>
<div id="blastp-best-reciprocal-hit" class="section level3">
<h3>BLASTp best reciprocal hit</h3>
<p>The BLAST best reciprocal hit is a bi-directional BLAST best hit search of a <code>query organism A</code> against a <code>subject organism B</code> based on the <code>e-value</code>.</p>
<p>The Algorithm for BLAST best reciprocal hit runs as follows:</p>
<ol style="list-style-type: decimal">
<li><p><code>bh_A &lt;- best_hit(A,B)</code></p></li>
<li><p><code>bh_B &lt;- best_hit(B,A)</code></p></li>
<li><p><code>join(bh_A,bh_B)</code> by tupel <code>(query_id, subject_id)</code></p></li>
</ol>
<p>In other words, only in case the tuple <code>(query_id, subject_id)</code> is returned as best hit based on the <code>e-value</code> in both BLAST directions, the corresponding tupel <code>(query_id, subject_id)</code> is retained as orthologous gene pair.</p>
<p>As demonstrated before a simple call of <code>orthologs()</code> using <code>ortho_detection = &quot;RBH&quot;</code> and one computing core <code>comp_core = 1</code> can be performed as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># perform orthology inference using BLAST reciprocal best hit</span>
<span class="co"># and fasta sequence files storing protein sequences</span>
<span class="kw">orthologs</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">seq_type        =</span> <span class="st">&quot;protein&quot;</span>, 
          <span class="dt">ortho_detection =</span> <span class="st">&quot;RBH&quot;</span>, 
          <span class="dt">clean_folders   =</span> <span class="ot">FALSE</span>,
          <span class="dt">comp_cores      =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>
      query_id            subject_id evalue
1  AT1G01010.1 333554|PACid:16033839  0e+00
2  AT1G01020.1 470181|PACid:16064328 7e-166
3  AT1G01030.1 470180|PACid:16054974  0e+00
4  AT1G01040.1 333551|PACid:16057793  0e+00
5  AT1G01050.1 909874|PACid:16064489 2e-160
6  AT1G01060.3 470177|PACid:16043374  0e+00
7  AT1G01070.1 918864|PACid:16052578  0e+00
8  AT1G01080.1 909871|PACid:16053217 1e-178
9  AT1G01090.1 470171|PACid:16052860  0e+00
10 AT1G01110.2 333544|PACid:16034284  0e+00
11 AT1G01120.1 918858|PACid:16049140  0e+00
12 AT1G01140.3 470161|PACid:16036015  0e+00
13 AT1G01150.1 918855|PACid:16037307 3e-150
14 AT1G01160.1 918854|PACid:16044153  1e-93
15 AT1G01170.2 311317|PACid:16052302  3e-54
16 AT1G01180.1 909860|PACid:16056125  0e+00
17 AT1G01190.1 311315|PACid:16059488  0e+00
18 AT1G01200.1 470156|PACid:16041002 3e-172
19 AT1G01210.1 311313|PACid:16057125  7e-76
20 AT1G01220.1 470155|PACid:16047984  0e+00
</code></pre>
<p>When you need more parameters returned by <code>RBH</code>, you can specify the <code>detailed_output = TRUE</code> argument to obtain all BLAST parameters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)
<span class="kw">library</span>(dplyr)

<span class="co"># perform orthology inference using BLAST reciprocal best hit</span>
<span class="co"># and fasta sequence files storing protein sequences</span>
RBH &lt;-<span class="st"> </span><span class="kw">orthologs</span>(<span class="dt">query_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files     =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">seq_type          =</span> <span class="st">&quot;protein&quot;</span>, 
          <span class="dt">ortho_detection   =</span> <span class="st">&quot;RBH&quot;</span>,
          <span class="dt">detailed_output   =</span> <span class="ot">TRUE</span>, 
          <span class="dt">clean_folders     =</span> <span class="ot">FALSE</span>,
          <span class="dt">comp_cores        =</span> <span class="dv">1</span>)


<span class="kw">glimpse</span>(RBH)</code></pre></div>
<pre><code>Variables:
$ query_id      (chr) &quot;AT1G01010.1&quot;, &quot;AT1G01020.1&quot;, &quot;AT1G01030.1&quot;, &quot;AT1G01040.1&quot;...
$ subject_id    (chr) &quot;333554|PACid:16033839&quot;, &quot;470181|PACid:16064328&quot;, &quot;470180|...
$ perc_identity (dbl) 73.99, 91.06, 95.54, 91.98, 100.00, 89.51, 95.08, 90.33, 9...
$ alig_length   (dbl) 469, 246, 359, 1970, 213, 648, 366, 300, 434, 528, 529, 45...
$ mismatches    (dbl) 80, 22, 12, 85, 0, 58, 14, 22, 8, 34, 4, 6, 68, 19, 0, 20,...
$ gap_openings  (dbl) 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, 1, 3, 2, 1, 1, 1, 0, 0, 0
$ q_start       (dbl) 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2, 1, 1, 1, 1, 1
$ q_end         (dbl) 430, 246, 359, 1910, 213, 646, 366, 294, 429, 528, 529, 45...
$ s_start       (dbl) 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4, 1, 6, 1, 1, 1
$ s_end         (dbl) 466, 246, 355, 1963, 213, 640, 362, 299, 433, 528, 529, 45...
$ evalue        (dbl) 0e+00, 7e-166, 0e+00, 0e+00, 2e-160, 0e+00, 0e+00, 1e-178,...
$ bit_score     (dbl) 627, 454, 698, 3704, 437, 1037, 696, 491, 859, 972, 1092, ...</code></pre>
<p>In case you would like to store the corresponding <code>hit tables</code> returned by BLAST for subsequent analyses, you can specify the <code>clean_folders = FALSE</code> argument. The corresponding BLAST hit table can then be found in <code>file.path(tempdir(),&quot;_blast_db&quot;)</code>.</p>
<p>A detailed overview of further analyses that can be done with the corresponding BLAST output can be found in the <a href="https://github.com/HajkD/orthologr/blob/master/vignettes/blast.Rmd">BLAST vignette</a>.</p>
</div>
<div id="delta-blast-best-reciprocal-hit" class="section level3">
<h3>DELTA-BLAST best reciprocal hit</h3>
<p><a href="http://www.biomedcentral.com/content/pdf/1745-6150-7-12.pdf">Domain enhanced lookup time accelerated BLAST (DELTA-BLAST)</a> is a new BLAST algorithm provided by NCBI. It was introduced as a useful program for the detection of remote protein homologs.</p>
<p>You can download the Conserved Domain Database (CDD) file <code>cdd.tar.gz</code> from <a href="ftp://ftp.ncbi.nlm.nih.gov/pub/mmdb/cdd">NCBI</a> and store all <code>cdd_deltablast.* files</code> in a folder: <code>path/to/cdd_database/folder</code>.</p>
<p>More information on how to install and use DELTA-BLAST can be found <a href="http://www.ncbi.nlm.nih.gov/books/NBK1763/#CmdLineAppsManual.Performing_a_DELTABLAS">here</a>. Make sure that when using DELTA-BLAST in <code>orthologs()</code> the corresponding path to the <code>cdd_deltablast.* files</code> need to be specified in the <code>cdd.delta</code> argument.</p>
<p>The DELTA-BLAST best reciprocal hit is a bi-directional DELTA-BLAST best hit search of a <code>query organism A</code> against a <code>subject organism B</code> based on the <code>e-value</code>.</p>
<p>The Algorithm for DELTA-BLAST best reciprocal hit runs as follows:</p>
<ol style="list-style-type: decimal">
<li><p><code>bh_A &lt;- delta.blast(A,B)</code></p></li>
<li><p><code>bh_B &lt;- delta.blast(B,A)</code></p></li>
<li><p><code>join(bh_A,bh_B)</code> by tupel <code>(query_id, subject_id)</code></p></li>
</ol>
<p>In other words, only in case the tuple <code>(query_id, subject_id)</code> is returned as best hit based on the <code>e-value</code> in both DELTA-BLAST directions, the corresponding tupel <code>(query_id, subject_id)</code> is retained as orthologous gene pair.</p>
<p>As demonstrated before a simple call of <code>orthologs()</code> using <code>ortho_detection = &quot;DELTA&quot;</code>, the <code>cdd.path = &quot;path/to/cdd_database/folder&quot;</code>, and one computing core <code>comp_core = 1</code> can be performed as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># perform orthology inference using DELTA-BLAST reciprocal best hit</span>
<span class="co"># and fasta sequence files storing protein sequences</span>
<span class="kw">orthologs</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">seq_type        =</span> <span class="st">&quot;protein&quot;</span>, 
          <span class="dt">ortho_detection =</span> <span class="st">&quot;DELTA&quot;</span>,
          <span class="dt">cdd.path        =</span> <span class="st">&quot;path/to/cdd_database/folder&quot;</span>,
          <span class="dt">clean_folders   =</span> <span class="ot">FALSE</span>,
          <span class="dt">detailed_output =</span> <span class="ot">FALSE</span>,
          <span class="dt">comp_cores      =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>      query_id            subject_id evalue
1  AT1G01010.1 333554|PACid:16033839  0e+00
2  AT1G01020.1 470181|PACid:16064328  3e-88
3  AT1G01030.1 470180|PACid:16054974  0e+00
4  AT1G01040.1 333551|PACid:16057793  0e+00
5  AT1G01050.1 909874|PACid:16064489 2e-128
6  AT1G01060.3 470177|PACid:16043374  0e+00
7  AT1G01070.1 918864|PACid:16052578  2e-68
8  AT1G01080.1 909871|PACid:16053217  4e-77
9  AT1G01090.1 470171|PACid:16052860  0e+00
10 AT1G01110.2 333544|PACid:16034284  0e+00
11 AT1G01120.1 918858|PACid:16049140  0e+00
12 AT1G01140.3 470161|PACid:16036015 9e-179
13 AT1G01150.1 918855|PACid:16037307 3e-120
14 AT1G01160.1 918854|PACid:16044153  4e-45
15 AT1G01170.2 311317|PACid:16052302  5e-54
16 AT1G01180.1 909860|PACid:16056125 8e-156
17 AT1G01190.1 311315|PACid:16059488  0e+00
18 AT1G01200.1 470156|PACid:16041002 2e-114
19 AT1G01210.1 311313|PACid:16057125  2e-40
20 AT1G01220.1 470155|PACid:16047984  0e+00</code></pre>
<p>When you need more parameters returned by <code>DELTA-BLAST</code>, you can specify the <code>detailed_output = TRUE</code> argument to obtain all DELTA-BLAST parameters.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)
<span class="kw">library</span>(dplyr)

<span class="co"># perform orthology inference using DELTA-BLAST reciprocal best hit</span>
<span class="co"># and fasta sequence files storing protein sequences</span>
DELTA &lt;-<span class="st"> </span><span class="kw">orthologs</span>(<span class="dt">query_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files     =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">seq_type          =</span> <span class="st">&quot;protein&quot;</span>, 
          <span class="dt">ortho_detection   =</span> <span class="st">&quot;DELTA&quot;</span>,
          <span class="dt">cdd.path          =</span> <span class="st">&quot;path/to/cdd_database/folder&quot;</span>,
          <span class="dt">detailed_output   =</span> <span class="ot">TRUE</span>, 
          <span class="dt">clean_folders     =</span> <span class="ot">FALSE</span>,
          <span class="dt">comp_cores        =</span> <span class="dv">1</span>)


<span class="kw">glimpse</span>(DELTA)</code></pre></div>
<pre><code>Variables:
$ query_id                (chr) &quot;AT1G01010.1&quot;, &quot;AT1G01020.1&quot;, &quot;AT1G01030.1&quot;, &quot;AT1G01040.1&quot;, &quot;...
$ subject_id              (chr) &quot;333554|PACid:16033839&quot;, &quot;470181|PACid:16064328&quot;, &quot;470180|PAC...
$ perc_identity           (dbl) 73.77, 91.06, 95.26, 91.57, 100.00, 89.51, 93.44, 78.60, 96.7...
$ num_ident_matches       (dbl) 346, 224, 342, 1803, 213, 580, 342, 235, 420, 494, 525, 446, ...
$ alig_length             (dbl) 469, 246, 359, 1969, 213, 648, 366, 299, 434, 528, 529, 453, ...
$ mismatches              (dbl) 81, 22, 13, 95, 0, 58, 20, 59, 8, 34, 4, 6, 71, 32, 0, 20, 31...
$ gap_openings            (dbl) 9, 0, 2, 8, 0, 5, 2, 1, 3, 0, 0, 1, 3, 1, 1, 1, 1, 0, 0, 0
$ q_start                 (dbl) 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2, 1, 1, 1, 1, 1
$ q_end                   (dbl) 430, 246, 359, 1910, 213, 646, 366, 294, 429, 528, 529, 452, ...
$ s_start                 (dbl) 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4, 1, 6, 1, 1, 1
$ s_end                   (dbl) 466, 246, 355, 1963, 213, 640, 362, 299, 433, 528, 529, 453, ...
$ evalue                  (dbl) 0e+00, 3e-88, 0e+00, 0e+00, 2e-128, 0e+00, 2e-68, 4e-77, 0e+0...
$ bit_score               (dbl) 637, 257, 502, 2362, 357, 1003, 214, 232, 658, 763, 747, 504,...
$ score_raw               (dbl) 1643, 657, 1294, 6123, 917, 2592, 544, 592, 1698, 1971, 1928,...
$ query_coverage_per_subj (dbl) 100, 100, 100, 99, 100, 100, 100, 100, 100, 100, 100, 100, 83...</code></pre>
<p>In case you would like to store the corresponding <code>hit tables</code> returned by DELTA-BLAST for subsequent analyses, you can specify the <code>clean_folders = FALSE</code> argument. When <code>clean_folders</code> is set to <code>FALSE</code>, a folder named <code>_blast_db</code> will be created inside the <code>tempdir()</code> folder of your R session storing the corresponding DELTA-BLAST output. The corresponding DELTA-BLAST hit table can then be found in <code>file.path(tempdir(),&quot;_blast_db&quot;)</code>. A detailed overview of further analyses that can be done with the corresponding DELTA-BLAST output can be found in the <a href="https://github.com/HajkD/orthologr/blob/master/vignettes/blast.Rmd">BLAST vignette</a>.</p>
</div>
<div id="proteinortho" class="section level3">
<h3>ProteinOrtho</h3>
<p>ProteinOrtho is a orthology inference program to detect orthologous genes within different species. For this purpose it compares similarities of given gene sequences and clusters them to find significant groups. To enhance the prediction accuracy, the relative order of genes (synteny) can be used as additional feature for the discrimination of orthologs (<a href="https://www.bioinf.uni-leipzig.de/Software/proteinortho/manual.html">Lechner et al., 2011</a>).</p>
<div id="installing-proteinortho-for-use-in-orthologr" class="section level4">
<h4>Installing ProteinOrtho for use in orthologr</h4>
<p>Please follow the <a href="https://www.bioinf.uni-leipzig.de/Software/proteinortho/manual.html">installation instructions</a> from the ProteinOrtho web page.</p>
<p>When you have ProteinOrtho installed on your machine, make sure that you can execute the file <code>proteinortho5.pl</code> from your default execution <code>PATH</code>, e.g. <code>/usr/local/bin/</code> (MacOS, Linux).</p>
<p>You can test whether ProteinOrtho can be executed properly by typing the following command in your R session:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">system</span>(<span class="st">&quot;proteinortho5.pl -h&quot;</span>)</code></pre></div>
<p>This way you should see the following output:</p>
<pre><code>Usage: proteinortho5.pl [OPTIONS] FASTA1 FASTA2 [FASTA...]
Options: -e=          E-value for blast [default: 1e-05]
         -p=          blast program {blastn|blastp|blastn+|blastp+}
                      [default: blastp+]
         -project=    prefix for all result file names [default: myproject]
         -synteny     activate PoFF extension to separate similar sequences
                      by contextual adjacencies (requires .gff for each .fasta)
         -dups=       PoFF: number of reiterations for adjacencies heuristic,
                      to determine duplicated regions (default: 0)
         -cs=         PoFF: Size of a maximum common substring (MCS) for
                      adjacency matches (default: 3)
         -alpha=      PoFF: weight of adjacencies vs. sequence similarity
                      (default: 0.5)
         -keep        stores temporary blast results for reuse
         -force       forces recalculation of blast results in any case
         -cpus=       number of processors to use [default: auto]
         -selfblast   apply selfblast, detects paralogs without orthologs
         -singles     report singleton genes without any hit
         -identity=   min. percent identity of best blast hits [default: 25]
         -cov=        min. coverage of best blast alignments in % [default: 50]
         -conn=       min. algebraic connectivity [default: 0.1]
         -sim=        min. similarity for additional hits (0..1) [default: 0.95]
         -step=       1 -&gt; generate indices
                      2 -&gt; run blast (and ff-adj, if -synteny is set)
                      3 -&gt; clustering
                      0 -&gt; all (default)
         -blastpath=  path to your local blast (if not installed globally)
         -verbose     keeps you informed about the progress
         -clean       remove all unnecessary files after processing
         -graph       generate .graph files (pairwise orthology relations)
         -debug       gives detailed information for bug tracking

         More specific blast parameters can be defined by
         -blastParameters='[parameters]' (e.g. -blastParameters='-seg no')

         In case jobs should be distributed onto several machines, use
         -startat=    File number to start with (default: 0)
         -stopat=     File number to end with (default: -1)
 
Invalid command line option: '-h'!
</code></pre>
</div>
<div id="in-case-you-have-problems-with-proteinortho" class="section level4">
<h4>In case you have problems with ProteinOrtho</h4>
<p>If you cannot reproduce the previous result, then <code>proteinortho5.pl</code> cannot be executed properly from your default execution <code>PATH</code> (e.g. <code>/usr/local/bin/</code>).</p>
<p>In this case first you can try to store all files within the <code>proteinortho_v5.07</code> folder in <code>/usr/local/bin/</code> (MacOS and Linux) or your default <code>Program Files</code> folder path (Windows).</p>
<p>Furthermore, you can also try the following steps:</p>
<ul>
<li>specify the <code>path</code> argument inside the <code>orthologs()</code> function to pass the path to your <code>proteinortho_v5.07</code> folder</li>
<li>read the documentation of the <code>ProteinOrtho()</code> function: <code>?ProteinOrtho</code></li>
<li>formulate a bug report</li>
</ul>
<p><strong>Note:</strong> when using the synteny option of ProteinOrtho, we noticed that the necessary file <code>proteinortho5_clustering.cpp</code> stored in the <code>proteinortho_v5.07</code> folder needs to be stored in <code>/usr/local/bin/</code> and in our case also needed to be recompiled with <code>g++ proteinortho5_clustering.cpp -o proteinortho5_clustering</code> inside the <code>Terminal</code>.</p>
</div>
<div id="running-orthologs-with-proteinortho" class="section level4">
<h4>Running orthologs() with ProteinOrtho</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># perform orthology inference using ProteinOrtho</span>
<span class="co"># and fasta sequence files storing protein sequences</span>
<span class="kw">orthologs</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">seq_type        =</span> <span class="st">&quot;protein&quot;</span>, 
          <span class="dt">ortho_detection =</span> <span class="st">&quot;PO&quot;</span>, 
          <span class="dt">clean_folders   =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>
      ortho_lyra_aa.fasta ortho_thal_aa.fasta evalue_ab bitscore_ab evalue_ba bitscore_ba
 1: 311313|PACid:16057125         AT1G01210.1     7e-76         215     7e-76         215
 2: 311315|PACid:16059488         AT1G01190.1     0e+00        1036     0e+00        1036
 3: 311317|PACid:16052302         AT1G01170.2     4e-54         158     3e-54         158
 4: 333544|PACid:16034284         AT1G01110.2     0e+00         978     0e+00         972
 5: 333551|PACid:16057793         AT1G01040.1     0e+00        3704     0e+00        3704
 6: 333554|PACid:16033839         AT1G01010.1     0e+00         615     0e+00         627
 7: 470155|PACid:16047984         AT1G01220.1     0e+00        2106     0e+00        2106
 8: 470156|PACid:16041002         AT1G01200.1    3e-172         470    3e-172         470
 9: 470161|PACid:16036015         AT1G01140.3     0e+00         918     0e+00         918
10: 470171|PACid:16052860         AT1G01090.1     0e+00         859     0e+00         859
11: 470177|PACid:16043374         AT1G01060.3     0e+00        1033     0e+00        1037
12: 470180|PACid:16054974         AT1G01030.1     0e+00         698     0e+00         698
13: 470181|PACid:16064328         AT1G01020.1    6e-158         434    7e-166         454
14: 909860|PACid:16056125         AT1G01180.1     0e+00         599     0e+00         576
15: 909871|PACid:16053217         AT1G01080.1    5e-169         466    1e-178         491
16: 909874|PACid:16064489         AT1G01050.1    2e-160         437    2e-160         437
17: 918854|PACid:16044153         AT1G01160.1     2e-86         249     1e-93         268
18: 918855|PACid:16037307         AT1G01150.1    2e-150         421    3e-150         421
19: 918858|PACid:16049140         AT1G01120.1     0e+00        1092     0e+00        1092
20: 918864|PACid:16052578         AT1G01070.1     0e+00         696     0e+00         696
</code></pre>
<p>Again the <code>comp_cores</code> argument can be used to run ProteinOrtho in parallel.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># perform orthology inference using ProteinOrtho</span>
<span class="co"># and fasta sequence files storing protein sequences</span>
<span class="kw">orthologs</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">seq_type        =</span> <span class="st">&quot;protein&quot;</span>, 
          <span class="dt">comp_cores      =</span> <span class="dv">2</span>,
          <span class="dt">ortho_detection =</span> <span class="st">&quot;PO&quot;</span>, 
          <span class="dt">clean_folders   =</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>You can run ProteinOrtho with multiple subject organisms.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># defining 3 subject organisms: A. lyrata, B. rapa, and T. halophila</span>
subject_organisms &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">system.file</span>(<span class="st">'seqs/example_brapa_aa.faa'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                       <span class="kw">system.file</span>(<span class="st">'seqs/example_alyra_aa.faa'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                       <span class="kw">system.file</span>(<span class="st">'seqs/example_thalo_aa.faa'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))

<span class="kw">ProteinOrtho</span>(<span class="dt">query_file =</span> <span class="kw">system.file</span>(<span class="st">'seqs/example_athal_aa.faa'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_files =</span> subject_organisms,
          <span class="dt">seq_type      =</span> <span class="st">&quot;protein&quot;</span>)</code></pre></div>
<pre><code>
$comparison_1
       # example_athal_aa.faa  example_alyra_aa.faa evalue_ab bitscore_ab evalue_ba bitscore_ba
1  AT1G01010.1|PACid:19656964 333554|PACid:16033839       0.0         627       0.0         615
2  AT1G01020.1|PACid:19655142 470181|PACid:16064328    7e-166         454    6e-158         434
3  AT1G01030.1|PACid:19649747 470180|PACid:16054974       0.0         698       0.0         698
4  AT1G01040.1|PACid:19652432 333551|PACid:16057793       0.0        3704       0.0        3704
5  AT1G01050.1|PACid:19652974 909874|PACid:16064489    2e-160         437    2e-160         437
6  AT1G01070.1|PACid:19650807 918864|PACid:16052578       0.0         696       0.0         696
7  AT1G01080.1|PACid:19651136 909871|PACid:16053217    1e-178         491    5e-169         466
8  AT1G01090.1|PACid:19650207 470171|PACid:16052860       0.0         859       0.0         859
9  AT1G01110.2|PACid:19658169 333544|PACid:16034284       0.0         972       0.0         978
10 AT1G01120.1|PACid:19655257 918858|PACid:16049140       0.0        1092       0.0        1092
11 AT1G01160.1|PACid:19654307 918854|PACid:16044153     1e-93         268     2e-86         249
12 AT1G01180.1|PACid:19654493 909860|PACid:16056125       0.0         572       0.0         594
13 AT1G01190.1|PACid:19654322 311315|PACid:16059488       0.0        1036       0.0        1036
14 AT1G01200.1|PACid:19649914 470156|PACid:16041002    3e-172         470    3e-172         470
15 AT1G01220.1|PACid:19652890 470155|PACid:16047984       0.0        2106       0.0        2106
16 AT1G01225.1|PACid:19656715 470154|PACid:16066159    6e-168         461    1e-163         450
17 AT1G01230.1|PACid:19650274 470153|PACid:16038331    3e-117         324    3e-117         324
18 AT1G01240.3|PACid:19650438 470152|PACid:16047967       0.0         568       0.0         545
19 AT1G01250.1|PACid:19649318 470149|PACid:16044741    1e-124         345    1e-124         345
20 AT1G01260.3|PACid:19651080 470147|PACid:16058975       0.0        1108       0.0        1117

$comparison_2
     # example_brapa_aa.faa  example_alyra_aa.faa evalue_ab bitscore_ab evalue_ba bitscore_ba
1  Bra032623|PACid:22715924 918854|PACid:16044153     1e-93         268     2e-98         281
2  Bra032626|PACid:22713115 470155|PACid:16047984       0.0        1946       0.0        1946
3  Bra032627|PACid:22713604 470154|PACid:16066159    1e-159         439    1e-160         442
4  Bra032629|PACid:22714275 470152|PACid:16047967    3e-152         427    8e-136         385
5  Bra032630|PACid:22715365 470149|PACid:16044741    2e-106         298    4e-106         298
6  Bra033273|PACid:22692002 470147|PACid:16058975       0.0         846       0.0         858
7  Bra033276|PACid:22693754 470153|PACid:16038331    3e-111         310    3e-111         310
8  Bra033277|PACid:22693004 470156|PACid:16041002    5e-160         439    5e-160         439
9  Bra033278|PACid:22692201 311315|PACid:16059488       0.0         992       0.0         992
10 Bra033279|PACid:22694360 909860|PACid:16056125       0.0         508       0.0         516
11 Bra033283|PACid:22693298 918858|PACid:16049140       0.0        1015       0.0        1015
12 Bra033284|PACid:22692650 333544|PACid:16034284       0.0         842       0.0         807
13 Bra033286|PACid:22694322 470171|PACid:16052860       0.0         818       0.0         818
14 Bra033287|PACid:22693868 909871|PACid:16053217    3e-128         358    2e-128         360
15 Bra033290|PACid:22692027 918864|PACid:16052578       0.0         621       0.0         621
16 Bra033292|PACid:22694379 909874|PACid:16064489    4e-158         432    4e-158         432
17 Bra033293|PACid:22692618 333551|PACid:16057793       0.0        3257       0.0        3334
18 Bra033294|PACid:22691990 470180|PACid:16054974       0.0         528       0.0         574
19 Bra033295|PACid:22694612 470181|PACid:16064328    4e-147         407    1e-139         387
20 Bra033296|PACid:22693237 333554|PACid:16033839    7e-137         398    1e-131         385

$comparison_3
           # example_thalo_aa.faa  example_alyra_aa.faa evalue_ab bitscore_ab evalue_ba bitscore_ba
1  Thhalv10006531m|PACid:20187082 333551|PACid:16057793       0.0        3484       0.0        3514
2  Thhalv10006637m|PACid:20187150 470155|PACid:16047984       0.0        2022       0.0        2022
3  Thhalv10007139m|PACid:20185859 470147|PACid:16058975       0.0        1032       0.0         992
4  Thhalv10007299m|PACid:20188214 333544|PACid:16034284       0.0         872       0.0         844
5  Thhalv10007320m|PACid:20187751 311315|PACid:16059488       0.0         959       0.0         959
6  Thhalv10007351m|PACid:20187717 918858|PACid:16049140       0.0        1055       0.0        1055
7  Thhalv10007665m|PACid:20186865 333554|PACid:16033839       0.0         538       0.0         532
8  Thhalv10007695m|PACid:20187891 470171|PACid:16052860       0.0         844       0.0         844
9  Thhalv10008068m|PACid:20187490 470180|PACid:16054974       0.0         566       0.0         596
10 Thhalv10008123m|PACid:20187436 470152|PACid:16047967       0.0         530       0.0         523
11 Thhalv10008287m|PACid:20186971 909860|PACid:16056125       0.0         546       0.0         557
12 Thhalv10008355m|PACid:20185661 909871|PACid:16053217    2e-163         452    5e-170         469
13 Thhalv10008482m|PACid:20185881 918854|PACid:16044153    4e-100         288    2e-100         288
14 Thhalv10008506m|PACid:20187047 470154|PACid:16066159    1e-160         442    4e-162         446
15 Thhalv10008618m|PACid:20186467 470181|PACid:16064328    1e-153         423    6e-146         404
16 Thhalv10008767m|PACid:20187619 909874|PACid:16064489    1e-158         433    1e-158         433
17 Thhalv10008983m|PACid:20187021 470153|PACid:16038331    6e-116         320    6e-116         320
18 Thhalv10009311m|PACid:20187590 470156|PACid:16041002    2e-163         447    2e-163         447
19 Thhalv10009402m|PACid:20186843 470149|PACid:16044741    2e-119         332    2e-119         332
20 Thhalv10009501m|PACid:20185805 918864|PACid:16052578       0.0         639       0.0         630

$comparison_4
     # example_brapa_aa.faa       example_athal_aa.faa evalue_ab bitscore_ab evalue_ba bitscore_ba
1  Bra032623|PACid:22715924 AT1G01160.1|PACid:19654307     3e-88         254     7e-97         275
2  Bra032626|PACid:22713115 AT1G01220.1|PACid:19652890       0.0        1953       0.0        1953
3  Bra032627|PACid:22713604 AT1G01225.1|PACid:19656715    1e-157         435    3e-162         446
4  Bra032629|PACid:22714275 AT1G01240.3|PACid:19650438    3e-155         435    4e-142         401
5  Bra032630|PACid:22715365 AT1G01250.1|PACid:19649318    5e-103         290    6e-103         290
6  Bra033273|PACid:22692002 AT1G01260.3|PACid:19651080       0.0         870       0.0         856
7  Bra033276|PACid:22693754 AT1G01230.1|PACid:19650274    1e-110         308    8e-111         308
8  Bra033277|PACid:22693004 AT1G01200.1|PACid:19649914    8e-158         433    8e-158         433
9  Bra033278|PACid:22692201 AT1G01190.1|PACid:19654322       0.0         986       0.0         986
10 Bra033279|PACid:22694360 AT1G01180.1|PACid:19654493       0.0         509       0.0         509
11 Bra033283|PACid:22693298 AT1G01120.1|PACid:19655257       0.0        1014       0.0        1014
12 Bra033284|PACid:22692650 AT1G01110.2|PACid:19658169       0.0         843       0.0         803
13 Bra033286|PACid:22694322 AT1G01090.1|PACid:19650207       0.0         811       0.0         811
14 Bra033287|PACid:22693868 AT1G01080.1|PACid:19651136    1e-125         352    4e-126         353
15 Bra033290|PACid:22692027 AT1G01070.1|PACid:19650807       0.0         610       0.0         610
16 Bra033292|PACid:22694379 AT1G01050.1|PACid:19652974    4e-158         432    4e-158         432
17 Bra033293|PACid:22692618 AT1G01040.1|PACid:19652432       0.0        3408       0.0        3430
18 Bra033294|PACid:22691990 AT1G01030.1|PACid:19649747       0.0         533       0.0         573
19 Bra033295|PACid:22694612 AT1G01020.1|PACid:19655142    2e-138         385    9e-135         375
20 Bra033296|PACid:22693237 AT1G01010.1|PACid:19656964    1e-139         404    6e-135         392

$comparison_5
           # example_thalo_aa.faa       example_athal_aa.faa evalue_ab bitscore_ab evalue_ba bitscore_ba
1  Thhalv10006531m|PACid:20187082 AT1G01040.1|PACid:19652432       0.0        3628       0.0        3603
2  Thhalv10006637m|PACid:20187150 AT1G01220.1|PACid:19652890       0.0        2026       0.0        2026
3  Thhalv10007139m|PACid:20185859 AT1G01260.3|PACid:19651080       0.0        1036       0.0         987
4  Thhalv10007299m|PACid:20188214 AT1G01110.2|PACid:19658169       0.0         875       0.0         840
5  Thhalv10007320m|PACid:20187751 AT1G01190.1|PACid:19654322       0.0         951       0.0         951
6  Thhalv10007351m|PACid:20187717 AT1G01120.1|PACid:19655257       0.0        1055       0.0        1055
7  Thhalv10007665m|PACid:20186865 AT1G01010.1|PACid:19656964       0.0         515       0.0         521
8  Thhalv10007695m|PACid:20187891 AT1G01090.1|PACid:19650207       0.0         842       0.0         842
9  Thhalv10008068m|PACid:20187490 AT1G01030.1|PACid:19649747       0.0         584       0.0         608
10 Thhalv10008123m|PACid:20187436 AT1G01240.3|PACid:19650438       0.0         515       0.0         530
11 Thhalv10008287m|PACid:20186971 AT1G01180.1|PACid:19654493       0.0         549       0.0         549
12 Thhalv10008355m|PACid:20185661 AT1G01080.1|PACid:19651136    9e-157         435    1e-172         476
13 Thhalv10008482m|PACid:20185881 AT1G01160.1|PACid:19654307     5e-90         261     2e-99         284
14 Thhalv10008506m|PACid:20187047 AT1G01225.1|PACid:19656715    1e-159         440    1e-164         452
15 Thhalv10008618m|PACid:20186467 AT1G01020.1|PACid:19655142    6e-141         391    1e-141         393
16 Thhalv10008767m|PACid:20187619 AT1G01050.1|PACid:19652974    1e-158         433    1e-158         433
17 Thhalv10008983m|PACid:20187021 AT1G01230.1|PACid:19650274    2e-115         319    2e-115         319
18 Thhalv10009311m|PACid:20187590 AT1G01200.1|PACid:19649914    2e-161         442    2e-161         442
19 Thhalv10009402m|PACid:20186843 AT1G01250.1|PACid:19649318    4e-119         331    6e-119         331
20 Thhalv10009501m|PACid:20185805 AT1G01070.1|PACid:19650807       0.0         633       0.0         625

$comparison_6
           # example_thalo_aa.faa     example_brapa_aa.faa evalue_ab bitscore_ab evalue_ba bitscore_ba
1  Thhalv10006531m|PACid:20187082 Bra033293|PACid:22692618       0.0        3495       0.0        3449
2  Thhalv10006637m|PACid:20187150 Bra032626|PACid:22713115       0.0        1963       0.0        1963
3  Thhalv10007139m|PACid:20185859 Bra033273|PACid:22692002       0.0         872       0.0         850
4  Thhalv10007299m|PACid:20188214 Bra033284|PACid:22692650       0.0         848       0.0         849
5  Thhalv10007320m|PACid:20187751 Bra033278|PACid:22692201       0.0         968       0.0         968
6  Thhalv10007351m|PACid:20187717 Bra033283|PACid:22693298       0.0        1036       0.0        1036
7  Thhalv10007665m|PACid:20186865 Bra033296|PACid:22693237    2e-150         432    1e-147         425
8  Thhalv10007695m|PACid:20187891 Bra033286|PACid:22694322       0.0         830       0.0         830
9  Thhalv10008068m|PACid:20187490 Bra033294|PACid:22691990       0.0         587       0.0         570
10 Thhalv10008123m|PACid:20187436 Bra032629|PACid:22714275    6e-140         396    2e-153         431
11 Thhalv10008287m|PACid:20186971 Bra033279|PACid:22694360       0.0         508       0.0         508
12 Thhalv10008355m|PACid:20185661 Bra033287|PACid:22693868    3e-129         361    7e-129         360
13 Thhalv10008482m|PACid:20185881 Bra032623|PACid:22715924    3e-115         325    2e-114         323
14 Thhalv10008506m|PACid:20187047 Bra032627|PACid:22713604    6e-167         459    2e-169         465
15 Thhalv10008618m|PACid:20186467 Bra033295|PACid:22694612    2e-141         392    3e-141         392
16 Thhalv10008767m|PACid:20187619 Bra033292|PACid:22694379    2e-158         432    2e-158         432
17 Thhalv10008983m|PACid:20187021 Bra033276|PACid:22693754    9e-111         308    1e-110         308
18 Thhalv10009311m|PACid:20187590 Bra033277|PACid:22693004    2e-159         437    2e-159         437
19 Thhalv10009402m|PACid:20186843 Bra032630|PACid:22715365    1e-106         299    9e-107         300
20 Thhalv10009501m|PACid:20185805 Bra033290|PACid:22692027       0.0         667       0.0         665

$proteinortho_tbl
   X..Species Genes Alg..Conn.       example_athal_aa.faa  example_alyra_aa.faa     example_brapa_aa.faa
1           4     4          1 AT1G01010.1|PACid:19656964 333554|PACid:16033839 Bra033296|PACid:22693237
2           4     4          1 AT1G01020.1|PACid:19655142 470181|PACid:16064328 Bra033295|PACid:22694612
3           4     4          1 AT1G01030.1|PACid:19649747 470180|PACid:16054974 Bra033294|PACid:22691990
4           4     4          1 AT1G01040.1|PACid:19652432 333551|PACid:16057793 Bra033293|PACid:22692618
5           4     4          1 AT1G01050.1|PACid:19652974 909874|PACid:16064489 Bra033292|PACid:22694379
6           4     4          1 AT1G01070.1|PACid:19650807 918864|PACid:16052578 Bra033290|PACid:22692027
7           4     4          1 AT1G01080.1|PACid:19651136 909871|PACid:16053217 Bra033287|PACid:22693868
8           4     4          1 AT1G01090.1|PACid:19650207 470171|PACid:16052860 Bra033286|PACid:22694322
9           4     4          1 AT1G01110.2|PACid:19658169 333544|PACid:16034284 Bra033284|PACid:22692650
10          4     4          1 AT1G01120.1|PACid:19655257 918858|PACid:16049140 Bra033283|PACid:22693298
11          4     4          1 AT1G01160.1|PACid:19654307 918854|PACid:16044153 Bra032623|PACid:22715924
12          4     4          1 AT1G01180.1|PACid:19654493 909860|PACid:16056125 Bra033279|PACid:22694360
13          4     4          1 AT1G01190.1|PACid:19654322 311315|PACid:16059488 Bra033278|PACid:22692201
14          4     4          1 AT1G01200.1|PACid:19649914 470156|PACid:16041002 Bra033277|PACid:22693004
15          4     4          1 AT1G01220.1|PACid:19652890 470155|PACid:16047984 Bra032626|PACid:22713115
16          4     4          1 AT1G01225.1|PACid:19656715 470154|PACid:16066159 Bra032627|PACid:22713604
17          4     4          1 AT1G01230.1|PACid:19650274 470153|PACid:16038331 Bra033276|PACid:22693754
18          4     4          1 AT1G01240.3|PACid:19650438 470152|PACid:16047967 Bra032629|PACid:22714275
19          4     4          1 AT1G01250.1|PACid:19649318 470149|PACid:16044741 Bra032630|PACid:22715365
20          4     4          1 AT1G01260.3|PACid:19651080 470147|PACid:16058975 Bra033273|PACid:22692002
             example_thalo_aa.faa
1  Thhalv10007665m|PACid:20186865
2  Thhalv10008618m|PACid:20186467
3  Thhalv10008068m|PACid:20187490
4  Thhalv10006531m|PACid:20187082
5  Thhalv10008767m|PACid:20187619
6  Thhalv10009501m|PACid:20185805
7  Thhalv10008355m|PACid:20185661
8  Thhalv10007695m|PACid:20187891
9  Thhalv10007299m|PACid:20188214
10 Thhalv10007351m|PACid:20187717
11 Thhalv10008482m|PACid:20185881
12 Thhalv10008287m|PACid:20186971
13 Thhalv10007320m|PACid:20187751
14 Thhalv10009311m|PACid:20187590
15 Thhalv10006637m|PACid:20187150
16 Thhalv10008506m|PACid:20187047
17 Thhalv10008983m|PACid:20187021
18 Thhalv10008123m|PACid:20187436
19 Thhalv10009402m|PACid:20186843
20 Thhalv10007139m|PACid:20185859
</code></pre>
<p>The resulting output is a list storing all combinations of pairwise comparisons as data frames inside each list element. The last list element defines the gene relationships in terms of gene ids.</p>
<p>In case you would like to use the <code>synteny</code> option of ProteinOrtho, we refer to the <code>ProteinOrtho()</code> function. When performing synteny, a *.gff file needs to be stored in a folder named <code>_ProteinOrtho</code> that needs to be created inside the current working directory of your R session (<code>dir.create(&quot;_ProteinOrtho&quot;)</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># it is also possible to run ProteinOrtho using the synteny option</span>
<span class="co"># make sure you can provide a *.gff file</span>
<span class="co"># see https://www.bioinf.uni-leipzig.de/Software/proteinortho/manual.html for details</span>
<span class="kw">ProteinOrtho</span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
             <span class="dt">subject_files   =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
             <span class="dt">seq_type        =</span> <span class="st">&quot;protein&quot;</span>,
             <span class="dt">format          =</span> <span class="st">&quot;fasta&quot;</span>,
             <span class="dt">po_params       =</span> <span class="st">&quot;-synteny&quot;</span>,
             <span class="dt">clean_folders   =</span> <span class="ot">FALSE</span>)</code></pre></div>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
