<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparative Genomics with R • orthologr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Comparative Genomics with R">
<meta property="og:description" content="Large scale orthology inference and dN/dS estimation between multiple species. ">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="articles/dNdS_estimation.html">dNdS estimation using orthologr</a>
    </li>
    <li>
      <a href="articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
<div id="orthologr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#orthologr" class="anchor"></a>orthologr</h1></div>
<div id="comparative-genomics-with-r" class="section level2">
<h2 class="hasAnchor">
<a href="#comparative-genomics-with-r" class="anchor"></a>Comparative Genomics with R</h2>
<div id="motivation" class="section level3">
<h3 class="hasAnchor">
<a href="#motivation" class="anchor"></a>Motivation</h3>
<p>The comparative method is a powerful approach in genomics research. Based on our knowledge about the phylogenetic relationships between species, we can study the evolution, diversification, and constraints of biological processes by comparing genomes, genes, and other genomic loci across species. The <code>orthologr</code> package aims to provide a framework to perform <strong>large scale</strong> comparative genomics studies with R. <code>Orthologr</code> aims to be as easy to use as possible - from genomic data retrieval to orthology inference and dNdS estimation between several genomes.</p>
<p>In combination with the R package <a href="https://github.com/ropensci/biomartr">biomartr</a>, users can retrieve genomes, proteomes, or coding sequences for several species and use them as input for orthology inference and dN/dS estimation with <code>orthologr</code>. The advantage of using <code>biomartr</code> in combination with <code>orthologr</code> is that users can join the new wave of research that promotes and facilitates <a href="https://github.com/ropensci/biomartr#short-package-description">computational reproducibility in genomics studies</a> and solve the issue of comparing genomes with different genome assembly qualities (also referred to as <a href="https://github.com/ropensci/biomartr#short-package-description">genome version crisis</a>).</p>
</div>
<div id="citation" class="section level3">
<h3 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h3>
<p><strong>Please cite the following paper in which I introduce <code>orthologr</code> when using this package for your own research. This will allow me to continue working on this software tool and will motivate me to extend its functionality and usability in the next years. Many thanks in advance :)</strong></p>
<blockquote>
<p>Drost et al. 2015. <strong>Evidence for Active Maintenance of Phylotranscriptomic Hourglass Patterns in Animal and Plant Embryogenesis</strong>. <em>Mol. Biol. Evol.</em> 32 (5): 1221-1231. <a href="http://mbe.oxfordjournals.org/content/32/5/1221.abstract?sid=767aea12-1eb3-40be-8c6a-e2861f159b46">doi:10.1093/molbev/msv012</a></p>
</blockquote>
</div>
<div id="short-package-description" class="section level3">
<h3 class="hasAnchor">
<a href="#short-package-description" class="anchor"></a>Short package description</h3>
<p>In detail, <code>orthologr</code> allows users to perform orthology inference and dN/dS estimation between two genomes or between several genomes. The following methods to infer orthologous relationships between genes of entire genomes are available in this package:</p>
<ul>
<li><a href="https://hajkd.github.io/orthologr/articles/blast.html#the-blast_best-function"><strong>BLAST best hit</strong></a></li>
<li><a href="https://hajkd.github.io/orthologr/articles/blast.html#the-blast_rec-function"><strong>BLAST best reciprocal hit</strong></a></li>
<li><a href="https://hajkd.github.io/orthologr/articles/orthology_inference.html"><strong>OrthoFinder2</strong></a></li>
</ul>
<p>The most useful implementation in <code>orthologr</code> is the ability to compute synonymous versus non-synonymous substitution rates (dN/dS) for all orthologous genes between two <strong>entire genomes</strong>. Available dN/dS estimation methods are:</p>
<ul>
<li>“NG”: Nei, M. and Gojobori, T. (1986)</li>
<li>“LWL”: Li, W.H., et al. (1985)</li>
<li>“LPB”: Li, W.H. (1993) and Pamilo, P. and Bianchi, N.O. (1993)</li>
<li>“MLWL” (Modified LWL), MLPB (Modified LPB): Tzeng, Y.H., et al. (2004)</li>
<li>“YN”: Yang, Z. and Nielsen, R. (2000)</li>
<li>“MYN” (Modified YN): Zhang, Z., et al. (2006)</li>
<li>“GMYN”: Wang, D.P., et al. Biology Direct. (2009)</li>
<li>“GY”: Goldman, N. and Yang, Z. (1994)</li>
<li>“MS”: (Model Selection): based on a set of candidate models, Posada, D. (2003)</li>
<li>“MA” (Model Averaging): based on a set of candidate models, Posada, D. (2003)</li>
<li>“ALL”: All models toghether</li>
</ul>
<p>Please find more details <a href="https://hajkd.github.io/orthologr/articles/dNdS_estimation.html">here</a>.</p>
</div>
</div>
<div id="install-orthologr" class="section level2">
<h2 class="hasAnchor">
<a href="#install-orthologr" class="anchor"></a>Install <code>orthologr</code>
</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ns-load">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))
    <span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"BiocManager"</span>)
BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>()
<span class="co"># Install package dependencies</span>
BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"Biostrings"</span>, <span class="dt">version =</span> <span class="st">"3.8"</span>)
<span class="co"># install orthologr from GitHub</span>
BiocManager<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocManager/topics/install">install</a></span>(<span class="st">"HajkD/orthologr"</span>)</code></pre></div>
</div>
<div id="use-cases" class="section level2">
<h2 class="hasAnchor">
<a href="#use-cases" class="anchor"></a>Use Cases</h2>
<p>Learn <code>orthologr</code> by reading these tutorials:</p>
<ul>
<li><a href="https://hajkd.github.io/orthologr/articles/Install.html">Install Prerequisite Tools</a></li>
<li><a href="https://hajkd.github.io/orthologr/articles/blast.html">Perform BLAST Searches</a></li>
<li><a href="https://hajkd.github.io/orthologr/articles/sequence_alignments.html">Perform Sequence Alignments</a></li>
<li><a href="https://hajkd.github.io/orthologr/articles/orthology_inference.html">Perform Orthology Inference</a></li>
<li><a href="https://hajkd.github.io/orthologr/articles/dNdS_estimation.html">Perform dNdS Estimation</a></li>
<li><a href="https://hajkd.github.io/orthologr/articles/divergence_stratigraphy.html">Perform Divergence Stratigraphy</a></li>
</ul>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<div id="small-example-with-internal-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#small-example-with-internal-dataset" class="anchor"></a>Small example with internal dataset</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(orthologr)

<span class="co"># Detect orthologous genes between a query species and a subject species</span>
<span class="co"># and compute the synonymous versus non-synonymous substitution rates (dN/dS)</span>
<span class="co"># following this paradigm:</span>
<span class="co"># 1) reciprocal best hit for orthology inference (RBH)</span>
<span class="co"># 2) Needleman-Wunsch for pairwise amino acid alignments</span>
<span class="co"># 3) pal2nal for codon alignments</span>
<span class="co"># 4) Comeron for dNdS estimation</span>
<span class="co"># 5) multi-core processing 'comp_cores = 1'</span>
<span class="kw"><a href="reference/dNdS.html">dNdS</a></span>(<span class="dt">query_file      =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
     <span class="dt">subject_file    =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
     <span class="dt">delete_corrupt_cds =</span> <span class="ot">TRUE</span>, <span class="co"># coding sequences that cannot be divided by 3 (triplets) will be removed</span>
     <span class="dt">ortho_detection =</span> <span class="st">"RBH"</span>, <span class="co"># perform BLAST best reciprocal hit orthology inference</span>
     <span class="dt">aa_aln_type     =</span> <span class="st">"pairwise"</span>, <span class="co"># perform pairwise global alignments of AA seqs </span>
     <span class="dt">aa_aln_tool     =</span> <span class="st">"NW"</span>, <span class="co"># using Needleman-Wunsch</span>
     <span class="dt">codon_aln_tool  =</span> <span class="st">"pal2nal"</span>, <span class="co"># perform codon alignments using the tool Pal2Nal</span>
     <span class="dt">dnds_est.method =</span> <span class="st">"Comeron"</span>, <span class="co"># use Comeron's method for dN/dS inference</span>
     <span class="dt">comp_cores      =</span> <span class="dv">1</span> )</code></pre></div>
<pre><code>   query_id subject_id      dN    dS   dNdS perc_identity alig_length mismatches gap_openings
   &lt;chr&gt;    &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;         &lt;dbl&gt;       &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt;
 1 AT1G010… 333554|PA… 0.106   0.254 0.420           74.0         469         80            8
 2 AT1G010… 470181|PA… 0.0402  0.104 0.388           91.1         246         22            0
 3 AT1G010… 470180|PA… 0.0150  0.126 0.118           95.5         359         12            2
 4 AT1G010… 333551|PA… 0.0135  0.116 0.116           92.0        1970         85           10
 5 AT1G010… 909874|PA… 0       0.175 0              100           213          0            0
 6 AT1G010… 470177|PA… 0.0449  0.113 0.397           89.5         648         58            5
 7 AT1G010… 918864|PA… 0.0183  0.106 0.173           95.1         366         14            2
 8 AT1G010… 909871|PA… 0.0340  0.106 0.322           90.3         300         22            2
 9 AT1G010… 470171|PA… 0.00910 0.218 0.0417          96.8         434          8            3
10 AT1G011… 333544|PA… 0.0325  0.122 0.266           93.6         528         34            0
11 AT1G011… 918858|PA… 0.00307 0.133 0.0232          99.2         529          4            0
12 AT1G011… 470161|PA… 0.00567 0.131 0.0432          98.4         453          6            1
13 AT1G011… 918855|PA… 0.13    0.203 0.641           72.6         285         68            3
14 AT1G011… 918854|PA… 0.105   0.280 0.373           84.9         179         19            2
15 AT1G011… 311317|PA… 0       0.306 0               85.6          97          0            1
16 AT1G011… 909860|PA… 0.0297  0.176 0.168           92.6         310         20            1
17 AT1G011… 311315|PA… 0.0287  0.162 0.177           94.2         533         30            1
18 AT1G012… 470156|PA… 0.0190  0.168 0.114           95.8         238         10            0
19 AT1G012… 311313|PA… 0.0207  0.154 0.134           95.3         107          5            0
20 AT1G012… 470155|PA… 0.0157  0.153 0.102           96.7        1056         35            0
# ... with 6 more variables: q_start &lt;dbl&gt;, q_end &lt;dbl&gt;, s_start &lt;dbl&gt;, s_end &lt;dbl&gt;,
#   evalue &lt;dbl&gt;, bit_score &lt;dbl&gt;</code></pre>
<p>When running your own query file, please specify <code>query_file = "path/to/your/cds.fasta</code> instead of <code><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file(..., package = "orthologr")</a></code>. The command <code><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file(..., package = "orthologr")</a></code> merely references the path to the example file stored in the <code>orthologr</code> package itself.</p>
</div>
<div id="example-computing-dnds-values-for-all-orthologous-genes-between-two-genomes" class="section level3">
<h3 class="hasAnchor">
<a href="#example-computing-dnds-values-for-all-orthologous-genes-between-two-genomes" class="anchor"></a>Example: Computing dN/dS values for all orthologous genes between two genomes</h3>
<p>First, users can retrieve all coding sequences from entire genomes using the <a href="https://github.com/ropensci/biomartr">biomartr</a> package (<a href="https://ropensci.github.io/biomartr/articles/Sequence_Retrieval.html#cds-retrieval">see details here</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"biomartr"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(biomartr)
<span class="co"># download all coding sequences for Mus musculus</span>
Mmusculus_file &lt;-<span class="st"> </span>biomartr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomartr/topics/getCDS">getCDS</a></span>(<span class="dt">organism =</span> <span class="st">"Mus musculus"</span>, <span class="dt">path =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())
<span class="co"># download all coding sequences for Homo sapiens</span>
Hsapiens_file &lt;-<span class="st"> </span>biomartr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomartr/topics/getCDS">getCDS</a></span>(<span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>, <span class="dt">path =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/getwd">getwd</a></span>())
<span class="co"># compute dN/dS values for Homo sapiens versus Mus musculus</span>
Hs_vs_Mm_dNdS &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="reference/dNdS.html">dNdS</a></span>(<span class="dt">query_file      =</span> Hsapiens_file,
       <span class="dt">subject_file    =</span> Mmusculus_file,
       <span class="dt">delete_corrupt_cds =</span> <span class="ot">FALSE</span>,
       <span class="dt">ortho_detection =</span> <span class="st">"RBH"</span>, 
       <span class="dt">aa_aln_type     =</span> <span class="st">"pairwise"</span>,
       <span class="dt">aa_aln_tool     =</span> <span class="st">"NW"</span>, 
       <span class="dt">codon_aln_tool  =</span> <span class="st">"pal2nal"</span>, 
       <span class="dt">dnds_est.method =</span> <span class="st">"Comeron"</span>, 
       <span class="dt">comp_cores      =</span> <span class="dv">1</span> )
<span class="co"># store result in Excel readable csv file</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages</a></span>(<span class="st">"readr"</span>)
readr<span class="op">::</span><span class="kw"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_excel_csv</a></span>(Hs_vs_Mm_dNdS, <span class="st">"Hs_vs_Mm_dNdS.csv"</span>)</code></pre></div>
<p>Users can find the corresponding map at <a href="https://github.com/HajkD/dNdS_database" class="uri">https://github.com/HajkD/dNdS_database</a>.</p>
<p><strong>This way, users can compute dN/dS values for any pairwise genome comparison.</strong></p>
</div>
<div id="on-windows-systems" class="section level3">
<h3 class="hasAnchor">
<a href="#on-windows-systems" class="anchor"></a>On Windows Systems</h3>
<p>In some cases (when working with <strong>WINDOWS</strong> machines), the installation via <code>devtools</code> will not work properly. In this case users can try the follwing steps:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># On Windows, this won't work - see ?build_github_devtools</span>
<span class="kw">install_github</span>(<span class="st">"HajkD/orthologr"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)

<span class="co"># When working with Windows, first users need to install the</span>
<span class="co"># R package: rtools -&gt; install.packages("rtools")</span>

<span class="co"># Afterwards users can install devtools -&gt; install.packages("devtools")</span>
<span class="co"># and then they can run:</span>

devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"HajkD/orthologr"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)

<span class="co"># and then call it from the library</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"orthologr"</span>, <span class="dt">lib.loc =</span> <span class="st">"C:/Program Files/R/R-3.1.1/library"</span>)</code></pre></div>
</div>
<div id="troubleshooting-on-windows-machines" class="section level3">
<h3 class="hasAnchor">
<a href="#troubleshooting-on-windows-machines" class="anchor"></a>Troubleshooting on Windows Machines</h3>
<ul>
<li>Install <code>orthologr</code> on a Win 8 laptop: <a href="https://github.com/HajkD/orthologr/issues/1">solution</a> ( Thanks to Andres Romanowski )</li>
</ul>
</div>
</div>
<div id="interfaces-implemented-in-orthologr" class="section level2">
<h2 class="hasAnchor">
<a href="#interfaces-implemented-in-orthologr" class="anchor"></a>Interfaces implemented in <code>orthologr</code>:</h2>
<div id="perform-blast-searches-with-r" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-blast-searches-with-r" class="anchor"></a>Perform BLAST searches with R</h3>
<ul>
<li>
<code><a href="reference/advanced_blast.html">advanced_blast()</a></code>: Perform an advanced BLAST+ search</li>
<li>
<code><a href="reference/advanced_makedb.html">advanced_makedb()</a></code>: Create a BLASTable database with makeblastdb (advanced options)</li>
<li>
<code><a href="reference/blast.html">blast()</a></code>: Perform a BLAST+ search</li>
<li>
<code><a href="reference/blast.nr.html">blast.nr()</a></code>: Perform a BLASTP search against NCBI nr</li>
<li>
<code><a href="reference/blast_best.html">blast_best()</a></code>: Perform a BLAST+ best hit search</li>
<li>
<code><a href="reference/blast_rec.html">blast_rec()</a></code>: Perform a BLAST+ best reciprocal hit (BRH) search</li>
<li>
<code><a href="reference/delta.blast.html">delta.blast()</a></code>: Perform a DELTA-BLAST Search</li>
</ul>
</div>
<div id="perform-pairwise-and-multiple-sequence-alignements-with-r" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-pairwise-and-multiple-sequence-alignements-with-r" class="anchor"></a>Perform Pairwise and Multiple Sequence Alignements with R</h3>
<ul>
<li>
<code><a href="reference/multi_aln.html">multi_aln()</a></code>: Compute Multiple Sequence Alignments based on the <code>clustalw</code>, <code>t_coffee</code>, <code>muscle</code>, <code>clustalo</code>, and <code>mafft</code> programs.</li>
<li>
<code><a href="reference/pairwise_aln.html">pairwise_aln()</a></code>: Compute Pairwise Alignments</li>
<li>
<code><a href="reference/codon_aln.html">codon_aln()</a></code>: Compute a Codon Alignment</li>
</ul>
</div>
<div id="perform-orthology-inference-with-r" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-orthology-inference-with-r" class="anchor"></a>Perform Orthology Inference with R</h3>
<ul>
<li>
<code><a href="reference/orthologs.html">orthologs()</a></code>: Main Orthology Inference Function</li>
<li>
<code><a href="reference/ProteinOrtho.html">ProteinOrtho()</a></code>: Orthology Inference with ProteinOrtho</li>
</ul>
</div>
<div id="perform-population-genomics-with-r" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-population-genomics-with-r" class="anchor"></a>Perform Population Genomics with R</h3>
<ul>
<li>
<code><a href="reference/dNdS.html">dNdS()</a></code>: Compute dNdS values for two organisms</li>
<li>
<code><a href="reference/substitutionrate.html">substitutionrate()</a></code>: Internal function for dNdS computations</li>
</ul>
</div>
<div id="read-and-write-cds-genomes-and-proteomes" class="section level3">
<h3 class="hasAnchor">
<a href="#read-and-write-cds-genomes-and-proteomes" class="anchor"></a>Read and Write CDS, Genomes, and Proteomes</h3>
<ul>
<li>
<code><a href="reference/read.cds.html">read.cds()</a></code>: Read the CDS of a given organism</li>
<li>
<code><a href="reference/read.genome.html">read.genome()</a></code>: Read the genome of a given organism</li>
<li>
<code><a href="reference/read.proteome.html">read.proteome()</a></code>: Read the proteome of a given organism</li>
<li>
<code><a href="reference/write.proteome.html">write.proteome()</a></code>: Save a proteome in fasta format</li>
</ul>
</div>
</div>
<div id="discussions-and-bug-reports" class="section level2">
<h2 class="hasAnchor">
<a href="#discussions-and-bug-reports" class="anchor"></a>Discussions and Bug Reports</h2>
<p>I would be very happy to learn more about potential improvements of the concepts and functions provided in this package.</p>
<p>Furthermore, in case you find some bugs, need additional (more flexible) functionality of parts of this package, or want to contribute to this project please let me know:</p>
<p><a href="https://github.com/HajkD/orthologr/issues" class="uri">https://github.com/HajkD/orthologr/issues</a></p>
</div>
<div id="licenses" class="section level2">
<h2 class="hasAnchor">
<a href="#licenses" class="anchor"></a>Licenses</h2>
<p>The <code>orthologr</code> package includes source code that has been published under following licenses:</p>
<div id="gestimator" class="section level3">
<h3 class="hasAnchor">
<a href="#gestimator" class="anchor"></a>gestimator</h3>
<pre><code>All files included in `orthologr` that were taken from gestimator are 
also part of libsequence.

libsequence is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.

libsequence is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
long with libsequence.  If not, see &lt;http://www.gnu.org/licenses/&gt;.


Modified by Sarah Scharfenberg and Hajk-Georg Drost (2014) to work 
in orthologr without using external libraries from libsequence.

All changes are also free under the terms of GNU General Public License
version 2 of the License, or any later version.
</code></pre>
</div>
<div id="kaks_calculator" class="section level3">
<h3 class="hasAnchor">
<a href="#kaks_calculator" class="anchor"></a>KaKs_Calculator</h3>
<p>In <code>orthologr</code> the file <code>parseFastaIntoAXT.pl</code> is stored in <code>/inst/KaKs_Calc_parser</code>.</p>
<pre><code>The parseFastaIntoAXT.pl script is freely available under GNU GPL v3 
Licence and included in the KaKs_Calculator project that can be found at 
https://code.google.com/p/kaks-calculator/
</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/HajkD/orthologr">https://​github.com/​HajkD/​orthologr</a>
</li>
<li>Report a bug at <br><a href="https://github.com/HajkD/orthologr/issues">https://​github.com/​HajkD/​orthologr/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2">GPL-2</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing orthologr</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Hajk-Georg Drost <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  </div>
</div>

      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
