<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BLAST searches using orthologr • orthologr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BLAST searches using orthologr">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">dNdS estimation using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>BLAST searches using orthologr</h1>
            
            <h4 class="date">2018-12-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/HajkD/orthologr/blob/master/vignettes/blast.Rmd"><code>vignettes/blast.Rmd</code></a></small>
      <div class="hidden name"><code>blast.Rmd</code></div>

    </div>

    
    
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting Started</h2>
<p>The <code>orthologr</code> package provides several interface functions to perform BLAST searches.</p>
<p>First, users need to make sure that they have <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">BLAST</a> installed on their machine. Please follow <a href="https://github.com/HajkD/orthologr/blob/master/vignettes/Install.Rmd#install-blast">these</a> instructions to install BLAST on your manine.</p>
</div>
<div id="performing-blast-searches" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-blast-searches" class="anchor"></a>Performing BLAST Searches</h2>
<p>The <code>orthologr</code> package stores 20 example genes (orthologs) between <em>Arabidopsis thaliana</em> and <em>Arabidopsis lyrata</em>. The following example BLAST search shall illustrate a simple search with standard parameters provided by the <code><a href="../reference/blast.html">blast()</a></code> function.</p>
<p>When running the subsequent functions please make sure you can call BLAST+ from your console either in the standard <code>PATH</code> or in case you have BLAST+ installed in a separate folder, please specify the <code>path</code> argument that can be passed to <code><a href="../reference/blast.html">blast()</a></code>.</p>
<p>To check whether BLAST+ can be executed from the default <code>PATH</code> (<code>usr/bin/local</code> on UNIX systems), you can run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"blastp -version"</span>)</code></pre></div>
<p>This should return something like this:</p>
<pre><code>blastp: 2.2.29+
Package: blast 2.2.29, build Dec 10 2013 15:51:59
</code></pre>
<p>If everything works properly, you can get started with you first BLAST+ search.</p>
</div>
<div id="the-blast-function" class="section level2">
<h2 class="hasAnchor">
<a href="#the-blast-function" class="anchor"></a>The blast() function</h2>
<p>The <code><a href="../reference/blast.html">blast()</a></code> function provides the easiest way to perform a BLAST search.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)

<span class="co"># performing a BLAST search using blastp (default)</span>
hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast.html">blast</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))


<span class="kw"><a href="http://dplyr.tidyverse.org/reference/reexports.html">glimpse</a></span>(hit_tbl)</code></pre></div>
<pre><code>Variables:
$ query_id      (chr) "AT1G01010.1", "AT1G01020.1", "AT1G01030.1",...
$ subject_id    (chr) "333554|PACid:16033839", "470181|PACid:16064...
$ perc_identity (dbl) 73.99, 91.06, 95.54, 91.98, 100.00, 89.51, 9...
$ alig_length   (dbl) 469, 246, 359, 1970, 213, 648, 366, 300, 434...
$ mismatches    (dbl) 80, 22, 12, 85, 0, 58, 14, 22, 8, 34, 4, 6, ...
$ gap_openings  (dbl) 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, 1, 3, 2, 1...
$ q_start       (dbl) 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2,...
$ q_end         (dbl) 430, 246, 359, 1910, 213, 646, 366, 294, 429...
$ s_start       (dbl) 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4...
$ s_end         (dbl) 466, 246, 355, 1963, 213, 640, 362, 299, 433...
$ evalue        (dbl) 0e+00, 7e-166, 0e+00, 0e+00, 2e-160, 0e+00, ...
$ bit_score     (dbl) 627, 454, 698, 3704, 437, 1037, 696, 491, 85...
</code></pre>
<p>As you can see, the hit table shows the output of the BLAST+ search. The <code><a href="../reference/blast.html">blast()</a></code> function runs <code>blastp</code> as default BLAST+ algorithm. Different BLAST+ algorithms can be selected by specifying the <code>blast_algorithm</code> argument, e.g. <code>blast_algorithm = "tblastn"</code>. See <code><a href="../reference/blast.html">?blast</a></code> for further details. The <code><a href="../reference/blast.html">blast()</a></code> function returns the BLAST arguments: <code>query_id</code>, <code>subject_id</code>, <code>perc_identity</code>, <code>alig_length</code>, <code>mismatches</code>, <code>gap_openings</code>, <code>q_start</code>, <code>q_end</code>, <code>s_start</code>, <code>s_end</code>, <code>evalue</code>, and <code>bit_score</code>.</p>
<p>Since <code><a href="../reference/blast.html">blast()</a></code> stores the hit table returned by BLAST in a data.table object, you can access each column, using the <a href="https://github.com/Rdatatable/data.table/wiki">data.table notation</a>.</p>
<p>In case you need to specify the <code>PATH</code> to BLAST+ please use the <code>path</code> argument:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># performing a BLAST search using blastp (default)</span>
hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast.html">blast</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">path         =</span> <span class="st">"/path/to/blastp"</span>)


hit_tbl</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># access columns: query_id, subject_id, evalue, and bit_score</span>
hit_tbl[ , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(query_id, subject_id, evalue, bit_score)]</code></pre></div>
<pre><code>
       query_id            subject_id evalue bit_score
 1: AT1G01010.1 333554|PACid:16033839  0e+00       627
 2: AT1G01020.1 470181|PACid:16064328 7e-166       454
 3: AT1G01030.1 470180|PACid:16054974  0e+00       698
 4: AT1G01040.1 333551|PACid:16057793  0e+00      3704
 5: AT1G01050.1 909874|PACid:16064489 2e-160       437
 6: AT1G01060.3 470177|PACid:16043374  0e+00      1037
 7: AT1G01070.1 918864|PACid:16052578  0e+00       696
 8: AT1G01080.1 909871|PACid:16053217 1e-178       491
 9: AT1G01090.1 470171|PACid:16052860  0e+00       859
10: AT1G01110.2 333544|PACid:16034284  0e+00       972
11: AT1G01120.1 918858|PACid:16049140  0e+00      1092
12: AT1G01140.3 470161|PACid:16036015  0e+00       918
13: AT1G01150.1 918855|PACid:16037307 3e-150       421
14: AT1G01160.1 918854|PACid:16044153  1e-93       268
15: AT1G01170.2 311317|PACid:16052302  3e-54       158
16: AT1G01180.1 909860|PACid:16056125  0e+00       576
17: AT1G01190.1 311315|PACid:16059488  0e+00      1036
18: AT1G01200.1 470156|PACid:16041002 3e-172       470
19: AT1G01210.1 311313|PACid:16057125  7e-76       215
20: AT1G01220.1 470155|PACid:16047984  0e+00      2106
</code></pre>
<p>The <code><a href="../reference/blast.html">blast()</a></code> function also allows you to pass additional parameters to the BLAST+ search using the <code>blast_params</code> argument. In the following example, a remote BLAST+ search is performed.</p>
<pre><code><a href="../reference/blast.html">
blast(query_file = system.file('seqs/ortho_thal_cds.fasta', package = 'orthologr'),
                 subject_file = system.file('seqs/ortho_lyra_cds.fasta', package = 'orthologr'),
                 blast_params = "-max_target_seqs 1")</a></code></pre>
<p>In all cases the default <code>e-value</code> BLAST+ searches is <code>1E-5</code> and the default <code>blast_algorithm</code> is <code>blastp</code>.</p>
<p>Since BLAST+ searches can be computationally expensive, it is possible to specify the <code>comp_cores</code> argument when working with an multicore machine.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># BLAST computations using the comp_cores parameter: here with 2 cores</span>
<span class="kw"><a href="../reference/blast.html">blast</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
      <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
      <span class="dt">comp_cores   =</span> <span class="dv">2</span>)</code></pre></div>
<p>The <code>query_file</code> and <code>subject_file</code> arguments specify the path to the corresponding fasta files storing the <code>CDS</code> files, <code>amino acid</code> files, or <code>genome</code> files of the query organism and subject organism of interest. Make sure that when using <code>CDS</code>files, <code>amino acid</code> files, or <code>genome</code> files the corresponding argument <code>seq_type</code> must be adapted according to the input data format.</p>
<p>Use :</p>
<ul>
<li>
<code>CDS</code> files -&gt; <code>seq_type = "cds"</code>
</li>
<li>
<code>amino acid</code> files -&gt; <code>seq_type = "protein"</code>
</li>
<li>
<code>genome</code> files -&gt; <code>seq_type = "dna"</code>
</li>
</ul>
<p>The <code>format</code> argument specifies the input file format, e.g. “fasta” or “gbk”. The <code>blast_algorithm</code> argument specifies the BLAST program (algorithm) that shall be used to perform BLAST searches, e.g. “blastp”,“blastn”,“tblastn”,etc. Again, the <code>eval</code> argument defines the default e-value that shall be chosen as best hit threshold.</p>
<div id="using-the-split-apply-combine-strategy-for-a-blast-hit-table" class="section level3">
<h3 class="hasAnchor">
<a href="#using-the-split-apply-combine-strategy-for-a-blast-hit-table" class="anchor"></a>Using the split-apply-combine strategy for a BLAST hit table</h3>
<p>All <code>blast</code> functions implemented in <code>orthologr</code> can easily be processed using the <a href="http://www.jstatsoft.org/v40/i01/paper">split-apply-combine strategy</a> to detect for example <code>one-to-one</code>, <code>one-to-many</code>, and <code>many-to-many</code> gene homology relationships.</p>
<p>Here a simple example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># install.packages(c("plyr","dplyr"), dependencies = TRUE)</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(plyr)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)

<span class="co"># perform a blastp search of 20 A. thaliana genes against 1000 A. lyrata genes</span>
hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast.html">blast</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))

<span class="co"># determine 'one-to-many' and 'one-to-one' gene relationships</span>
rel_hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/plyr/topics/ddply">ddply</a></span>(<span class="dt">.data =</span> hit_tbl, <span class="dt">.variables =</span> <span class="st">"query_id"</span>, <span class="dt">.fun =</span> nrow)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>(rel_hit_tbl)[<span class="dv">2</span>] &lt;-<span class="st"> "n_genes"</span>

rel_hit_tbl</code></pre></div>
<pre><code>      query_id n_genes
1  AT1G01010.1       4
2  AT1G01020.1       1
3  AT1G01030.1       1
4  AT1G01040.1       1
5  AT1G01050.1       1
6  AT1G01060.3       2
7  AT1G01070.1       3
8  AT1G01080.1       4
9  AT1G01090.1       1
10 AT1G01110.2       1
11 AT1G01120.1       3
12 AT1G01140.3      36
13 AT1G01150.1       1
14 AT1G01160.1       1
15 AT1G01170.2       1
16 AT1G01180.1       1
17 AT1G01190.1       6
18 AT1G01200.1       8
19 AT1G01210.1       1
20 AT1G01220.1       1
</code></pre>
<p>Now you can sort genes into classes: <code>one-to-one</code> and <code>one-to-many</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># classify into 'one-to-one' relationships</span>
one_to_one &lt;-<span class="st"> </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(rel_hit_tbl,n_genes <span class="op">==</span><span class="st"> </span><span class="dv">1</span>)

<span class="co"># classify into 'one-to-many' relationships</span>
one_to_many &lt;-<span class="st"> </span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(rel_hit_tbl,n_genes <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at one_to_one</span>
one_to_one </code></pre></div>
<pre><code>      query_id n_genes
1  AT1G01020.1       1
2  AT1G01030.1       1
3  AT1G01040.1       1
4  AT1G01050.1       1
5  AT1G01090.1       1
6  AT1G01110.2       1
7  AT1G01150.1       1
8  AT1G01160.1       1
9  AT1G01170.2       1
10 AT1G01180.1       1
11 AT1G01210.1       1
12 AT1G01220.1       1
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at one_to_many</span>
one_to_many</code></pre></div>
<pre><code>     query_id n_genes
1 AT1G01010.1       4
2 AT1G01060.3       2
3 AT1G01070.1       3
4 AT1G01080.1       4
5 AT1G01120.1       3
6 AT1G01140.3      36
7 AT1G01190.1       6
8 AT1G01200.1       8
</code></pre>
<p>Now we can treat classes: <code>one_to_one</code> and <code>one_to_many</code> differently:</p>
</div>
<div id="one-to-one-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#one-to-one-genes" class="anchor"></a>one-to-one genes</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at the evalue, perc_identity, and alig_length of one_to_one genes</span>
oo_genes &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(hit_tbl,query_id <span class="op">%in%</span><span class="st"> </span>one_to_one[ , <span class="st">"query_id"</span>])

oo_genes[ , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(query_id,subject_id,evalue,perc_identity,alig_length)]</code></pre></div>
<pre><code>      query_id            subject_id evalue perc_identity alig_length
1  AT1G01020.1 470181|PACid:16064328 3e-164         91.06         246
2  AT1G01030.1 470180|PACid:16054974  0e+00         95.54         359
3  AT1G01040.1 333551|PACid:16057793  0e+00         91.98        1970
4  AT1G01050.1 909874|PACid:16064489 1e-158        100.00         213
5  AT1G01090.1 470171|PACid:16052860  0e+00         96.77         434
6  AT1G01110.2 333544|PACid:16034284  0e+00         93.56         528
7  AT1G01150.1 918855|PACid:16037307 1e-148         72.63         285
8  AT1G01160.1 918854|PACid:16044153  5e-92         84.92         179
9  AT1G01170.2 311317|PACid:16052302  1e-52         85.57          97
10 AT1G01180.1 909860|PACid:16056125  0e+00         92.58         310
11 AT1G01210.1 311313|PACid:16057125  3e-74         95.33         107
12 AT1G01220.1 470155|PACid:16047984  0e+00         96.69        1056
</code></pre>
<p>Now you could filter for additional criteria to define a first set of true orthologs. In this example we define true orthologs as <code>one_to_one</code> genes having a minimum alignment length of 300, a perc_identity of &gt; 80 percent and an e-value &lt; 1E-5.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at the evalue, perc_identity, and alig_length of one_to_one genes</span>
oo_genes &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(hit_tbl,query_id <span class="op">%in%</span><span class="st"> </span>one_to_one[ , <span class="st">"query_id"</span>])

true_orthologs &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(oo_genes,evalue <span class="op">&lt;</span><span class="st"> </span><span class="fl">1e-5</span>, perc_identity <span class="op">&gt;</span><span class="st"> </span><span class="dv">80</span>, alig_length <span class="op">&gt;</span><span class="st"> </span><span class="dv">300</span>)

true_orthologs[ , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(query_id,subject_id,evalue,perc_identity,alig_length)]</code></pre></div>
<pre><code>      query_id            subject_id evalue perc_identity alig_length
1: AT1G01030.1 470180|PACid:16054974      0         95.54         359
2: AT1G01040.1 333551|PACid:16057793      0         91.98        1970
3: AT1G01090.1 470171|PACid:16052860      0         96.77         434
4: AT1G01110.2 333544|PACid:16034284      0         93.56         528
5: AT1G01180.1 909860|PACid:16056125      0         92.58         310
6: AT1G01220.1 470155|PACid:16047984      0         96.69        1056
</code></pre>
<p>This way we could filter out a high confidence set of orthologous genes from the <code>one_to_one</code> class of genes.</p>
<p>In reality most orthology inference programs and methods perform way more complicated and sophisticated analyses to distinguish true orthologs from true paralogs (in-paralogs, out-paralogs, etc.). These subsequent analyses can also be performed using the above introduced split-apply-combine strategy.</p>
<p>Note, that you can perform self-BLAST searches <code><a href="../reference/blast.html">blast(query,query)</a></code> and <code><a href="../reference/blast.html">blast(subject,subject)</a></code> to distinguish between orthologous and paralogous genes.</p>
<p>Now we continue with the <code>one_to_many</code> class of genes.</p>
</div>
<div id="one-to-many-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#one-to-many-genes" class="anchor"></a>one-to-many genes</h3>
<p>Here we want to address the question how to deal with multiple hits returned by BLAST+ .</p>
<p>Again we investigate all <code>one_to_many</code> genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">one_to_many</code></pre></div>
<pre><code>     query_id n_genes
1 AT1G01010.1       4
2 AT1G01060.3       2
3 AT1G01070.1       3
4 AT1G01080.1       4
5 AT1G01120.1       3
6 AT1G01140.3      36
7 AT1G01190.1       6
8 AT1G01200.1       8
</code></pre>
<p>When looking at gene_id <code>AT1G01200.1</code> we see that it was found 8 times in the corresponding subject set of <em>A. lyrata</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hit_tbl[<span class="st">"AT1G01200.1"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(query_id,subject_id,evalue,perc_identity,alig_length)]</code></pre></div>
<pre><code>      query_id            subject_id evalue perc_identity alig_length
1: AT1G01200.1 470156|PACid:16041002 2e-170         95.80         238
2: AT1G01200.1 909905|PACid:16035105  7e-06         21.64         171
3: AT1G01200.1 910431|PACid:16035207  8e-74         52.97         219
4: AT1G01200.1 918732|PACid:16054958  2e-50         44.56         193
5: AT1G01200.1 919287|PACid:16060536  1e-68         58.10         179
6: AT1G01200.1 919355|PACid:16050170  7e-72         53.30         212
7: AT1G01200.1 919721|PACid:16036935  9e-80         59.31         204
8: AT1G01200.1 919852|PACid:16055066  4e-07         24.03         154
</code></pre>
<p>Now we have to decide which hit shall be considered as potential <em>ortholog</em>. In this example <code>subject_id</code> <code>470156|PACid:16041002</code> has the highest <code>perc_identity</code> as well as the lowest e-value <code>2e-170</code>. So a straightforward approach would be to choose subject gene <code>470156|PACid:16041002</code> as potential ortholog of query gene <code>AT1G01200.1</code>.</p>
<p>We can validate this approach by running a reciprocal best hit search with <code><a href="../reference/blast_rec.html">blast_rec()</a></code>and compare the output of gene <code>AT1G01200.1</code> with our choice <code>470156|PACid:16041002</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbh_hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast_rec.html">blast_rec</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                         <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))


rbh_hit_tbl</code></pre></div>
<pre><code>      query_id            subject_id evalue
1  AT1G01010.1 333554|PACid:16033839  0e+00
2  AT1G01020.1 470181|PACid:16064328 3e-164
3  AT1G01030.1 470180|PACid:16054974  0e+00
4  AT1G01040.1 333551|PACid:16057793  0e+00
5  AT1G01050.1 909874|PACid:16064489 1e-158
6  AT1G01060.3 470177|PACid:16043374  0e+00
7  AT1G01070.1 918864|PACid:16052578  0e+00
8  AT1G01080.1 909871|PACid:16053217 5e-177
9  AT1G01090.1 470171|PACid:16052860  0e+00
10 AT1G01110.2 333544|PACid:16034284  0e+00
11 AT1G01120.1 918858|PACid:16049140  0e+00
12 AT1G01140.3 470161|PACid:16036015  0e+00
13 AT1G01150.1 918855|PACid:16037307 1e-148
14 AT1G01160.1 918854|PACid:16044153  5e-92
15 AT1G01170.2 311317|PACid:16052302  1e-52
16 AT1G01180.1 909860|PACid:16056125  0e+00
17 AT1G01190.1 311315|PACid:16059488  0e+00
18 AT1G01200.1 470156|PACid:16041002 2e-170
19 AT1G01210.1 311313|PACid:16057125  3e-74
20 AT1G01220.1 470155|PACid:16047984  0e+00
</code></pre>
<p>When we now look at gene <code>AT1G01200.1</code> we find that indeed subject gene <code>470156|PACid:16041002</code> has been detected as potential ortholog having the same evalue <code>2e-170</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbh_hit_tbl[<span class="st">"AT1G01200.1"</span> , ]</code></pre></div>
<pre><code>     query_id            subject_id evalue
1 AT1G01200.1 470156|PACid:16041002 2e-170
</code></pre>
<p>Nevertheless there might be cases in which it is hard to decide for or against the best hit compared with all other hits.</p>
<p>For example we can investigate gene <code>AT1G01070.1</code> :</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hit_tbl[<span class="st">"AT1G01070.1"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(query_id,subject_id,evalue,perc_identity,alig_length)]</code></pre></div>
<pre><code>      query_id            subject_id evalue perc_identity alig_length
1: AT1G01070.1 918864|PACid:16052578  0e+00         95.08         366
2: AT1G01070.1 919693|PACid:16048878  2e-67         32.87         356
3: AT1G01070.1 919961|PACid:16062329  0e+00         79.29         338
</code></pre>
<p>Here both e-values for subject genes <code>918864|PACid:16052578</code> and <code>919961|PACid:16062329</code> are the same and only <code>perc_identity</code> and <code>alig_length</code> differ. The <em>reciprocal best hit</em> approach chose gene <code>918864|PACid:16052578</code> which also had the highest <code>perc_identity</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rbh_hit_tbl[<span class="st">"AT1G01070.1"</span> , ]</code></pre></div>
<pre><code>     query_id            subject_id evalue
1 AT1G01070.1 918864|PACid:16052578      0
</code></pre>
<p>But since the <code><a href="../reference/blast_rec.html">blast_rec()</a></code> function was implemented to choose the bidirectional best hit based on the e-value, in border line cases a different gene as expected could be chosen.</p>
<p>An alternative analysis that can be performed with these three candidate subject genes is the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read CDS sequences of the 20 example query genes of A. thaliana</span>
Ath.cds &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read.cds.html">read.cds</a></span>(<span class="dt">file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                    <span class="dt">format =</span> <span class="st">"fasta"</span>)

<span class="co"># read CDS sequences of the 1000 example subject genes of A. lyrata</span>
Aly.cds &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read.cds.html">read.cds</a></span>(<span class="dt">file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                    <span class="dt">format =</span> <span class="st">"fasta"</span>)


<span class="co"># show the sequence of gene AT1G01070.1</span>
Ath.cds[<span class="st">"AT1G01070.1"</span> , seqs]</code></pre></div>
<pre><code>[1] "atggctggagatatgcaaggagtgagagtagtagaaaaatattcaccggtcatagtgatggtgatgtcaaatgta
gcgatgggttcggtgaatgcacttgtgaagaaagctcttgatgttggtgtgaaccatatggtcattggtgcttatcgaat
ggctatttccgctttaattttggttccctttgcctatgtcttggaaaggaaaacaagaccacaaataacgtttaggctaa
tggtcgatcatttcgtcagtggccttctcggggcgagtttgatgcagtttttctttttgcttggtctgtcgtacacgtca
gcaactgtttcgtgtgctttggtaagcatgttgcctgcaatcaccttcgctttggcccttattttcaggactgaaaatgt
gaagattctaaagaccaaagcaggaatgttgaaggtgattggaactttgatctgtataagtggagctttgttcttaacat
tttacaaaggcccacaaatatcaaactctcactctcactctcacggtggggcttcccacaacaacaacgatcaagacaag
gccaataattggcttcttggatgtctttatttaaccataggaacagtgttgctatctctatggatgttgtttcaagggac
tttaagtattaagtacccttgcaaatactcgagcacttgtcttatgtcaattttcgcggcatttcaatgtgctctcttga
gcctttacaagagcagagacgttaatgattggatcatagatgatagattcgttatcaccgtcatcatatacgctggagtg
gtaggacaagcaatgacgacggttgcaacaacatgggggattaaaaaattaggagctgtgttcgcatcggcgtttttccc
acttactctcatttcggctactctatttgatttcctcattttacacactcctttataccttggaagtgtgattggatcac
tagtgaccataacgggtctctacatgttcttgtgggggaagaacaaagaaacggaatcatcaactgcattgtcttcagga
atggataacgaagctcaatatactactcctaataaggataacgactctaagtcgcccgtttaa"
</code></pre>
<p>Now you can perform a global alignment between the CDS sequences of <code>AT1G01070.1</code> and the three subject genes as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(Biostrings)

<span class="co"># perform 3 global alignments between:  AT1G01070.1 and 918864|PACid:16052578, </span>
<span class="co"># 919693|PACid:16048878, 919961|PACid:16062329</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(Aly.cds[ hit_tbl[ <span class="st">"AT1G01070.1"</span>, subject_id], seqs ], pairwiseAlignment, 
       <span class="dt">pattern =</span> Ath.cds[<span class="st">"AT1G01070.1"</span> , seqs],
       <span class="dt">type    =</span> <span class="st">"global"</span> )</code></pre></div>
<pre><code>$...
Global PairwiseAlignmentsSingleSubject (1 of 1)
pattern: [1] atggctggagatatgcaaggagtgagagta...aaggataacgactctaagtcgcccgtttaa 
subject: [1] atgggtgaaggtatgattggagtgagagta...aaggataacgactctaagtcgcccgtttaa 
score: 1768.965 

$...
Global PairwiseAlignmentsSingleSubject (1 of 1)
pattern: [1] atggctgga---gatatgcaaggagtgaga...----cgac----tctaagtcgcccgtttaa 
subject: [1] atggctaaatcagatatgc------tg---...ggttccacaaggtctatatcgcc---ttaa 
score: -2318.726 

$...
Global PairwiseAlignmentsSingleSubject (1 of 1)
pattern: [1] atggctggagatatgcaaggagtgagagta...aaggataacgactctaagtcgcccgtttaa 
subject: [1] atgagtgaggatatgggaggagtgaaagta...----------------------------aa 
score: 486.462 
</code></pre>
<p><strong>Note</strong>: To obtain the score value, you need to specify the <code>scoreOnly = TRUE</code> in the <code>pairwiseAlignment</code> function.</p>
<p>As you can see, subject gene <code>918864|PACid:16052578</code> also has the highest global alignment score <code>1768.965</code> based on the Needleman-Wunsch algorithm. This strategy might help you to differentiate between border line cases.</p>
<p>The examples shown above shall demonstrate the use cases that can be performed using the <code>blast</code> functions implemented in <code>orthologr</code>.</p>
<p>Another useful analysis can be to take the length of the initial query genes into account using the <code><a href="https://www.rdocumentation.org/packages/base/topics/nchar">nchar()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show the length distribution of all genes</span>
<span class="co"># stored in "Ath.cds"</span>
Ath.cds[ , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nchar">nchar</a></span>(seqs)]</code></pre></div>
<pre><code>
[1] 1290  738 1077 5730  639 1938 1098  882 1287 1584 1587 1356 1038  588  252 1437 1608  714  321 3168
</code></pre>
<p>Or the length of a specific gene:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Ath.cds[<span class="st">"AT1G01070.1"</span> , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nchar">nchar</a></span>(seqs)]</code></pre></div>
<pre><code>[1] 1098</code></pre>
<p>This way you can easily visualize the length distribution of genes stored in your query organism file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Ath.cds &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read.cds.html">read.cds</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                    <span class="dt">format =</span> <span class="st">"fasta"</span>)
                    
                    
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/hist">hist</a></span>(Ath.cds[ , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nchar">nchar</a></span>(seqs)], <span class="dt">breaks =</span> <span class="dv">100</span>)</code></pre></div>
</div>
</div>
<div id="the-blast_best-function" class="section level2">
<h2 class="hasAnchor">
<a href="#the-blast_best-function" class="anchor"></a>The blast_best() function</h2>
<p>For some analyses it is sufficient to perform BLAST+ best hit searches. The <code><a href="../reference/blast_best.html">blast_best()</a></code> function is optimized to perform BLAST+ best hit searches (only based on the minimum e-value) and returns the best hit when performing a BLAST+ search of a query organisms (or set of query genes) against a subject organism (or set of subject genes).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># performing gene orthology inference using the best hit (BH) method</span>
<span class="kw"><a href="../reference/blast_best.html">blast_best</a></span>(<span class="dt">query_file    =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
           <span class="dt">subject_file  =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
           <span class="dt">clean_folders =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>       query_id            subject_id evalue
 1: AT1G01010.1 333554|PACid:16033839  0e+00
 2: AT1G01020.1 470181|PACid:16064328 7e-166
 3: AT1G01030.1 470180|PACid:16054974  0e+00
 4: AT1G01040.1 333551|PACid:16057793  0e+00
 5: AT1G01050.1 909874|PACid:16064489 2e-160
 6: AT1G01060.3 470177|PACid:16043374  0e+00
 7: AT1G01070.1 918864|PACid:16052578  0e+00
 8: AT1G01080.1 909871|PACid:16053217 1e-178
 9: AT1G01090.1 470171|PACid:16052860  0e+00
10: AT1G01110.2 333544|PACid:16034284  0e+00
11: AT1G01120.1 918858|PACid:16049140  0e+00
12: AT1G01140.3 470161|PACid:16036015  0e+00
13: AT1G01150.1 918855|PACid:16037307 3e-150
14: AT1G01160.1 918854|PACid:16044153  1e-93
15: AT1G01170.2 311317|PACid:16052302  3e-54
16: AT1G01180.1 909860|PACid:16056125  0e+00
17: AT1G01190.1 311315|PACid:16059488  0e+00
18: AT1G01200.1 470156|PACid:16041002 3e-172
19: AT1G01210.1 311313|PACid:16057125  7e-76
20: AT1G01220.1 470155|PACid:16047984  0e+00
</code></pre>
<p>The <code><a href="../reference/blast_best.html">blast_best()</a></code> function returns: <code>query_id</code>, <code>subject_id</code>, and <code>eval</code>.</p>
<p>In case you need more parameters returned by a BLAST+ best hit search, you can specify the <code>detailed_output</code> argument (<code>detailed_output = TRUE</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># BLAST+ best hit search</span>
best_hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast_best.html">blast_best</a></span>(<span class="dt">query_file      =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                           <span class="dt">subject_file    =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                           <span class="dt">detailed_output =</span> <span class="ot">TRUE</span>)

dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/reexports.html">glimpse</a></span>(best_hit_tbl)</code></pre></div>
<pre><code>Variables:
$ query_id      (chr) "AT1G01010.1", "AT1G01020.1", "AT1G01030.1",...
$ subject_id    (chr) "333554|PACid:16033839", "470181|PACid:16064...
$ perc_identity (dbl) 73.99, 91.06, 95.54, 91.98, 100.00, 89.51, 9...
$ alig_length   (dbl) 469, 246, 359, 1970, 213, 648, 366, 300, 434...
$ mismatches    (dbl) 80, 22, 12, 85, 0, 58, 14, 22, 8, 34, 4, 6, ...
$ gap_openings  (dbl) 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, 1, 3, 2, 1...
$ q_start       (dbl) 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2,...
$ q_end         (dbl) 430, 246, 359, 1910, 213, 646, 366, 294, 429...
$ s_start       (dbl) 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4...
$ s_end         (dbl) 466, 246, 355, 1963, 213, 640, 362, 299, 433...
$ evalue        (dbl) 0e+00, 7e-166, 0e+00, 0e+00, 2e-160, 0e+00, ...
$ bit_score     (dbl) 627, 454, 698, 3704, 437, 1037, 696, 491, 85...
</code></pre>
</div>
<div id="the-blast_rec-function" class="section level2">
<h2 class="hasAnchor">
<a href="#the-blast_rec-function" class="anchor"></a>The blast_rec() function</h2>
<p>The <code><a href="../reference/blast_rec.html">blast_rec()</a></code> function was implemented to optimize BLAST+ reciprocal best hit searches (only based on the minimum e-value). BLAST+ reciprocal best hit searches are used to perform orthology inference.</p>
<p>Running <code><a href="../reference/blast_rec.html">blast_rec()</a></code> using default parameter settings:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># performing gene orthology inference using the reciprocal best hit (RBH) method</span>
<span class="kw"><a href="../reference/blast_rec.html">blast_rec</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))</code></pre></div>
<pre><code>      query_id            subject_id evalue
1  AT1G01010.1 333554|PACid:16033839  0e+00
2  AT1G01020.1 470181|PACid:16064328 7e-166
3  AT1G01030.1 470180|PACid:16054974  0e+00
4  AT1G01040.1 333551|PACid:16057793  0e+00
5  AT1G01050.1 909874|PACid:16064489 2e-160
6  AT1G01060.3 470177|PACid:16043374  0e+00
7  AT1G01070.1 918864|PACid:16052578  0e+00
8  AT1G01080.1 909871|PACid:16053217 1e-178
9  AT1G01090.1 470171|PACid:16052860  0e+00
10 AT1G01110.2 333544|PACid:16034284  0e+00
11 AT1G01120.1 918858|PACid:16049140  0e+00
12 AT1G01140.3 470161|PACid:16036015  0e+00
13 AT1G01150.1 918855|PACid:16037307 3e-150
14 AT1G01160.1 918854|PACid:16044153  1e-93
15 AT1G01170.2 311317|PACid:16052302  3e-54
16 AT1G01180.1 909860|PACid:16056125  0e+00
17 AT1G01190.1 311315|PACid:16059488  0e+00
18 AT1G01200.1 470156|PACid:16041002 3e-172
19 AT1G01210.1 311313|PACid:16057125  7e-76
20 AT1G01220.1 470155|PACid:16047984  0e+00
</code></pre>
<p>Again you can specify the <code>detailed_output</code> argument to get more parameters returned by <code><a href="../reference/blast_rec.html">blast_rec()</a></code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># running blast_rec() using detailed_output = TRUE</span>
rbh &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast_rec.html">blast_rec</a></span>(<span class="dt">query_file      =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file    =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">detailed_output =</span> <span class="ot">TRUE</span>)


dplyr<span class="op">::</span><span class="kw"><a href="http://dplyr.tidyverse.org/reference/reexports.html">glimpse</a></span>(rbh)</code></pre></div>
<pre><code>Variables:
$ query_id      (chr) "AT1G01010.1", "AT1G01020.1", "AT1G01030.1",...
$ subject_id    (chr) "333554|PACid:16033839", "470181|PACid:16064...
$ perc_identity (dbl) 73.99, 91.06, 95.54, 91.98, 100.00, 89.51, 9...
$ alig_length   (dbl) 469, 246, 359, 1970, 213, 648, 366, 300, 434...
$ mismatches    (dbl) 80, 22, 12, 85, 0, 58, 14, 22, 8, 34, 4, 6, ...
$ gap_openings  (dbl) 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, 1, 3, 2, 1...
$ q_start       (dbl) 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2,...
$ q_end         (dbl) 430, 246, 359, 1910, 213, 646, 366, 294, 429...
$ s_start       (dbl) 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4...
$ s_end         (dbl) 466, 246, 355, 1963, 213, 640, 362, 299, 433...
$ evalue        (dbl) 0e+00, 7e-166, 0e+00, 0e+00, 2e-160, 0e+00, ...
$ bit_score     (dbl) 627, 454, 698, 3704, 437, 1037, 696, 491, 85...
</code></pre>
<div id="the-set_blast-function" class="section level3">
<h3 class="hasAnchor">
<a href="#the-set_blast-function" class="anchor"></a>The set_blast() function</h3>
<p>The <code><a href="../reference/set_blast.html">set_blast()</a></code>function reads a file storing a specific sequence type, such as “cds”, “protein”, or “dna” in a standard sequence file format such as “fasta”, etc. and depending of the makedb parameter either creates a blast-able database, or returns the corresponding protein sequences as data.table object for further BLAST searches.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="kw"><a href="../reference/set_blast.html">set_blast</a></span>(<span class="dt">file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))[[<span class="dv">1</span>]] , <span class="dv">2</span>)</code></pre></div>
<pre><code>
       geneids
1: AT1G01010.1
2: AT1G01020.1

seqs
1: atggaggatcaagttgggtttgggttccgtccgaacgacgaggagctcgttggtcactatctccgtaacaaaatcgaaggaaacact
agccgcgacgttgaagtagccatcagcgaggtcaacatctgtagctacgatccttggaacttgcgcttccagtcaaagtacaaatcgaga
gatgctatgtggtacttcttctctcgtagagaaaacaacaaagggaatcgacagagcaggacaacggtttctggtaaatggaagcttacc
ggagaatctgttgaggtcaaggaccagtggggattttgtagtgagggctttcgtggtaagattggtcataaaagggttttggtgttcctc
gatggaagataccctgacaaaaccaaatctgattgggttatccacgagttccactacgacctcttaccagaacatcagaggacatatgtc
atctgcagacttgagtacaagggtgatgatgcggacattctatctgcttatgcaatagatcccactcccgcttttgtccccaatatgact
agtagtgcaggttctgtggtcaaccaatcacgtcaacgaaattcaggatcttacaacacttactctgagtatgattcagcaaatcatggc
cagcagtttaatgaaaactctaacattatgcagcagcaaccacttcaaggatcattcaaccctctccttgagtatgattttgcaaatcac
ggcggtcagtggctgagtgactatatcgacctgcaacagcaagttccttacttggcaccttatgaaaatgagtcggagatgatttggaag
catgtgattgaagaaaattttgagtttttggtagatgaaaggacatctatgcaacagcattacagtgatcaccggcccaaaaaacctgtg
tctggggttttgcctgatgatagcagtgatactgaaactggatcaatgattttcgaagacacttcgagctccactgatagtgttggtagt
tcagatgaaccgggccatactcgtatagatgatattccatcattgaacattattgagcctttgcacaattataaggcacaagagcaacca
aagcagcagagcaaagaaaaggtgataagttcgcagaaaagcgaatgcgagtggaaaatggctgaagactcgatcaagatacctccatcc
accaacacggtgaagcagagctggattgttttggagaatgcacagtggaactatctcaagaacatgatcattggtgtcttgttgttcatc
tccgtcattagttggatcattcttgttggttaa
2:
atggcggcgagtgaacacagatgcgtgggatgtggttttagggtaaagtcattgttcattcaatactctccgggtaacattcgtctcatg
aaatgcggaaattgcaaggaagtagcagatgagtacatcgagtgtgaacgcatgattattttcatcgatttaatccttcacagaccaaag
gtatatagacacgtcctctacaatgcaattaatccagcaactgtcaatattcagcatctgttgtggaagttggtcttcgcctatcttctt
ctagactgttatagaagcttgctactgagaaaaagtgatgaagaatcgagcttttctgatagccctgttcttctatctataaaggttctg
attggtgtcttatctgcaaacgctgcatttatcatctcttttgccattgcgactaagggtttgctaaatgaagtttccagaagaagagag
attatgttggggatattcatctctagttacttcaagatatttctgcttgcgatgttggtatgggaattcccaatgtcagtgatttttttt
gtcgatatacttctcttaacatcaaactccatggctcttaaagtgatgactgaatcaacaatgaccagatgcatagccgtatgcttaatc
gcgcacttgattagattcttggtgggtcagatttttgagccgacaatatttttgatacaaattggatctctgttgcaatatatgtcttat
tttttcagaatcgtatga

aa
1: MEDQVGFGFRPNDEELVGHYLRNKIEGNTSRDVEVAISEVNICSYDPWNLRFQSKYKSRDAMWYFFSRRENNKGNRQSRTTVSGKWK
LTGESVEVKDQWGFCSEGFRGKIGHKRVLVFLDGRYPDKTKSDWVIHEFHYDLLPEHQRTYVICRLEYKGDDADILSAYAIDPTPAFVPN
MTSSAGSVVNQSRQRNSGSYNTYSEYDSANHGQQFNENSNIMQQQPLQGSFNPLLEYDFANHGGQWLSDYIDLQQQVPYLAPYENESEMI
WKHVIEENFEFLVDERTSMQQHYSDHRPKKPVSGVLPDDSSDTETGSMIFEDTSSSTDSVGSSDEPGHTRIDDIPSLNIIEPLHNYKAQE
QPKQQSKEKVISSQKSECEWKMAEDSIKIPPSTNTVKQSWIVLENAQWNYLKNMIIGVLLFISVISWIILVG*
2:
MAASEHRCVGCGFRVKSLFIQYSPGNIRLMKCGNCKEVADEYIECERMIIFIDLILHRPKVYRHVLYNAINPATVNIQHLLWKLVFAYLL
LDCYRSLLLRKSDEESSFSDSPVLLSIKVLIGVLSANAAFIISFAIATKGLLNEVSRRREIMLGIFISSYFKIFLLAMLVWEFPMSVIFF
VDILLLTSNSMALKVMTESTMTRCIAVCLIAHLIRFLVGQIFEPTIFLIQIGSLLQYMSYFFRIV*</code></pre>
</div>
<div id="the-advanced_makedb-function" class="section level3">
<h3 class="hasAnchor">
<a href="#the-advanced_makedb-function" class="anchor"></a>The advanced_makedb() function</h3>
<p>The <code>advanced_makedb</code> function provides a simple, but powerful interface between the R language and <code>makeblastdb</code>. You can specify the <code>params</code> argument to pass all parameters defined for <code>makeblastdb</code> to the corresponding <code>makeblastdb</code> call.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># make the A. thaliana genome to a blast-able database</span>
<span class="kw"><a href="../reference/advanced_makedb.html">advanced_makedb</a></span>( <span class="dt">database_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_aa.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">params        =</span> <span class="st">"-input_type fasta -dbtype prot -hash_index"</span> )</code></pre></div>
<pre><code>Building a new DB, current time: 11/10/2014 16:56:58
New DB name:   _blast_db/ortho_thal_aa.fasta
New DB title:  _blast_db/ortho_thal_aa.fasta
Sequence type: Protein
Keep Linkouts: T
Keep MBits: T
Maximum file size: 1000000000B
Adding sequences from FASTA; added 20 sequences in 0.0583858 seconds.
</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#getting-started">Getting Started</a></li>
      <li><a href="#performing-blast-searches">Performing BLAST Searches</a></li>
      <li><a href="#the-blast-function">The blast() function</a></li>
      <li><a href="#the-blast_best-function">The blast_best() function</a></li>
      <li><a href="#the-blast_rec-function">The blast_rec() function</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.2.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
