<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform an advanced BLAST+ search — advanced_blast • orthologr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Perform an advanced BLAST+ search — advanced_blast" />

<meta property="og:description" content="This function performs a BLAST+ search of a given set of sequences and a given set of parameters against a BLAST database." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">dNdS estimation using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform an advanced BLAST+ search</h1>
    <small>Source: <a href='https://github.com/HajkD/orthologr/blob/master/R/advanced_blast.R'><code>R/advanced_blast.R</code></a></small>
    <div class="hidden name"><code>advanced_blast.Rd</code></div>
    </div>

    
    <p>This function performs a BLAST+ search of a given set of sequences and a given set of parameters against a BLAST database.</p>
    

    <pre class="usage"><span class='fu'>advanced_blast</span>(<span class='no'>query_file</span>, <span class='no'>subject_file</span>, <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"cds"</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"fasta"</span>,
  <span class='kw'>blast_algorithm</span> <span class='kw'>=</span> <span class='st'>"blastp"</span>, <span class='kw'>path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>blast_params</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>makedb_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>, <span class='kw'>taxonomy</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>db_path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>sql_database</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>session_name</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>write.only</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>clean_folders</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>query_file</th>
      <td><p>a character string specifying the path to the sequence file of interest (query organism).</p></td>
    </tr>
    <tr>
      <th>subject_file</th>
      <td><p>a character string specifying the path to the sequence file of interest (subject organism).
In case a NCBI database is used, it is possible to specify <code>subject_file</code> = <code>"nr"</code>, <code>"plaza"</code>, <code>"cdd_delta"</code>. Other databases
will be added in future versions of <span class="pkg">orthologr</span>. When specifying <code>subject_file</code> = "nr", "plaza", ...
make sure the corresponding database ("nr", "plaza","cdd_delta" ...) is stored in the blast working directory <code>"_blast_db"</code>.</p></td>
    </tr>
    <tr>
      <th>seq_type</th>
      <td><p>a character string specifying the sequence type stored in the input file.
Options are are: <code>seq_type</code> = <code>"cds"</code>, <code>seq_type</code> = <code>"protein"</code>, or <code>seq_type</code> = <code>"dna"</code>.
In case of <code>seq_type</code> = <code>"cds"</code>, sequence are translated to protein sequences,
in case of <code>seq_type</code> = <code>"dna"</code>, cds prediction is performed on the corresponding sequences which subsequently are
translated to protein sequences. Default is <code>seq_type</code> = <code>"cds"</code>.</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p>a character string specifying the file format of the sequence file, e.g. <code>format</code> = <code>"fasta"</code>, <code>format</code> = <code>"gbk"</code>.</p></td>
    </tr>
    <tr>
      <th>blast_algorithm</th>
      <td><p>a character string specifying the BLAST algorithm that shall be used, 
e.g. <code>blast_algorithm</code> = <code>"blastp"</code>, <code>blast_algorithm</code> = <code>"blastn"</code>, <code>blast_algorithm</code> = <code>"tblastn"</code>,
<code>blast_algorithm</code> = <code>"deltablast"</code>.</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>a character string specifying the path to the BLAST program (in case you don't use the default path).</p></td>
    </tr>
    <tr>
      <th>blast_params</th>
      <td><p>a character string listing the input paramters that shall be passed to the executing BLAST program. Default is <code>NULL</code>, implicating
that a set of default parameters is used when running BLAST.</p></td>
    </tr>
    <tr>
      <th>makedb_type</th>
      <td><p>a character string specifying the sequence type stored in the BLAST database
that is generated using 'makeblastdb'. Options are: "protein" and "nucleotide". Default is <code>makedb_type</code> = <code>"protein"</code>.</p></td>
    </tr>
    <tr>
      <th>taxonomy</th>
      <td><p>a logical value specifying whether the subject taxonomy shall be returned based on NCBI Taxonomy queries. 
Default is <code>taxonomy</code> = <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>db_path</th>
      <td><p>a character string specidying the path to the local BLAST database.</p></td>
    </tr>
    <tr>
      <th>sql_database</th>
      <td><p>a logical value specifying whether an SQL database shall be created to store the BLAST output.
This is only useful when the amount of data does not fit in-memory anymore.</p></td>
    </tr>
    <tr>
      <th>session_name</th>
      <td><p>a character string specifying the name of the BLAST output. Default is <code>session_name</code> = <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>write.only</th>
      <td><p>a logical value specifying whether the resulting blast output should be returned as
data.table or database connection. Default is <code>write.only</code> = <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>clean_folders</th>
      <td><p>a boolean value spefiying whether all internall folders storing the output of used programs
shall be removed. Default is <code>clean_folders</code> = <code>FALSE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>the <code>advanced_blast</code> function creates a folder named '_blast_db' and stores
the hit table returned by BLAST in this folder.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The <code>advanced_blast</code> function allows you to use the BLAST+ stand-alone specific notation
within the R environment. The <code>blast_params</code> argument allows the user to specify BLAST+ specific command line
parameters and the resulting hit table is then returned as <span class="pkg">data.table</span> object.</p>
<p>Optionally the <code>sql_database = TRUE</code> argument allows you to store the hit table in an SQLite database
using the flexible query notation provided by the <span class="pkg">dplyr</span> (&gt;= 0.3) package.</p>
<p>The <code>advanced_blast</code> function works very similar to the command line version of BLAST+,
so when specifying a <code>blast_algorithm</code> such as <em>blastp</em>, <em>tblastn</em>, <em>deltablast</em>, etc.
the corresponding databases must be stored and accessible accoring to the BLAST stand-alone notation.</p>
    <p>Following BLAST programs and algorithms can be assigned to <code>blast_algorithm</code>:</p>
<ul>
<li><p>"blastp"</p></li>
<li><p>"blastn"</p></li>
<li><p>"megablast"</p></li>
<li><p>"psiblast"</p></li>
<li><p>"phiblast"</p></li>
<li><p>"deltablast"</p></li>
<li><p>"blastx"</p></li>
<li><p>"tblastn"</p></li>
<li><p>"tblastx"</p></li>
</ul>
    <p>The intention of this function is to provide the user with an easy to use interface function
to BLAST+.</p>
<p>When using <code>taxonomy</code> =  <code>TRUE</code>, then the taxdb.btd and taxdb.bti must be stored
in the same folder as BLAST-able database. In case you need to specify the path to your BLAST database,
use the <code>db_path</code> argument.</p>
<p>The <code>advanced_blast</code> function can also store the BLAST output CSV file in a SQLite database.
This works only with dplyr version &gt;= 0.3 . To store the BLAST output in an SQLite database and
to receive an SQLite connection as specified by <code>tbl</code> in <code>dplyr</code> please use the <code>sql_database</code> = <code>TRUE</code> argument.</p>
<p>In case you want to use deltablast, you need to all cdd_delta.* files into the folder "_blast_db".</p>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>This function is also able to return the subject taxonomy.</p>
<p>For this purpose the NCBI taxdb must be installed on your system:</p>
<p>ftp://ftp.ncbi.nlm.nih.gov/blast/db/taxdb.tar.gz</p>
<p>In case you want to use the taxonomy option, set <code>taxonomy</code> = <code>TRUE</code>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>BLAST Command Line Applications User Manual: <a href='http://www.ncbi.nlm.nih.gov/books/NBK1763/'>http://www.ncbi.nlm.nih.gov/books/NBK1763/</a></p>
<p>Blast Program Selection Guide: <a href='http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=ProgSelectionGuide'>http://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=ProgSelectionGuide</a></p>
<p>BLAST web interface: <a href='http://blast.ncbi.nlm.nih.gov/Blast.cgi'>http://blast.ncbi.nlm.nih.gov/Blast.cgi</a></p>
<p>Altschul, S.F., Gish, W., Miller, W., Myers, E.W. and Lipman, D.J. (1990) "Basic local alignment search tool." J. Mol. Biol. 215:403-410.
Gish, W. and States, D.J. (1993) "Identification of protein coding regions by database similarity search." Nature Genet. 3:266-272.</p>
<p>Madden, T.L., Tatusov, R.L., and Zhang, J. (1996) "Applications of network BLAST server" Meth. Enzymol. 266:131-141.</p>
<p>Altschul, S.F., Madden, T.L., Schaeffer, A.A., Zhang, J., Zhang, Z., Miller, W., and Lipman, D.J. (1997) "Gapped BLAST and PSI-BLAST: a new generation of protein database search programs." Nucleic Acids Res. 25:3389-3402.</p>
<p>Zhang Z., Schwartz S., Wagner L., and Miller W. (2000), "A greedy algorithm for aligning DNA sequences" J Comput Biol 2000; 7(1-2):203-14.</p>
<p>Zhang, J. and Madden, T.L. (1997) "PowerBLAST: A new network BLAST application for interactive or automated sequence analysis and annotation." Genome Res. 7:649-656.</p>
<p>Morgulis A., Coulouris G., Raytselis Y., Madden T.L., Agarwala R., and Schaeffer A.A. (2008) "Database indexing for production MegaBLAST searches." Bioinformatics 15:1757-1764.</p>
<p>Camacho C., Coulouris G., Avagyan V., Ma N., Papadopoulos J., Bealer K., and Madden T.L. (2008) "BLAST+: architecture and applications." BMC Bioinformatics 10:421.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='blast_best.html'>blast_best</a></code>, <code><a href='blast_rec.html'>blast_rec</a></code>, <code><a href='blast.html'>blast</a></code>, <code><a href='set_blast.html'>set_blast</a></code>, <code><a href='advanced_makedb.html'>advanced_makedb</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>

<span class='co'># performing a BLAST search using blastp (default)</span>
<span class='fu'>advanced_blast</span>(
  <span class='kw'>query_file</span>      <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
  <span class='kw'>subject_file</span>    <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_lyra_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
  <span class='kw'>blast_algorithm</span> <span class='kw'>=</span> <span class='st'>"blastp"</span>,
  <span class='kw'>blast_params</span>    <span class='kw'>=</span> <span class='st'>"-evalue 1E-5 -num_threads 2"</span> )

<span class='co'># performing a BLAST search using blastp (default) and starting with protein sequences</span>
<span class='fu'>advanced_blast</span>(
  <span class='kw'>query_file</span>      <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
  <span class='kw'>subject_file</span>    <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
  <span class='kw'>seq_type</span>        <span class='kw'>=</span> <span class='st'>"protein"</span>,
  <span class='kw'>blast_algorithm</span> <span class='kw'>=</span> <span class='st'>"blastp"</span>,
  <span class='kw'>blast_params</span>    <span class='kw'>=</span> <span class='st'>"-evalue 1E-5 -num_threads 2"</span> )



<span class='co'># DELTA-BLAST</span>
<span class='co'>#</span>
<span class='co'># you can also use deltablast to perform BLAST searches</span>
<span class='co'># make sure you have the cdd_deltablast.* files stored in the folder "_blast_db"     </span>
<span class='fu'>advanced_blast</span>(
  <span class='kw'>query_file</span>      <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
  <span class='kw'>subject_file</span>    <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_lyra_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
  <span class='kw'>blast_algorithm</span> <span class='kw'>=</span> <span class='st'>"deltablast"</span>,
  <span class='kw'>db_path</span>         <span class='kw'>=</span> <span class='st'>"path/to/cdd_files"</span>,
  <span class='kw'>blast_params</span>    <span class='kw'>=</span> <span class='st'>"-evalue 1E-5 -num_threads 2"</span>)



<span class='co'># when performing an advanced BLAST search, you can easily select the best hit using</span>
<span class='fu'>library</span>(<span class='no'>dplyr</span>)

<span class='no'>advB</span> <span class='kw'>&lt;-</span> <span class='fu'>advanced_blast</span>(
   <span class='kw'>query_file</span>   <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
   <span class='kw'>subject_file</span> <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
   <span class='kw'>seq_type</span>     <span class='kw'>=</span> <span class='st'>"cds"</span>,<span class='kw'>blast_algorithm</span> <span class='kw'>=</span> <span class='st'>"blastp"</span>,
   <span class='kw'>blast_params</span> <span class='kw'>=</span> <span class='st'>"-evalue 1E-5 -num_threads 1"</span>)

<span class='no'>best_hit</span> <span class='kw'>&lt;-</span> <span class='no'>advB</span> <span class='kw'>%&gt;%</span> <span class='fu'><a href='http://dplyr.tidyverse.org/reference/group_by.html'>group_by</a></span>(<span class='no'>query_id</span>) <span class='kw'>%&gt;%</span> <span class='fu'><a href='http://dplyr.tidyverse.org/reference/summarise.html'>summarise</a></span>(<span class='fu'>min</span>(<span class='no'>evalue</span>))


<span class='co'># using a SQLite database to store the BLAST output and</span>
<span class='co'># select some example data</span>

<span class='fu'>library</span>(<span class='no'>dplyr</span>)

<span class='no'>sqlE</span> <span class='kw'>&lt;-</span> <span class='fu'>advanced_blast</span>(
    <span class='kw'>query_file</span>      <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
    <span class='kw'>subject_file</span>    <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
    <span class='kw'>seq_type</span>        <span class='kw'>=</span> <span class='st'>"cds"</span>,
    <span class='kw'>blast_algorithm</span> <span class='kw'>=</span> <span class='st'>"blastp"</span>,
    <span class='kw'>blast_params</span>    <span class='kw'>=</span> <span class='st'>"-evalue 1E-5 -num_threads 1"</span>,
    <span class='kw'>sql_database</span>    <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='fu'>head</span>(<span class='no'>sqlE</span>)

<span class='fu'><a href='http://dplyr.tidyverse.org/reference/reexports.html'>glimpse</a></span>(<span class='no'>sqlE</span>)

<span class='co'># select all rows that have an evalue &gt; 1e-15</span>
<span class='fu'><a href='http://dplyr.tidyverse.org/reference/filter.html'>filter</a></span>(<span class='no'>sqlE</span>,<span class='no'>evalue</span> <span class='kw'>&gt;</span> <span class='fl'>1e-15</span>)

<span class='co'># select the best hit using the evalue criterion</span>
<span class='no'>sqlE</span> <span class='kw'>%&gt;%</span> <span class='fu'><a href='http://dplyr.tidyverse.org/reference/group_by.html'>group_by</a></span>(<span class='no'>query_id</span>) <span class='kw'>%&gt;%</span> <span class='fu'><a href='http://dplyr.tidyverse.org/reference/summarise.html'>summarise</a></span>(<span class='kw'>best_hit_eval</span> <span class='kw'>=</span> <span class='fu'>min</span>(<span class='no'>evalue</span>))


<span class='co'># using advanced_blast() with external BLAST path                                           </span>
<span class='fu'>advanced_blast</span>(
 <span class='kw'>query_file</span>      <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
 <span class='kw'>subject_file</span>    <span class='kw'>=</span> <span class='fu'>system.file</span>(<span class='st'>'seqs/ortho_lyra_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
 <span class='kw'>blast_algorithm</span> <span class='kw'>=</span> <span class='st'>"blastp"</span>,
 <span class='kw'>blast_params</span>    <span class='kw'>=</span> <span class='st'>"-evalue 1E-5 -num_threads 2"</span>,
 <span class='kw'>path</span>            <span class='kw'>=</span> <span class='st'>"path/to/blastp/"</span>)

<span class='co'># }</span><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Hajk-Georg Drost

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

