<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Perform a BLAST+ best hit search — blast_best • orthologr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Perform a BLAST+ best hit search — blast_best" />

<meta property="og:description" content="This function performs a BLAST+ search (best hit) of a given set of protein sequences against a given database." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">dNdS estimation using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Perform a BLAST+ best hit search</h1>
    <small class="dont-index">Source: <a href='https://github.com/HajkD/orthologr/blob/master/R/blast_best.R'><code>R/blast_best.R</code></a></small>
    <div class="hidden name"><code>blast_best.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function performs a BLAST+ search (best hit) of a given set of protein sequences against a given database.</p>
    
    </div>

    <pre class="usage"><span class='fu'>blast_best</span>(<span class='no'>query_file</span>, <span class='no'>subject_file</span>, <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"cds"</span>,
  <span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"fasta"</span>, <span class='kw'>blast_algorithm</span> <span class='kw'>=</span> <span class='st'>"blastp"</span>,
  <span class='kw'>delete_corrupt_cds</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>eval</span> <span class='kw'>=</span> <span class='st'>"1E-5"</span>, <span class='kw'>max.target.seqs</span> <span class='kw'>=</span> <span class='fl'>10000</span>,
  <span class='kw'>path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>comp_cores</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>blast_params</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>clean_folders</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>save.output</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>query_file</th>
      <td><p>a character string specifying the path to the CDS file of interest (query organism).</p></td>
    </tr>
    <tr>
      <th>subject_file</th>
      <td><p>a character string specifying the path to the CDS file of interest (subject organism).</p></td>
    </tr>
    <tr>
      <th>seq_type</th>
      <td><p>a character string specifying the sequence type stored in the input file.
Options are are: "cds", "protein", or "dna". In case of "cds", sequence are translated to protein sequences,
in case of "dna", cds prediction is performed on the corresponding sequences which subsequently are
translated to protein sequences. Default is <code>seq_type</code> = "cds".</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p>a character string specifying the file format of the sequence file, e.g. <code>format</code> = <code>"fasta"</code>.
Default is <code>format</code> = <code>"fasta"</code>.</p></td>
    </tr>
    <tr>
      <th>blast_algorithm</th>
      <td><p>a character string specifying the BLAST algorithm that shall be used, e.g. 
<code>blast_algorithm</code> = <code>"blastp"</code>, <code>blast_algorithm</code> = <code>"blastn"</code>, <code>blast_algorithm</code> = <code>"tblastn"</code> .</p></td>
    </tr>
    <tr>
      <th>delete_corrupt_cds</th>
      <td><p>a logical value indicating whether sequences with corrupt base triplets should be removed from the input <code>file</code>. This is the case when the length of coding sequences cannot be divided by 3 and thus the coding sequence contains at least one corrupt base triplet.</p></td>
    </tr>
    <tr>
      <th>eval</th>
      <td><p>a numeric value specifying the E-Value cutoff for BLAST hit detection.</p></td>
    </tr>
    <tr>
      <th>max.target.seqs</th>
      <td><p>a numeric value specifying the number of aligned sequences to keep.
Please be aware that <code>max.target.seqs</code> selects best hits based on the database entry and not by the best e-value. See details here: https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty833/5106166 .</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>a character string specifying the path to the BLAST program (in case you don't use the default path).</p></td>
    </tr>
    <tr>
      <th>comp_cores</th>
      <td><p>a numeric value specifying the number of cores to be used for multicore BLAST computations.</p></td>
    </tr>
    <tr>
      <th>blast_params</th>
      <td><p>a character string listing the input paramters that shall be passed to the executing BLAST program. Default is <code>NULL</code>, implicating
that a set of default parameters is used when running BLAST.</p></td>
    </tr>
    <tr>
      <th>clean_folders</th>
      <td><p>a boolean value spefiying whether all internall folders storing the output of used programs
shall be removed. Default is <code>clean_folders</code> = <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>save.output</th>
      <td><p>a path to the location were the BLAST output shall be stored. E.g. <code>save.output</code> = <code><a href='https://www.rdocumentation.org/packages/base/topics/getwd'>getwd()</a></code>
to store it in the current working directory, or <code>save.output</code> = <code><a href='https://www.rdocumentation.org/packages/base/topics/file.path'>file.path(put,your,path,here)</a></code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data.table as returned by the <code>blast</code> function, storing the geneids
of orthologous genes (best hit) in the first column and the amino acid sequences in the second column.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Given a set of protein sequences (query sequences), a best hit blast search (BH BLAST) is being performed.</p>
<p>Internally to perform best hit searches, the BLAST+ parameter settings:</p>
<p><code>"-best_hit_score_edge 0.05 -best_hit_overhang 0.25 -max_target_seqs 1"</code></p>
<p>are used to speed up best hit computations.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Altschul, S.F., Gish, W., Miller, W., Myers, E.W. &amp; Lipman, D.J. (1990) "Basic local alignment search tool." J. Mol. Biol. 215:403-410.</p>
<p>Gish, W. &amp; States, D.J. (1993) "Identification of protein coding regions by database similarity search." Nature Genet. 3:266-272.</p>
<p>Madden, T.L., Tatusov, R.L. &amp; Zhang, J. (1996) "Applications of network BLAST server" Meth. Enzymol. 266:131-141.</p>
<p>Altschul, S.F., Madden, T.L., Schaeffer, A.A., Zhang, J., Zhang, Z., Miller, W. &amp; Lipman, D.J. (1997) "Gapped BLAST and PSI-BLAST: a new generation of protein database search programs." Nucleic Acids Res. 25:3389-3402.</p>
<p>Zhang Z., Schwartz S., Wagner L., &amp; Miller W. (2000), "A greedy algorithm for aligning DNA sequences" J Comput Biol 2000; 7(1-2):203-14.</p>
<p>Zhang, J. &amp; Madden, T.L. (1997) "PowerBLAST: A new network BLAST application for interactive or automated sequence analysis and annotation." Genome Res. 7:649-656.</p>
<p>Morgulis A., Coulouris G., Raytselis Y., Madden T.L., Agarwala R., &amp; Schaeffer A.A. (2008) "Database indexing for production MegaBLAST searches." Bioinformatics 15:1757-1764.</p>
<p>Camacho C., Coulouris G., Avagyan V., Ma N., Papadopoulos J., Bealer K., &amp; Madden T.L. (2008) "BLAST+: architecture and applications." BMC Bioinformatics 10:421.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='blast.html'>blast</a></code>, <code><a href='blast_rec.html'>blast_rec</a></code>, <code><a href='advanced_blast.html'>advanced_blast</a></code>, <code><a href='set_blast.html'>set_blast</a></code>, <code><a href='advanced_makedb.html'>advanced_makedb</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'># performing gene orthology inference using the best hit (BH) method</span>
<span class='fu'>blast_best</span>(<span class='kw'>query_file</span>   <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
           <span class='kw'>subject_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>))



<span class='co'># performing gene orthology inference using the best hit (BH) method starting with protein sequences</span>
<span class='fu'>blast_best</span>(<span class='kw'>query_file</span>   <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
           <span class='kw'>subject_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
           <span class='kw'>seq_type</span>     <span class='kw'>=</span> <span class='st'>"protein"</span>)



<span class='co'># save the BLAST output file to the current working directory</span>
<span class='fu'>blast_best</span>(<span class='kw'>query_file</span>   <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
           <span class='kw'>subject_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
           <span class='kw'>seq_type</span>     <span class='kw'>=</span> <span class='st'>"protein"</span>,
           <span class='kw'>save.output</span>  <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/getwd'>getwd</a></span>())





<span class='co'># use multicore processing</span>
<span class='fu'>blast_best</span>(<span class='kw'>query_file</span>   <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
           <span class='kw'>subject_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
           <span class='kw'>comp_cores</span>   <span class='kw'>=</span> <span class='fl'>2</span>)



<span class='co'># performing gene orthology inference using the best hit (BH) method and external</span>
<span class='co'># blastp path</span>
<span class='fu'>blast_best</span>(<span class='kw'>query_file</span>   <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
           <span class='kw'>subject_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
           <span class='kw'>path</span>         <span class='kw'>=</span> <span class='st'>"path/to/blastp/"</span>)


<span class='co'># }</span><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Hajk-Georg Drost</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

