<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compute dNdS Values For A Given Pairwise Alignment — compute_dnds • orthologr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
<meta property="og:title" content="Compute dNdS Values For A Given Pairwise Alignment — compute_dnds" />

<meta property="og:description" content="This function takes a vector containing the
query amino acid sequence, subject amino acid sequence, query CDS sequence, and subject CDS sequence
and then runs the following pipieline:

1) Multiple-Alignment of query amino acid sequence and subject amino acid sequence
2) Codon-Alignment of the amino acid alignment returned by 1) and query CDS sequence + subject CDS sequence
3) dNdS estimation of the codon alignment returned by 2)
" />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->




  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">dNdS estimation using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute dNdS Values For A Given Pairwise Alignment</h1>
    <small>Source: <a href='https://github.com/HajkD/orthologr/blob/master/R/compute_dnds.R'><code>R/compute_dnds.R</code></a></small>
    <div class="hidden name"><code>compute_dnds.Rd</code></div>
    </div>

    
    <p>This function takes a vector containing the
query amino acid sequence, subject amino acid sequence, query CDS sequence, and subject CDS sequence
and then runs the following pipieline:</p>
<ul>
<li><p>1) Multiple-Alignment of query amino acid sequence and subject amino acid sequence</p></li>
<li><p>2) Codon-Alignment of the amino acid alignment returned by 1) and query CDS sequence + subject CDS sequence</p></li>
<li><p>3) dNdS estimation of the codon alignment returned by 2)</p></li>
</ul>
    

    <pre class="usage"><span class='fu'>compute_dnds</span>(<span class='no'>complete_tbl</span>, <span class='kw'>aa_aln_type</span> <span class='kw'>=</span> <span class='st'>"multiple"</span>,
  <span class='kw'>aa_aln_tool</span> <span class='kw'>=</span> <span class='st'>"clustalw"</span>, <span class='kw'>aa_aln_path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>aa_aln_params</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>codon_aln_tool</span> <span class='kw'>=</span> <span class='st'>"pal2nal"</span>, <span class='kw'>dnds_est.method</span> <span class='kw'>=</span> <span class='st'>"YN"</span>,
  <span class='kw'>kaks_calc_path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>comp_cores</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>clean_folders</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>complete_tbl</th>
      <td><p>a data.table object storing the query_id, subject_id, query_cds (sequence), 
subject_cds (sequence), query_aa (sequence), and subject_aa (sequence) of the organisms that shall be compared.</p></td>
    </tr>
    <tr>
      <th>aa_aln_type</th>
      <td><p>a character string specifying the amino acid alignement type: <code>aa_aln_type</code> = "multiple" or <code>aa_aln_type</code> = <code>"pairwise"</code>.
Default is <code>aa_aln_type</code> = "multiple".</p></td>
    </tr>
    <tr>
      <th>aa_aln_tool</th>
      <td><p>a character string specifying the multiple alignment tool that shall be used for pairwise protein alignments.</p></td>
    </tr>
    <tr>
      <th>aa_aln_path</th>
      <td><p>a character string specifying the path to the corresponding multiple alignment tool.</p></td>
    </tr>
    <tr>
      <th>aa_aln_params</th>
      <td><p>a character string specifying additional parameters that shall be passed to the multiple alignment system call.</p></td>
    </tr>
    <tr>
      <th>codon_aln_tool</th>
      <td><p>a character string specifying the codon alignment tool that shall be used for codon alignments. Default is <code>codon_aln_tool</code> = <code>"pal2nal"</code>.</p></td>
    </tr>
    <tr>
      <th>dnds_est.method</th>
      <td><p>a character string specifying the dNdS estimation method, e.g. "Comeron","Li", "YN", etc. See Details for all options.</p></td>
    </tr>
    <tr>
      <th>kaks_calc_path</th>
      <td><p>a character string specifying the execution path to KaKs_Calculator. Default is <code>kaks_calc_path</code> = <code>NULL</code>
(meaning that KaKs_Calculator is stored and executable in your default <code>PATH</code>).</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>a logical value specifying whether a successful interface call shall be printed out.</p></td>
    </tr>
    <tr>
      <th>comp_cores</th>
      <td><p>a numeric value specifying the number of cores that shall be used to perform parallel computations on a multicore machine.</p></td>
    </tr>
    <tr>
      <th>clean_folders</th>
      <td><p>a boolean value spefiying whether all internall folders storing the output of used programs
shall be removed. Default is <code>clean_folders</code> = <code>FALSE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function takes the amino acid and CDS sequences two orthologous genes
and writes the corresponding amino acid and CDS sequences as fasta file into
the internal folder environment. The resulting fasta files (two files) store the 
amino acid sequence of the query_id and subject_id (file one) and the CDS sequence of
the query_id and subject_id (file two). These fasta files are then used to pass through the following pipeline:</p>
<p>1) Multiple-Alignment or Pairwise-Alignment of query amino acid sequence and subject amino acid sequence</p>
<p>2) Codon-Alignment of the amino acid alignment returned by 1) and query CDS sequence + subject CDS sequence</p>
<p>3) dNdS estimation of the codon alignment returned by 2)</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p><a href='http://www.r-bloggers.com/the-wonders-of-foreach/'>http://www.r-bloggers.com/the-wonders-of-foreach/</a></p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='multi_aln.html'>multi_aln</a></code>, <code><a href='substitutionrate.html'>substitutionrate</a></code>, <code><a href='dNdS.html'>dNdS</a></code></p></div>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
          </ul>

    <h2>Author</h2>
    
Hajk-Georg Drost and Sarah Scharfenberg

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

