<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compute Multiple Sequence Alignments — multi_aln • orthologr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Compute Multiple Sequence Alignments — multi_aln" />

<meta property="og:description" content="This function takes a FASTA file containing DNA or amino acid sequences
that shall be aligned and computes a multiple alignment using a defined multiple alignment tool." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute Multiple Sequence Alignments</h1>
    <small class="dont-index">Source: <a href='https://github.com/HajkD/orthologr/blob/master/R/multi_aln.R'><code>R/multi_aln.R</code></a></small>
    <div class="hidden name"><code>multi_aln.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function takes a FASTA file containing DNA or amino acid sequences
that shall be aligned and computes a multiple alignment using a defined multiple alignment tool.</p>
    
    </div>

    <pre class="usage"><span class='fu'>multi_aln</span>(<span class='no'>file</span>, <span class='no'>tool</span>, <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>multi_aln_name</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>params</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>clean_folders</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>file</th>
      <td><p>a character string specifying the path to the file storing the sequences in FASTA format.</p></td>
    </tr>
    <tr>
      <th>tool</th>
      <td><p>a character string specifying the program that should be used: "clustalw", "t_coffee", "muscle", "clustalo", and "mafft".</p></td>
    </tr>
    <tr>
      <th>get_aln</th>
      <td><p>a logical value indicating whether the produced alignment should be returned.</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>a character string specifying the path to the multiple alignment program (in case you don't use the default path).</p></td>
    </tr>
    <tr>
      <th>multi_aln_name</th>
      <td><p>a character string specifying the name of the stored alignment file. 
Default is <code>multi_aln_name</code> = <code>NULL</code> denoting a default name: 'toolname.aln' .</p></td>
    </tr>
    <tr>
      <th>params</th>
      <td><p>a character string listing the input paramters that shall be passed to corresponding alignment tool specified in the <code>tool</code> argument. 
Default is <code>NULL</code>, implicating that a set of default parameters is used when running the correponding tool, e.g. clustalw. 
Example: <code>params</code> = "-PWMATRIX=BLOSUM -TYPE=PROTEIN" for <code>tool</code> = "clustalw.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>a logical value specifying whether the output of the corresponding multiple alignment tool shall be printed out to the console.
Default is <code>quiet</code> = <code>FALSE</code>.</p></td>
    </tr>
    <tr>
      <th>clean_folders</th>
      <td><p>a boolean value spefiying whether all internall folders storing the output of used programs
shall be removed. Default is <code>clean_folders</code> = <code>FALSE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>In case the argument <code>get_aln</code> is set <code>TRUE</code>, an object of class alignment of the seqinr package is returned.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function provides an interface between R and common multiple alignment programs
such as "clustalw", "t_coffee", "muscle", "clustalo", and "mafft".</p>
<ul>
<li><p>CLUSTALW :</p>
    <p>Different operating systems perform different execution calls to the clustalw program:</p>
<ul>
<li><p>MacOS: 'clustalw2'</p></li>
<li><p>Linux: 'clustalw'</p></li>
<li><p>Windows: 'clustalw2.exe'</p></li>
</ul>
    <p>In case you use the default path to the clustalw program, depending on your operating system,
the following calls to clustalw should work properly on your system:</p>
<ul>
<li><p>MacOS: <code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("clustalw2 -help")</a></code></p></li>
<li><p>Linux: <code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("clustalw -help")</a></code></p></li>
<li><p>Windows: <code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("clustalw2.exe -help")</a></code></p></li>
</ul>
    <p>In case these procedures don't work properly, please use the <code>path</code> argument
to specify the 'clustalw' execution path on your system:</p>
<ul>
<li><p>MacOS: <code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("path/to/clustalw/clustalw2 -help")</a></code></p></li>
<li><p>Linux: <code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("path/to/clustalw/clustalw -help")</a></code></p></li>
<li><p>Windows: <code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("path/to/clustalw/clustalw2.exe -help")</a></code></p></li>
</ul></li>
<li><p>T_COFFEE :</p>
<p>In case you use the default path to the t_coffee program,
the following calls to clustalw should work properly on your system:</p>
<p><code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("t_coffee -version")</a></code></p>
<p>In case this procedures doesn't work properly, please use the <code>path</code> argument
to specify the 't_coffee' execution path on your system:</p>
<p><code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("path/to/t_coffee/t_coffee -version")</a></code></p></li>
<li><p>MUSCLE :</p>
<p>In case you use the default path to the muscle program,
the following calls to muscle should work properly on your system:</p>
<p><code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("muscle -help")</a></code></p>
<p>In case this procedures doesn't work properly, please use the <code>path</code> argument
to specify the 'muscle' execution path on your system:</p>
<p><code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("path/to/muscle/muscle -help")</a></code></p></li>
<li><p>CLUSTALO :</p>
<p>In case you use the default path to the clustalo program,
the following calls to clustalo should work properly on your system:</p>
<p><code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("clustalo --help")</a></code></p>
<p>In case this procedures doesn't work properly, please use the <code>path</code> argument
to specify the 'clustalo' execution path on your system:</p>
<p><code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("path/to/clustalo/clustalo --help")</a></code></p></li>
<li><p>MAFFT :</p>
<p>In case you use the default path to the mafft program,
the following calls to mafft should work properly on your system:</p>
<p><code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("mafft -help")</a></code></p>
<p>In case this procedures doesn't work properly, please use the <code>path</code> argument
to specify the 'mafft' execution path on your system:</p>
<p><code><a href='https://www.rdocumentation.org/packages/base/topics/system'>system("path/to/mafft/mafft -help")</a></code></p></li>
</ul>
    
    <h2 class="hasAnchor" id="note"><a class="anchor" href="#note"></a>Note</h2>

    <p>Note that when using the <code>clustalw.params</code>, ... parameters, make sure the corresponding alignment tool
returns a file in clustal format *.aln. This is only important when <code>get_aln</code> = <code>TRUE</code>.</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <ul>
<li><p>CLUSTALW:</p>
<p>Larkin MA, Blackshields G, Brown NP, Chenna R, McGettigan PA, McWilliam H, Valentin F,
Wallace IM, Wilm A, Lopez R, Thompson JD, Gibson TJ, Higgins DG. (2007).
Clustal W and Clustal X version 2.0. Bioinformatics, 23, 2947-2948.</p>
<p><a href='http://www.clustal.org/clustal2/'>http://www.clustal.org/clustal2/</a></p>
<p><a href='http://www.ebi.ac.uk/Tools/msa/clustalw2/help/'>http://www.ebi.ac.uk/Tools/msa/clustalw2/help/</a></p></li>
<li><p>T_COFFEE</p>
<p>T-Coffee: A novel method for multiple sequence alignments. 
Notredame, Higgins, Heringa, JMB, 302(205-217). 2000.</p>
<p><a href='http://www.tcoffee.org/Projects/tcoffee/'>http://www.tcoffee.org/Projects/tcoffee/</a></p>
<p><a href='http://www.tcoffee.org/Projects/tcoffee/documentation/t_coffee_tutorial.pdf'>http://www.tcoffee.org/Projects/tcoffee/documentation/t_coffee_tutorial.pdf</a></p></li>
<li><p>MUSCLE:</p>
<p>Edgar, R.C. (2004) MUSCLE: multiple sequence alignment with high accuracy and high throughput. Nucleic Acids Res. 32(5):1792-1797.</p>
<p>Edgar, R.C. (2004) MUSCLE: a multiple sequence alignment method with reduced time and space complexity. BMC Bioinformatics, (5) 113.</p>
<p><a href='http://www.drive5.com/muscle/'>http://www.drive5.com/muscle/</a></p>
<p><a href='http://www.drive5.com/muscle/manual/'>http://www.drive5.com/muscle/manual/</a></p></li>
<li><p>CLUSTALO:</p>
<p>Sievers F, Wilm A, Dineen DG, Gibson TJ, Karplus K, Li W, Lopez R, McWilliam H, Remmert M, Soeding J, Thompson JD, Higgins DG (2011). 
Fast, scalable generation of high-quality protein multiple sequence alignments using Clustal Omega. Molecular Systems Biology 7:539 doi:10.1038/msb.2011.75</p>
<p><a href='http://www.clustal.org/omega/'>http://www.clustal.org/omega/</a></p>
<p><a href='http://www.clustal.org/omega/README'>http://www.clustal.org/omega/README</a></p></li>
<li><p>MAFFT :</p>
<p>Katoh, Standley 2013 (Molecular Biology and Evolution 30:772-780)  
MAFFT multiple sequence alignment software version 7: improvements in performance and usability.</p>
<p><a href='http://mafft.cbrc.jp/alignment/software/'>http://mafft.cbrc.jp/alignment/software/</a></p>
<p><a href='http://mafft.cbrc.jp/alignment/software/manual/manual.html'>http://mafft.cbrc.jp/alignment/software/manual/manual.html</a></p>
<p><a href='http://mafft.cbrc.jp/alignment/software/tips0.html'>http://mafft.cbrc.jp/alignment/software/tips0.html</a></p></li>
</ul>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'># CLUSTALW Example:</span>

<span class='co'># in case the default execution path of clustalw runs properly on your system</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"clustalw"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># in case the default execution path of clustalw is not set within the default path</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"clustalw"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
          <span class='kw'>path</span>    <span class='kw'>=</span> <span class='st'>"path/to/clustalw/"</span>)

<span class='co'># running clustalw using additional parameters</span>
<span class='co'># details: system("clustalw2 -help")</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"clustalw"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
          <span class='kw'>params</span>  <span class='kw'>=</span> <span class='st'>"-PWMATRIX=BLOSUM -TYPE=PROTEIN"</span>)


<span class='co'># T_COFFEE Example:</span>

<span class='co'># in case the default execution path of t_coffee runs properly on your system</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"t_coffee"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># in case the default execution path of t_coffee is not set within the default path</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"t_coffee"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
          <span class='kw'>path</span>    <span class='kw'>=</span> <span class='st'>"path/to/t_coffee/"</span>)

<span class='co'># running t_coffee using additional parameters</span>
<span class='co'># details: http://www.tcoffee.org/Projects/tcoffee/#DOCUMENTATION</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"t_coffee"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
          <span class='kw'>params</span>  <span class='kw'>=</span> <span class='st'>"-mode expresso"</span>)



<span class='co'># MUSCLE Example:  </span>

<span class='co'># in case the default execution path of muscle runs properly on your system</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"muscle"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># in case the default execution path of muscle is not set within the default path</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"muscle"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
          <span class='kw'>path</span>    <span class='kw'>=</span> <span class='st'>"path/to/muscle/"</span>)

<span class='co'># running muscle using additional parameters</span>
<span class='co'># details: http://www.drive5.com/muscle/manual/</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"muscle"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
          <span class='kw'>params</span>  <span class='kw'>=</span> <span class='st'>"-diags -clwstrict"</span>)


<span class='co'># CLUSTALO Example:  </span>

<span class='co'># in case the default execution path of clustalo runs properly on your system</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"clustalo"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># in case the default execution path of clustalo is not set within the default path</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"clustalo"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
          <span class='kw'>path</span>    <span class='kw'>=</span> <span class='st'>"path/to/clustalo/"</span>)

<span class='co'># running clustalo using additional parameters</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"clustalo"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
          <span class='kw'>params</span>  <span class='kw'>=</span> <span class='st'>"--outfmt clu"</span>)



<span class='co'># MAFFT Example:  </span>

<span class='co'># in case the default execution path of mafft runs properly on your system</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"mafft"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># in case the default execution path of mafft is not set within the default path</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"mafft"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
          <span class='kw'>path</span>    <span class='kw'>=</span> <span class='st'>"path/to/mafft/"</span>)

<span class='co'># running mafft using additional parameters</span>
<span class='co'># details: http://www.drive5.com/mafft/manual/</span>
<span class='fu'>multi_aln</span>(<span class='kw'>file</span>    <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>tool</span>    <span class='kw'>=</span> <span class='st'>"mafft"</span>,
          <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
          <span class='kw'>params</span>  <span class='kw'>=</span> <span class='st'>"--maxiterate 1 --clustalout"</span>)


<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#note">Note</a></li>

      <li><a href="#references">References</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Hajk-Georg Drost and Sarah Scharfenberg</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

