<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Main Orthology Inference Function — orthologs • orthologr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Main Orthology Inference Function — orthologs" />

<meta property="og:description" content="This function takes nucleotide or protein sequences for a set of organisms 
and performs orthology inference to detect orthologous genes within the given organisms
based on selected orthology inference programs." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Main Orthology Inference Function</h1>
    <small class="dont-index">Source: <a href='https://github.com/HajkD/orthologr/blob/master/R/orthologs.R'><code>R/orthologs.R</code></a></small>
    <div class="hidden name"><code>orthologs.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function takes nucleotide or protein sequences for a set of organisms 
and performs orthology inference to detect orthologous genes within the given organisms
based on selected orthology inference programs.</p>
    
    </div>

    <pre class="usage"><span class='fu'>orthologs</span>(<span class='no'>query_file</span>, <span class='no'>subject_files</span>, <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>,
  <span class='kw'>outgroup_file</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>eval</span> <span class='kw'>=</span> <span class='st'>"1E-5"</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"fasta"</span>,
  <span class='kw'>ortho_detection</span> <span class='kw'>=</span> <span class='st'>"RBH"</span>, <span class='kw'>delete_corrupt_cds</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>cdd.path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>add_params</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>comp_cores</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>clean_folders</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>query_file</th>
      <td><p>a character string specifying the path to the sequence file of interest (query organism).</p></td>
    </tr>
    <tr>
      <th>subject_files</th>
      <td><p>a character string specifying the paths to the sequence files of interest (subject organisms).
Different orthology inference methods can detect orthologs using multiple subject organisms, e.g. "OrthoMCL", and "PO" (ProteinOrtho).</p></td>
    </tr>
    <tr>
      <th>seq_type</th>
      <td><p>a character string specifying the sequence type stored in the input file.
Options are are: "cds", "protein", or "dna". In case of "cds", sequence are translated to protein sequences,
in case of "dna", cds prediction is performed on the corresponding sequences which subsequently are
translated to protein sequences. Default is <code>seq_type</code> = "protein".</p></td>
    </tr>
    <tr>
      <th>outgroup_file</th>
      <td><p>a character string specifying the paths to the sequence files of interest (outgroup organisms).
This argument is only used by <code>InParanoid</code>.</p></td>
    </tr>
    <tr>
      <th>eval</th>
      <td><p>a numeric value specifying the E-Value cutoff for BLAST hit detection.</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p>a character string specifying the file format of the sequence file, e.g. "fasta", "gbk". Default is "fasta".</p></td>
    </tr>
    <tr>
      <th>ortho_detection</th>
      <td><p>a character string specifying the orthology inference method that shall be performed
to detect orthologous genes. Default is <code>ortho_detection</code> = "RBH" (BLAST reciprocal best hit).
Further methods are: "RBH" (BLAST reciprocal best hit), "PO" (ProteinOrtho), and "OrthoMCL.</p></td>
    </tr>
    <tr>
      <th>delete_corrupt_cds</th>
      <td><p>a logical value indicating whether sequences with corrupt base triplets should be removed from the input <code>file</code>. This is the case when the length of coding sequences cannot be divided by 3 and thus the coding sequence contains at least one corrupt base triplet.</p></td>
    </tr>
    <tr>
      <th>cdd.path</th>
      <td><p>path to the cdd database folder (specify when using <code>ortho_detection</code> = <code>"DELTA"</code>).</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>a character string specifying the path to the corresponding orthology inference tool.
For "BH" and "RBH": path to BLAST, "PO": path to ProteinOrtho 5.07, "OrthoMCL": path to OrthoMCL.</p></td>
    </tr>
    <tr>
      <th>add_params</th>
      <td><p>a character string specifying additional parameters that shall be handed to the orthology inference method (tool).
Default is <code>add_params</code> = <code>NULL</code>.</p></td>
    </tr>
    <tr>
      <th>comp_cores</th>
      <td><p>a numeric value specifying the number of cores to be used for multicore computations.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>a logical value specifying whether a successful interface call shall be printed out.</p></td>
    </tr>
    <tr>
      <th>clean_folders</th>
      <td><p>a boolean value spefiying whether all internall folders storing the output of used programs
shall be removed. Default is <code>clean_folders</code> = <code>FALSE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A data.table storing the query_ids of orthologous genes in the first column, the subject_ids of orthologous genes
in the second column and the amino acid sequences in the third column.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function takes sequence files of a query organism and a subject organism and performs orthology inference
using a defined orthology inference method to dectect orthologous genes.</p>
<p>The following interfaces are implemented in the <code>orthologs</code> function:</p> 
<p>BLAST based methods:</p>
<ul>
<li><p>BLAST best hit (BH)</p></li>
<li><p>BLAST reciprocal best hit (RBH)</p></li>
<li><p>DELTA-BLAST reciprocal best hit (DELTA)</p></li>
</ul>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>BLAST: http://blast.ncbi.nlm.nih.gov/blastcgihelp.shtml</p>
<p>ProteinOrtho: https://www.bioinf.uni-leipzig.de/Software/proteinortho/</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='blast_rec.html'>blast_rec</a></code>, <code><a href='dNdS.html'>dNdS</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>

<span class='co'>### BLAST Best Hit</span>

<span class='co'># perform orthology inference using BLAST best hit</span>
<span class='co'># and fasta sequence files storing protein sequences</span>
<span class='fu'>orthologs</span>(<span class='kw'>query_file</span>      <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>subject_files</span>   <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>seq_type</span>        <span class='kw'>=</span> <span class='st'>"protein"</span>,
          <span class='kw'>ortho_detection</span> <span class='kw'>=</span> <span class='st'>"BH"</span>)


<span class='co'>### BLAST Reciprocal Best Hit</span>

<span class='co'># perform orthology inference using BLAST reciprocal best hit</span>
<span class='co'># and fasta sequence files storing protein sequences</span>
<span class='fu'>orthologs</span>(<span class='kw'>query_file</span>      <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>subject_files</span>   <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>seq_type</span>        <span class='kw'>=</span> <span class='st'>"protein"</span>,
          <span class='kw'>ortho_detection</span> <span class='kw'>=</span> <span class='st'>"RBH"</span>)


<span class='co'># multicore version          </span>
<span class='fu'>orthologs</span>(<span class='kw'>query_file</span>      <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>subject_files</span>   <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
          <span class='kw'>seq_type</span>        <span class='kw'>=</span> <span class='st'>"protein"</span>,
          <span class='kw'>ortho_detection</span> <span class='kw'>=</span> <span class='st'>"RBH"</span>,
          <span class='kw'>comp_cores</span>      <span class='kw'>=</span> <span class='fl'>2</span>)



<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Hajk-Georg Drost</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

