<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compute Pairwise Alignments — pairwise_aln • orthologr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Compute Pairwise Alignments — pairwise_aln" />

<meta property="og:description" content="This function takes a FASTA file containing two DNA or amino acid sequences
that shall be aligned and computes a paiwise alignment using a defined alignment method." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute Pairwise Alignments</h1>
    <small class="dont-index">Source: <a href='https://github.com/HajkD/orthologr/blob/master/R/pairwise_aln.R'><code>R/pairwise_aln.R</code></a></small>
    <div class="hidden name"><code>pairwise_aln.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function takes a FASTA file containing two DNA or amino acid sequences
that shall be aligned and computes a paiwise alignment using a defined alignment method.</p>
    
    </div>

    <pre class="usage"><span class='fu'>pairwise_aln</span>(<span class='no'>file</span>, <span class='kw'>tool</span> <span class='kw'>=</span> <span class='st'>"NW"</span>, <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>,
  <span class='kw'>get_aln</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>pairwise_aln_name</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>quiet</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>file</th>
      <td><p>a character string specifying the path to the file storing the sequences in FASTA format.</p></td>
    </tr>
    <tr>
      <th>tool</th>
      <td><p>a character string specifying the program/algorithm that should be used: <code>tool</code> = <code>"NW"</code>.</p></td>
    </tr>
    <tr>
      <th>seq_type</th>
      <td><p>a character string specifying the sequence type stored within the given FASTA file. Options are
<code>seq_type</code> = <code>"protein"</code>, <code>seq_type</code> = <code>"cds"</code>, <code>seq_type</code> = <code>"dna"</code>.
 Default is <code>seq_type</code> = <code>"protein"</code>.</p></td>
    </tr>
    <tr>
      <th>get_aln</th>
      <td><p>a logical value indicating whether the produced alignment should be returned.</p></td>
    </tr>
    <tr>
      <th>pairwise_aln_name</th>
      <td><p>a character string specifying the name of the stored alignment file. 
Default is <code>pairwise_aln_name</code> = <code>NULL</code> denoting a default name: 'toolname_seq_type.aln' .</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>a character string specifying the path to the pairwise alignment program (in case you don't use the default path).</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>a logical value specifying whether a successful interface call shall be printed out.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>In case the argument <code>get_aln</code> is set <code>TRUE</code>, an object of class alignment of the seqinr package is returned.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function provides an interface between R and common pairwise alignment computation methods.</p>
<p>The current version of this function computes pairwise alignments based on the <code><a href='https://www.rdocumentation.org/packages/Biostrings/topics/pairwiseAlignment'>pairwiseAlignment</a></code>
function implemented in the <span class="pkg">Biostrings</span> package.</p>
<p>The default pairwise alignment method is based on the <em>Needleman-Wunsch Algorithm</em>.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>

<span class='co'># Needleman-Wunsch Example:  </span>

<span class='co'># in case Biostrings works properly</span>
<span class='fu'>pairwise_aln</span>( <span class='kw'>file</span>     <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/aa_seqs.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
              <span class='kw'>tool</span>     <span class='kw'>=</span> <span class='st'>"NW"</span>,
              <span class='kw'>get_aln</span>  <span class='kw'>=</span> <span class='fl'>TRUE</span>,
              <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>)



<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Sarah Scharfenberg and Hajk-Georg Drost</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

