<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Orthology Inference with ProteinOrtho — ProteinOrtho • orthologr</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Orthology Inference with ProteinOrtho — ProteinOrtho" />

<meta property="og:description" content="This function takes a query organism and a set of subject organisms
and performs orthology inference using ProteinOrtho as orthology detection program (methodology)." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">dNdS estimation using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Orthology Inference with ProteinOrtho</h1>
    <small class="dont-index">Source: <a href='https://github.com/HajkD/orthologr/blob/master/R/ProteinOrtho.R'><code>R/ProteinOrtho.R</code></a></small>
    <div class="hidden name"><code>ProteinOrtho.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function takes a query organism and a set of subject organisms
and performs orthology inference using ProteinOrtho as orthology detection program (methodology).</p>
    
    </div>

    <pre class="usage"><span class='fu'>ProteinOrtho</span>(<span class='no'>query_file</span>, <span class='no'>subject_files</span>, <span class='kw'>po_params</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>eval</span> <span class='kw'>=</span> <span class='st'>"1E-5"</span>, <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>, <span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"fasta"</span>,
  <span class='kw'>po_path</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>comp_cores</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>delete_files</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>query_file</th>
      <td><p>a character string specifying the path to the sequence file of interest (query organism).</p></td>
    </tr>
    <tr>
      <th>subject_files</th>
      <td><p>a character string specifying the paths to the sequence files of interest (subject organisms).</p></td>
    </tr>
    <tr>
      <th>po_params</th>
      <td><p>a character string specifying additional parameters that shall be handed to the ProteinOrtho call,
e.g. <code>po_params</code> = <code>"-identity=35 -cov=75"</code>. See https://www.bioinf.uni-leipzig.de/Software/proteinortho/manual.html
for parameter details.</p></td>
    </tr>
    <tr>
      <th>eval</th>
      <td><p>a numeric value specifying the E-Value cutoff for BLAST hit detection.</p></td>
    </tr>
    <tr>
      <th>seq_type</th>
      <td><p>a character string specifying the sequence type stored in the input file.
Options are: "protein", or "dna". ProteinOrtho can only handle protein or nucleotide sequences stored in fasta files.</p></td>
    </tr>
    <tr>
      <th>format</th>
      <td><p>when using ProteinOrtho you can only specify <code>format</code> = <code>"fasta"</code>.</p></td>
    </tr>
    <tr>
      <th>po_path</th>
      <td><p>a character string specifying the execution path to the <code>proteinortho5.pl</code> file (in case it is not sored
in the standard execution path).</p></td>
    </tr>
    <tr>
      <th>comp_cores</th>
      <td><p>a numeric value specifying the number of cores that shall be
used to run ProteinOrtho.</p></td>
    </tr>
    <tr>
      <th>delete_files</th>
      <td><p>a boolean value specifying whether the folder '_ProteinOrtho' that stored the
ProteinOrtho output files shall be removed after the analysis. Default is <code>delete_files</code> = <code>FALSE</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>a data.frame storing the ProteinOrtho output.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>This function provides an interface to the ProteinOrtho program
and performs orthology inference using the ProteinOrtho methodology.</p>
<p>ProteinOrtho is a orthology inference program to detect orthologous genes within different species. 
For this purpose it compares similarities of given gene sequences and clusters them to find significant groups. 
To enhance the prediction accuracy, the relative order of genes (synteny) can be used as additional feature for the discrimination of orthologs.
(source: https://www.bioinf.uni-leipzig.de/Software/proteinortho/manual.html).</p>
    
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p>Lechner M et al. (2011) Proteinortho: Detection of (Co-)orthologs in large-scale analysis. BMC Bioinformatics 2011, 12:124.</p>
<p>https://www.bioinf.uni-leipzig.de/Software/proteinortho/manual.html</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='orthologs.html'>orthologs</a></code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'># finding orthologs between Arabidopsis thaliana and Arabidopsis lyrata genes</span>
<span class='fu'>ProteinOrtho</span>(<span class='kw'>query_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>subject_files</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>,<span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"fasta"</span>)



<span class='co'># finding orthologs between Arabidopsis thaliana and Arabidopsis lyrata genes</span>
<span class='co'># using multicore processing</span>
<span class='fu'>ProteinOrtho</span>(<span class='kw'>query_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>subject_files</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>,<span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"fasta"</span>, <span class='kw'>comp_cores</span> <span class='kw'>=</span> <span class='fl'>2</span>)




<span class='co'># it is also possible to run ProteinOrtho using the synteny option</span>
<span class='co'># make sure you can provide a *.gff file</span>
<span class='co'># see https://www.bioinf.uni-leipzig.de/Software/proteinortho/manual.html for details</span>
<span class='fu'>ProteinOrtho</span>(<span class='kw'>query_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>subject_files</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>,<span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"fasta"</span>,<span class='kw'>po_params</span> <span class='kw'>=</span> <span class='st'>"-synteny"</span>)



 <span class='co'># defining 3 subject organisms: A. lyrata, B. rapa, and T. halophila</span>
  <span class='no'>subject_organisms</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/example_brapa_aa.faa'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
                         <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/example_alyra_aa.faa'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
                         <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/example_thalo_aa.faa'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>))

  <span class='fu'>ProteinOrtho</span>(<span class='kw'>query_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/example_athal_aa.faa'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
            <span class='kw'>subject_files</span> <span class='kw'>=</span> <span class='no'>subject_organisms</span>,
            <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>)



<span class='co'># in case you need to specify the BLAST path or want to add </span>
<span class='co'># additional parameters to the BLAST run, please use the 'po_params' argument</span>
<span class='fu'>ProteinOrtho</span>(<span class='kw'>query_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>subject_files</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>,<span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"fasta"</span>,
             <span class='kw'>po_params</span> <span class='kw'>=</span> <span class='st'>"-blastpath='path/to/blastp' -blastParameters='-max_target_seqs 1'"</span>)




<span class='co'># finding orthologs between Arabidopsis thaliana and Arabidopsis lyrata genes</span>
<span class='co'># using "dna" sequences</span>
<span class='fu'>ProteinOrtho</span>(<span class='kw'>query_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>subject_files</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_cds.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"dna"</span>,<span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"fasta"</span>)




<span class='co'># you can also clean up all output files generated by ProteinOrtho</span>
<span class='co'># using the 'delete_files' argument</span>
<span class='fu'>ProteinOrtho</span>(<span class='kw'>query_file</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_thal_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>subject_files</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>'seqs/ortho_lyra_aa.fasta'</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>'orthologr'</span>),
             <span class='kw'>seq_type</span> <span class='kw'>=</span> <span class='st'>"protein"</span>,<span class='kw'>format</span> <span class='kw'>=</span> <span class='st'>"fasta"</span>, <span class='kw'>delete_files</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)



<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#references">References</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Hajk-Georg Drost</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

