#' @title Generate ortholog tables by gene locus and splice varaint
#' @description Given an input dNdS table generated by \code{\link{dNdS}} and
#' annotation files for the query and subject species in \code{gtf} or \code{gff}
#' file format, this function selects the best BLAST hit to represent either a gene locus (e.g. the splice variant of the gene locus with lowest e-value) or the best BLAST hit for a splice varaint.
#' @param dNdS_file file path to a \code{dNdS_tbl} generated with \code{\link{dNdS}} and
#' stored conform with \code{\link{read.dnds.tbl}}.
#' @param annotation_file_query file path to the annotation file of the query species in \code{gtf} or \code{gff} file format.
#' @param annotation_file_subject file path to the annotation file of the subject species in \code{gtf} or \code{gff} file format.
#' @param format a vector of length 2 storing the annotation file formats of the query annotation file and subject annotation file: either \code{gtf} or \code{gff} format. E.g. \code{format = c("gtf","gtf")}.
#' @author Hajk-Georg Drost
#' @export
generate_ortholog_tables <-
        function(dNdS_file,
                 annotation_file_query,
                 annotation_file_subject,
                 format) {
                if (!file.exists(dNdS_file))
                        stop("Please provide a valid path to the dNdS_file.", call. = FALSE)
                
                qry_species_name <-
                        unlist(stringr::str_split(basename(annotation_file_query), "[.]"))[1]
                sbj_species_name <-
                        unlist(stringr::str_split(basename(annotation_file_subject), "[.]"))[1]
                
                select_orthologs_by_gene_locus_output_name <-
                        paste0("orthologs_by_gene_",
                               qry_species_name,
                               "_vs_",
                               sbj_species_name,
                               ".csv")
                
                select_orthologs_by_splice_variant_output_name <-
                        paste0(
                                "orthologs_by_splice_variant_",
                                qry_species_name,
                                "_vs_",
                                sbj_species_name,
                                ".csv"
                        )
                
                # import dNdS file
                dNdS_tbl <-
                        read.dnds.tbl(dNdS_file)
                
                # select orthologs by gene_locus
                select_orthologs_by_gene_locus <- select_orthologs(
                        dnds_tbl = dNdS_tbl,
                        annotation_file_query = annotation_file_query,
                        annotation_file_subject = annotation_file_subject,
                        collapse_by = "gene_locus",
                        format = format
                )
                
                # select orthologs by splice_variant
                select_orthologs_splice_variant <- select_orthologs(
                        dnds_tbl = dNdS_tbl,
                        annotation_file_query = annotation_file_query,
                        annotation_file_subject = annotation_file_subject,
                        collapse_by = "splice_variant",
                        format = format
                )
                
                by_gene_output_folder <- paste0("orthologs_by_gene_locus_qry_", stringr::str_to_lower(qry_species_name))
                by_splice_variant_output_folder <- paste0("orthologs_by_splice_variant_qry_", stringr::str_to_lower(qry_species_name))
                
                message("\n")
                message(
                        "Store ortholog by gene locus table at '",
                        file.path(by_gene_output_folder, select_orthologs_by_gene_locus_output_name),
                        "'."
                )
                
                
                if (!dir.exists(by_gene_output_folder))
                        dir.create(by_gene_output_folder)
                
                if (!dir.exists(by_splice_variant_output_folder))
                        dir.create(by_splice_variant_output_folder)
                
                
                readr::write_delim(
                        select_orthologs_by_gene_locus,
                        file.path(by_gene_output_folder, select_orthologs_by_gene_locus_output_name),
                        col_names = TRUE,
                        delim = ";"
                )
                
                message("\n")
                message(
                        "Store ortholog by splice variant table at '",
                        file.path(by_splice_variant_output_folder, select_orthologs_by_splice_variant_output_name),
                        "'."
                )
                readr::write_delim(
                        select_orthologs_splice_variant,
                        file.path(by_splice_variant_output_folder, select_orthologs_by_splice_variant_output_name),
                        col_names = TRUE,
                        delim = ";"
                )
        }